{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Creating the \"PREDICT\" dataset based on various data sources"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Version 1.0.0 (December 28th 2022). Please run the notebooks \"FEATURELESS_dataset-v1.0.0.ipynb\" and \"TRANSCRIPT_dataset-v1.0.0\" beforehand."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Librairies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess as sb\n",
    "import os\n",
    "import pickle\n",
    "\n",
    "from multiprocessing import cpu_count\n",
    "from joblib import Parallel, delayed\n",
    "from time import time\n",
    "from itertools import product\n",
    "n_jobs=cpu_count()-2\n",
    "assert n_jobs > 0 and n_jobs < cpu_count()\n",
    "parallel = Parallel(n_jobs=n_jobs, backend='loky')\n",
    "\n",
    "import sys\n",
    "sys.path.insert(0, \"../../src/\")\n",
    "\n",
    "import paths_global\n",
    "import utils\n",
    "import data_processing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Local paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "root_folder=\"/media/kali/1b80f30d-2803-4260-a792-9ae206084252/Code/M30/data/\"\n",
      "data_folder=\"/media/kali/1b80f30d-2803-4260-a792-9ae206084252/RECeSS/cfdr/data/\"\n"
     ]
    }
   ],
   "source": [
    "## Where database files are stored\n",
    "print('root_folder=\"%s\"' % paths_global.root_folder)\n",
    "## Where intermediary files are stored\n",
    "print('data_folder=\"%s\"' % paths_global.data_folder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "predict_folder=\"/media/kali/1b80f30d-2803-4260-a792-9ae206084252/RECeSS/cfdr/data/PREDICT_v1.0.0/\"\n"
     ]
    }
   ],
   "source": [
    "predict_folder = paths_global.data_folder+\"PREDICT_v1.0.0/\"\n",
    "sb.Popen([\"mkdir\", \"-p\", predict_folder])\n",
    "## Where PREDICT dataset files are stored\n",
    "print('predict_folder=\"%s\"' % predict_folder)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Drug and disease identifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert os.path.exists(paths_global.data_folder+\"drugbankid2drugname.pck\")\n",
    "with open(paths_global.data_folder+\"drugbankid2drugname.pck\", \"rb\") as f:\n",
    "    di_drugbankid2drugname = pickle.load(f)\n",
    "    \n",
    "assert os.path.exists(paths_global.data_folder+\"omimid2diseasename.pck\")\n",
    "with open(paths_global.data_folder+\"omimid2diseasename.pck\", \"rb\") as f:\n",
    "    di_omimid2diseasename = pickle.load(f)\n",
    "    \n",
    "cids_file = paths_global.data_folder+\"medgenid2diseasename.pck\"\n",
    "if (not os.path.exists(cids_file)):\n",
    "    di_medgenid2diseasename = {}\n",
    "else:\n",
    "    with open(cids_file, \"rb\") as f:\n",
    "        di_medgenid2diseasename = pickle.load(f)\n",
    "        \n",
    "pubchem_file = paths_global.data_folder+\"pubchemid2drugname.pck\"\n",
    "if (not os.path.exists(pubchem_file)):\n",
    "    di_pubchemid2drugname = {}\n",
    "else:\n",
    "    with open(pubchem_file, \"rb\") as f:\n",
    "        di_pubchemid2drugname = pickle.load(f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## I. Matrix A : $N_S \\times N_D$ of drug-disease associations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>C1851649</th>\n",
       "      <th>C0042133</th>\n",
       "      <th>C5193005</th>\n",
       "      <th>C2676676</th>\n",
       "      <th>C1704272</th>\n",
       "      <th>C4722327</th>\n",
       "      <th>C1858361</th>\n",
       "      <th>C2676676.1</th>\n",
       "      <th>C4310232</th>\n",
       "      <th>C0029456</th>\n",
       "      <th>...</th>\n",
       "      <th>C0242770</th>\n",
       "      <th>C1880129</th>\n",
       "      <th>C0022661</th>\n",
       "      <th>C0236792</th>\n",
       "      <th>C1135191</th>\n",
       "      <th>C0149516</th>\n",
       "      <th>C1835407</th>\n",
       "      <th>C0016667</th>\n",
       "      <th>C0039445</th>\n",
       "      <th>C5203670</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>657181</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DB00010</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5311128</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DB00017</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5311065</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5917</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>442872</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>442021</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4843</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DB12466</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1601 rows × 1599 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         C1851649  C0042133  C5193005  C2676676  C1704272  C4722327  C1858361   \n",
       "657181        1.0       1.0       1.0       1.0       1.0       1.0       1.0  \\\n",
       "DB00010       0.0       0.0       0.0       0.0       0.0       0.0       1.0   \n",
       "5311128       1.0       0.0       0.0       1.0       0.0       1.0       0.0   \n",
       "DB00017       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "5311065       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "...           ...       ...       ...       ...       ...       ...       ...   \n",
       "5917          0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "442872        0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "442021        0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "4843          0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "DB12466       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "\n",
       "         C2676676.1  C4310232  C0029456  ...  C0242770  C1880129  C0022661   \n",
       "657181          0.0       0.0       0.0  ...       0.0       0.0       0.0  \\\n",
       "DB00010         0.0       0.0       0.0  ...       0.0       0.0       0.0   \n",
       "5311128         1.0       0.0       0.0  ...       0.0       0.0       0.0   \n",
       "DB00017         0.0       1.0       1.0  ...       0.0       0.0       0.0   \n",
       "5311065         0.0       0.0       0.0  ...       0.0       0.0       0.0   \n",
       "...             ...       ...       ...  ...       ...       ...       ...   \n",
       "5917            0.0       0.0       0.0  ...       0.0       0.0       0.0   \n",
       "442872          0.0       0.0       0.0  ...       0.0       0.0       0.0   \n",
       "442021          0.0       0.0       0.0  ...       0.0       0.0       0.0   \n",
       "4843            0.0       0.0       0.0  ...       0.0       0.0       0.0   \n",
       "DB12466         0.0       0.0       0.0  ...       0.0       0.0       0.0   \n",
       "\n",
       "         C0236792  C1135191  C0149516  C1835407  C0016667  C0039445  C5203670  \n",
       "657181        0.0       0.0       0.0       0.0       0.0       0.0       0.0  \n",
       "DB00010       0.0       0.0       0.0       0.0       0.0       0.0       0.0  \n",
       "5311128       0.0       0.0       0.0       0.0       0.0       0.0       0.0  \n",
       "DB00017       0.0       0.0       0.0       0.0       0.0       0.0       0.0  \n",
       "5311065       0.0       0.0       0.0       0.0       0.0       0.0       0.0  \n",
       "...           ...       ...       ...       ...       ...       ...       ...  \n",
       "5917          0.0       0.0       0.0       0.0       0.0       0.0       0.0  \n",
       "442872        0.0       0.0       0.0       0.0       0.0       0.0       0.0  \n",
       "442021        0.0       0.0       0.0       0.0       0.0       0.0       0.0  \n",
       "4843          0.0       0.0       0.0       0.0       0.0       0.0       0.0  \n",
       "DB12466       0.0       0.0       0.0       0.0       0.0       0.0      -1.0  \n",
       "\n",
       "[1601 rows x 1599 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "A = pd.read_csv(paths_global.data_folder+\"FEATURELESS_v1.0.0/all_ratings.csv\", index_col=0)\n",
    "A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sparsity = 0.35070326199346175%\n",
      "Ndrugs=1599\tNdiseases=1599\n",
      "8658 positive\t320 negative\t2547823 unknown matchings\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>8968</th>\n",
       "      <th>8969</th>\n",
       "      <th>8970</th>\n",
       "      <th>8971</th>\n",
       "      <th>8972</th>\n",
       "      <th>8973</th>\n",
       "      <th>8974</th>\n",
       "      <th>8975</th>\n",
       "      <th>8976</th>\n",
       "      <th>8977</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ind_id</th>\n",
       "      <td>C1851649</td>\n",
       "      <td>C0042133</td>\n",
       "      <td>C5193005</td>\n",
       "      <td>C2676676</td>\n",
       "      <td>C1704272</td>\n",
       "      <td>C4722327</td>\n",
       "      <td>C1858361</td>\n",
       "      <td>C0034013</td>\n",
       "      <td>C0014175</td>\n",
       "      <td>C1858361</td>\n",
       "      <td>...</td>\n",
       "      <td>C0014544</td>\n",
       "      <td>C0014544</td>\n",
       "      <td>C0014544</td>\n",
       "      <td>C0014544</td>\n",
       "      <td>C0014544</td>\n",
       "      <td>C0014544</td>\n",
       "      <td>C0014544</td>\n",
       "      <td>C0014544</td>\n",
       "      <td>C0014544</td>\n",
       "      <td>C5203670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>drug_id</th>\n",
       "      <td>657181</td>\n",
       "      <td>657181</td>\n",
       "      <td>657181</td>\n",
       "      <td>657181</td>\n",
       "      <td>657181</td>\n",
       "      <td>657181</td>\n",
       "      <td>657181</td>\n",
       "      <td>657181</td>\n",
       "      <td>657181</td>\n",
       "      <td>DB00010</td>\n",
       "      <td>...</td>\n",
       "      <td>2310</td>\n",
       "      <td>492405</td>\n",
       "      <td>4375</td>\n",
       "      <td>3352</td>\n",
       "      <td>3373</td>\n",
       "      <td>5917</td>\n",
       "      <td>442872</td>\n",
       "      <td>442021</td>\n",
       "      <td>4843</td>\n",
       "      <td>DB12466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rating</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 8978 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             0         1         2         3         4         5         6      \n",
       "ind_id   C1851649  C0042133  C5193005  C2676676  C1704272  C4722327  C1858361  \\\n",
       "drug_id    657181    657181    657181    657181    657181    657181    657181   \n",
       "rating        1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "\n",
       "             7         8         9     ...      8968      8969      8970   \n",
       "ind_id   C0034013  C0014175  C1858361  ...  C0014544  C0014544  C0014544  \\\n",
       "drug_id    657181    657181   DB00010  ...      2310    492405      4375   \n",
       "rating        1.0       1.0       1.0  ...      -1.0      -1.0      -1.0   \n",
       "\n",
       "             8971      8972      8973      8974      8975      8976      8977  \n",
       "ind_id   C0014544  C0014544  C0014544  C0014544  C0014544  C0014544  C5203670  \n",
       "drug_id      3352      3373      5917    442872    442021      4843   DB12466  \n",
       "rating       -1.0      -1.0      -1.0      -1.0      -1.0       1.0      -1.0  \n",
       "\n",
       "[3 rows x 8978 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ratings_A = utils.matrix2ratings(A, \"ind_id\", \"drug_id\", \"rating\")\n",
    "\n",
    "print(\"Sparsity = \"+str(utils.compute_sparsity(A))+\"%\")\n",
    "utils.print_dataset(ratings_A, \"ind_id\", \"drug_id\", \"rating\")\n",
    "ratings_A.T"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## II. Drug-drug similarity matrix\n",
    "\n",
    "We build a larger version of the similarity matrices mentioned in the [PREDICT](https://dx.doi.org/10.1038%2Fmsb.2011.26) paper."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### II.1. Drug-drug similarity: Chemical drug-drug similarities (from structures in DrugBank or in PubChem)\n",
    "\n",
    "This yields a dataframe `chemical_similarity_df` of size #drugs x #drugs, where identifiers are PubChem or DrugBank identifiers, which contains, when computable, drug-pairwise Tanimoto scores between SMILES fingerprints."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "chemical_similarity_file = predict_folder+\"chemical_similarity_df.csv\"\n",
    "\n",
    "if (not os.path.exists(chemical_similarity_file)):\n",
    "    drugbank_chem_file = paths_global.drugbank_folder+\"STRUCTURES/Structure External Links/structure links.csv\"\n",
    "    if (not os.path.exists(drugbank_chem_file)):\n",
    "        sb.call(\"unzip \"+path+\"drugbank_all_structure_links.csv.zip\", shell=True)\n",
    "    chemical_based = pd.read_csv(drugbank_chem_file, sep=\",\", index_col=0, header=0)[[\"SMILES\"]]\n",
    "    chemical_based = chemical_based.dropna()\n",
    "    ndrugs=chemical_based.shape[0]\n",
    "    \n",
    "    def job_function(i, j, run_id=None, seed=None):\n",
    "        if (str(seed)!=\"None\"):\n",
    "            np.random.seed(seed)\n",
    "        if (i<j):\n",
    "            return 0.\n",
    "        elif (i==j):\n",
    "            return 1.\n",
    "        else:\n",
    "            return data_processing.smiles_similarity(chemical_based.values[i,0], chemical_based.values[j,0])\n",
    "\n",
    "    start = time()\n",
    "    print(\"%d drugs, %d jobs\" % (ndrugs, n_jobs))\n",
    "    if (n_jobs==1):\n",
    "        chemical_similarity = [job_function(i, j) for i, j in product(range(ndrugs), range(ndrugs))]\n",
    "    else:\n",
    "        seeds = [np.random.randint(int(10e6)) for _ in range(ndrugs*ndrugs)]\n",
    "        chemical_similarity = parallel(delayed(job_function)(run_id//ndrugs, run_id%ndrugs, run_id, seed) for run_id, seed in enumerate(seeds))\n",
    "    \n",
    "    chemical_similarity = np.array(chemical_similarity).reshape((ndrugs, ndrugs))\n",
    "    chemical_similarity += chemical_similarity.T\n",
    "    chemical_similarity[range(ndrugs), range(ndrugs)] = 1.\n",
    "    end = time()-start\n",
    "\n",
    "    drug_index=chemical_based.index  \n",
    "    chemical_similarity_df = pd.DataFrame(chemical_similarity, index=drug_index, columns=drug_index)\n",
    "    chemical_similarity_df.to_csv(chemical_similarity_file)\n",
    "    print(\"Computation took %.2f sec. on %d jobs\" % (end, n_jobs))\n",
    "\n",
    "    drugnames = [di_drugbankid2drugname.get(cid, None) for cid in chemical_similarity_df.index]\n",
    "    di_drugname2pubchemid = {di_pubchemid2drugname[k] : k for k in di_pubchemid2drugname}\n",
    "    pubchem_ids = [di_drugname2pubchemid.get(d, None) for d in drugnames]\n",
    "    drug_ids = [ip for ip, p in enumerate(pubchem_ids) if (p)]\n",
    "    chemical_custom = chemical_similarity_df.iloc[drug_ids,drug_ids]\n",
    "    chemical_custom.index = [p for p in pubchem_ids if (p)]\n",
    "    chemical_custom.columns = chemical_custom.index\n",
    "\n",
    "    chemical_custom.index = chemical_custom.index.astype(str)\n",
    "    chemical_custom.columns = chemical_custom.columns.astype(str)\n",
    "\n",
    "    chemical_custom.to_csv(chemical_similarity_file)\n",
    "\n",
    "chemical_custom = pd.read_csv(chemical_similarity_file, index_col=0, header=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### II.2. Drug-drug similarity: Side effect drug-drug similarities (from reported SE in SIDER)\n",
    "\n",
    "This yields a dataframe `se_similarity_df` of size #drugs x #drugs, where identifiers are PubChem or DrugBank identifiers, which contains, when computable, drug-pairwise Jaccard scores between lists of SIDER-reported side effects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>5280453</th>\n",
       "      <th>6238</th>\n",
       "      <th>450503</th>\n",
       "      <th>15950376</th>\n",
       "      <th>1727</th>\n",
       "      <th>4775</th>\n",
       "      <th>3406</th>\n",
       "      <th>4075</th>\n",
       "      <th>3385</th>\n",
       "      <th>137</th>\n",
       "      <th>...</th>\n",
       "      <th>16135415</th>\n",
       "      <th>35370</th>\n",
       "      <th>60490</th>\n",
       "      <th>23994</th>\n",
       "      <th>60854</th>\n",
       "      <th>60857</th>\n",
       "      <th>5732</th>\n",
       "      <th>5734</th>\n",
       "      <th>5735</th>\n",
       "      <th>5311507</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5280453</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.112903</td>\n",
       "      <td>0.042553</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>0.149123</td>\n",
       "      <td>0.099237</td>\n",
       "      <td>0.135593</td>\n",
       "      <td>0.138743</td>\n",
       "      <td>0.143541</td>\n",
       "      <td>0.109677</td>\n",
       "      <td>...</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.161435</td>\n",
       "      <td>0.176471</td>\n",
       "      <td>0.032967</td>\n",
       "      <td>0.125326</td>\n",
       "      <td>0.163701</td>\n",
       "      <td>0.145897</td>\n",
       "      <td>0.173010</td>\n",
       "      <td>0.164234</td>\n",
       "      <td>0.172566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6238</th>\n",
       "      <td>0.112903</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.080000</td>\n",
       "      <td>0.145833</td>\n",
       "      <td>0.208333</td>\n",
       "      <td>0.086957</td>\n",
       "      <td>0.111111</td>\n",
       "      <td>0.056757</td>\n",
       "      <td>0.107955</td>\n",
       "      <td>0.113043</td>\n",
       "      <td>...</td>\n",
       "      <td>0.089888</td>\n",
       "      <td>0.085859</td>\n",
       "      <td>0.156627</td>\n",
       "      <td>0.020408</td>\n",
       "      <td>0.060274</td>\n",
       "      <td>0.076046</td>\n",
       "      <td>0.081169</td>\n",
       "      <td>0.068841</td>\n",
       "      <td>0.074219</td>\n",
       "      <td>0.105000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>450503</th>\n",
       "      <td>0.042553</td>\n",
       "      <td>0.080000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.093023</td>\n",
       "      <td>0.052632</td>\n",
       "      <td>0.063830</td>\n",
       "      <td>0.017391</td>\n",
       "      <td>0.040268</td>\n",
       "      <td>0.060241</td>\n",
       "      <td>...</td>\n",
       "      <td>0.033557</td>\n",
       "      <td>0.023392</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014620</td>\n",
       "      <td>0.025316</td>\n",
       "      <td>0.024476</td>\n",
       "      <td>0.024096</td>\n",
       "      <td>0.017316</td>\n",
       "      <td>0.034286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15950376</th>\n",
       "      <td>0.031250</td>\n",
       "      <td>0.145833</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.051724</td>\n",
       "      <td>0.085106</td>\n",
       "      <td>0.014409</td>\n",
       "      <td>0.033113</td>\n",
       "      <td>0.022989</td>\n",
       "      <td>...</td>\n",
       "      <td>0.033333</td>\n",
       "      <td>0.023256</td>\n",
       "      <td>0.075472</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.014577</td>\n",
       "      <td>0.029536</td>\n",
       "      <td>0.027972</td>\n",
       "      <td>0.019920</td>\n",
       "      <td>0.021645</td>\n",
       "      <td>0.028249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1727</th>\n",
       "      <td>0.149123</td>\n",
       "      <td>0.208333</td>\n",
       "      <td>0.093023</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.134146</td>\n",
       "      <td>0.152778</td>\n",
       "      <td>0.103448</td>\n",
       "      <td>0.153374</td>\n",
       "      <td>0.152381</td>\n",
       "      <td>...</td>\n",
       "      <td>0.198718</td>\n",
       "      <td>0.162011</td>\n",
       "      <td>0.219178</td>\n",
       "      <td>0.023810</td>\n",
       "      <td>0.095101</td>\n",
       "      <td>0.131148</td>\n",
       "      <td>0.120275</td>\n",
       "      <td>0.116279</td>\n",
       "      <td>0.107438</td>\n",
       "      <td>0.144385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60857</th>\n",
       "      <td>0.163701</td>\n",
       "      <td>0.076046</td>\n",
       "      <td>0.025316</td>\n",
       "      <td>0.029536</td>\n",
       "      <td>0.131148</td>\n",
       "      <td>0.115830</td>\n",
       "      <td>0.102767</td>\n",
       "      <td>0.309255</td>\n",
       "      <td>0.219048</td>\n",
       "      <td>0.108392</td>\n",
       "      <td>...</td>\n",
       "      <td>0.251634</td>\n",
       "      <td>0.278481</td>\n",
       "      <td>0.144578</td>\n",
       "      <td>0.008439</td>\n",
       "      <td>0.294382</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.380952</td>\n",
       "      <td>0.344444</td>\n",
       "      <td>0.380952</td>\n",
       "      <td>0.261538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5732</th>\n",
       "      <td>0.145897</td>\n",
       "      <td>0.081169</td>\n",
       "      <td>0.024476</td>\n",
       "      <td>0.027972</td>\n",
       "      <td>0.120275</td>\n",
       "      <td>0.104235</td>\n",
       "      <td>0.100334</td>\n",
       "      <td>0.296296</td>\n",
       "      <td>0.189041</td>\n",
       "      <td>0.098802</td>\n",
       "      <td>...</td>\n",
       "      <td>0.288690</td>\n",
       "      <td>0.250689</td>\n",
       "      <td>0.127946</td>\n",
       "      <td>0.003472</td>\n",
       "      <td>0.282787</td>\n",
       "      <td>0.380952</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.315271</td>\n",
       "      <td>0.389189</td>\n",
       "      <td>0.284916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5734</th>\n",
       "      <td>0.173010</td>\n",
       "      <td>0.068841</td>\n",
       "      <td>0.024096</td>\n",
       "      <td>0.019920</td>\n",
       "      <td>0.116279</td>\n",
       "      <td>0.157692</td>\n",
       "      <td>0.098113</td>\n",
       "      <td>0.330337</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>0.111486</td>\n",
       "      <td>...</td>\n",
       "      <td>0.312292</td>\n",
       "      <td>0.308176</td>\n",
       "      <td>0.120755</td>\n",
       "      <td>0.008032</td>\n",
       "      <td>0.348624</td>\n",
       "      <td>0.344444</td>\n",
       "      <td>0.315271</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.391813</td>\n",
       "      <td>0.278788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5735</th>\n",
       "      <td>0.164234</td>\n",
       "      <td>0.074219</td>\n",
       "      <td>0.017316</td>\n",
       "      <td>0.021645</td>\n",
       "      <td>0.107438</td>\n",
       "      <td>0.080769</td>\n",
       "      <td>0.106122</td>\n",
       "      <td>0.273942</td>\n",
       "      <td>0.193651</td>\n",
       "      <td>0.091873</td>\n",
       "      <td>...</td>\n",
       "      <td>0.271186</td>\n",
       "      <td>0.265176</td>\n",
       "      <td>0.139918</td>\n",
       "      <td>0.004348</td>\n",
       "      <td>0.285068</td>\n",
       "      <td>0.380952</td>\n",
       "      <td>0.389189</td>\n",
       "      <td>0.391813</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.244582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5311507</th>\n",
       "      <td>0.172566</td>\n",
       "      <td>0.105000</td>\n",
       "      <td>0.034286</td>\n",
       "      <td>0.028249</td>\n",
       "      <td>0.144385</td>\n",
       "      <td>0.140704</td>\n",
       "      <td>0.124352</td>\n",
       "      <td>0.207459</td>\n",
       "      <td>0.150000</td>\n",
       "      <td>0.138393</td>\n",
       "      <td>...</td>\n",
       "      <td>0.234615</td>\n",
       "      <td>0.243636</td>\n",
       "      <td>0.143590</td>\n",
       "      <td>0.005682</td>\n",
       "      <td>0.281796</td>\n",
       "      <td>0.261538</td>\n",
       "      <td>0.284916</td>\n",
       "      <td>0.278788</td>\n",
       "      <td>0.244582</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1279 rows × 1279 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           5280453      6238    450503  15950376      1727      4775   \n",
       "5280453   1.000000  0.112903  0.042553  0.031250  0.149123  0.099237  \\\n",
       "6238      0.112903  1.000000  0.080000  0.145833  0.208333  0.086957   \n",
       "450503    0.042553  0.080000  1.000000  0.071429  0.093023  0.052632   \n",
       "15950376  0.031250  0.145833  0.071429  1.000000  0.142857  0.051724   \n",
       "1727      0.149123  0.208333  0.093023  0.142857  1.000000  0.134146   \n",
       "...            ...       ...       ...       ...       ...       ...   \n",
       "60857     0.163701  0.076046  0.025316  0.029536  0.131148  0.115830   \n",
       "5732      0.145897  0.081169  0.024476  0.027972  0.120275  0.104235   \n",
       "5734      0.173010  0.068841  0.024096  0.019920  0.116279  0.157692   \n",
       "5735      0.164234  0.074219  0.017316  0.021645  0.107438  0.080769   \n",
       "5311507   0.172566  0.105000  0.034286  0.028249  0.144385  0.140704   \n",
       "\n",
       "              3406      4075      3385       137  ...  16135415     35370   \n",
       "5280453   0.135593  0.138743  0.143541  0.109677  ...  0.166667  0.161435  \\\n",
       "6238      0.111111  0.056757  0.107955  0.113043  ...  0.089888  0.085859   \n",
       "450503    0.063830  0.017391  0.040268  0.060241  ...  0.033557  0.023392   \n",
       "15950376  0.085106  0.014409  0.033113  0.022989  ...  0.033333  0.023256   \n",
       "1727      0.152778  0.103448  0.153374  0.152381  ...  0.198718  0.162011   \n",
       "...            ...       ...       ...       ...  ...       ...       ...   \n",
       "60857     0.102767  0.309255  0.219048  0.108392  ...  0.251634  0.278481   \n",
       "5732      0.100334  0.296296  0.189041  0.098802  ...  0.288690  0.250689   \n",
       "5734      0.098113  0.330337  0.200000  0.111486  ...  0.312292  0.308176   \n",
       "5735      0.106122  0.273942  0.193651  0.091873  ...  0.271186  0.265176   \n",
       "5311507   0.124352  0.207459  0.150000  0.138393  ...  0.234615  0.243636   \n",
       "\n",
       "             60490     23994     60854     60857      5732      5734   \n",
       "5280453   0.176471  0.032967  0.125326  0.163701  0.145897  0.173010  \\\n",
       "6238      0.156627  0.020408  0.060274  0.076046  0.081169  0.068841   \n",
       "450503    0.076923  0.000000  0.014620  0.025316  0.024476  0.024096   \n",
       "15950376  0.075472  0.100000  0.014577  0.029536  0.027972  0.019920   \n",
       "1727      0.219178  0.023810  0.095101  0.131148  0.120275  0.116279   \n",
       "...            ...       ...       ...       ...       ...       ...   \n",
       "60857     0.144578  0.008439  0.294382  1.000000  0.380952  0.344444   \n",
       "5732      0.127946  0.003472  0.282787  0.380952  1.000000  0.315271   \n",
       "5734      0.120755  0.008032  0.348624  0.344444  0.315271  1.000000   \n",
       "5735      0.139918  0.004348  0.285068  0.380952  0.389189  0.391813   \n",
       "5311507   0.143590  0.005682  0.281796  0.261538  0.284916  0.278788   \n",
       "\n",
       "              5735   5311507  \n",
       "5280453   0.164234  0.172566  \n",
       "6238      0.074219  0.105000  \n",
       "450503    0.017316  0.034286  \n",
       "15950376  0.021645  0.028249  \n",
       "1727      0.107438  0.144385  \n",
       "...            ...       ...  \n",
       "60857     0.380952  0.261538  \n",
       "5732      0.389189  0.284916  \n",
       "5734      0.391813  0.278788  \n",
       "5735      1.000000  0.244582  \n",
       "5311507   0.244582  1.000000  \n",
       "\n",
       "[1279 rows x 1279 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "se_similarity_file = predict_folder+\"se_similarity_df.csv\"\n",
    "\n",
    "if (not os.path.exists(se_similarity_file)):\n",
    "    drug_names = pd.read_csv(sider_folder+\"drug_names.tsv\", header=None, sep=\"\\t\", \n",
    "                             index_col=0)\n",
    "    drug_names.columns = [\"drug_name\"]\n",
    "    drug_names = drug_names.to_dict()[\"drug_name\"]\n",
    "    se = pd.read_csv(sider_folder+\"meddra_all_se.tsv\", header=None, sep=\"\\t\", \n",
    "                           index_col=0)\n",
    "    se.index = [drug_names[s] for s in se.index]\n",
    "    se = se[se.columns[[2, 3]]]\n",
    "    se.columns = ['MedDRA SE concept type', 'UMLS concept id for MedDRA term']\n",
    "    ## preferred term, unique\n",
    "    se = se.loc[se['MedDRA SE concept type'] == \"PT\"][se.columns[-1:]]\n",
    "\n",
    "    print(\"%d drugs\" % (len(se.index)))\n",
    "    start = time()\n",
    "    se_based = data_processing.sideeffect_similarity(se)\n",
    "    end = time()-start\n",
    "    print(\"Computation took %.2f sec.\" % end)\n",
    "    \n",
    "    drug_index=se.index\n",
    "    se_similarity_df = pd.DataFrame(se_based, index=drug_index, columns=drug_index)\n",
    "    \n",
    "    ## Standardize drug names\n",
    "    def standard_name_from_SIDER(drug):\n",
    "        if (drug == \"1,25(OH)2D3\"):\n",
    "            drug = \"Calcitriol\"\n",
    "        elif (drug == '5-ASA'):\n",
    "            drug = \"Mesalazine\"\n",
    "        elif (drug == \"18F-FDG\"):\n",
    "            drug = \"Fludeoxyglucose (18F)\"\n",
    "        elif (drug == \"18F-flutemetamol\"):\n",
    "            drug = \"Flutemetamol (18F)\"\n",
    "        elif (drug == \"4-AP\"):\n",
    "            drug = \"Dalfampridine\"\n",
    "        elif (drug == '4-PBA'):\n",
    "            drug = \"Phenylbutyric acid\"\n",
    "        elif (drug == \"4-methylpyrazole\"):\n",
    "            drug = \"Fomepizole\"\n",
    "        elif (drug == \"5-FU\"):\n",
    "            drug = \"Fluorouracil\"\n",
    "        elif (drug == \"5-aminolevulinic\"):\n",
    "            drug = \"Aminolevulinic acid\"\n",
    "        elif (drug == \"5-aza-2'-deoxycytidine\"):\n",
    "            drug = \"Decitabine\"\n",
    "        elif (drug == \"5-azacytidine\"):\n",
    "            drug = \"Azacitidine\"\n",
    "        elif (drug == \"5-fluorocytosine\"):\n",
    "            drug = \"Flucytosine\"\n",
    "        elif (drug == \"5-methyltetrahydrofolate\"):\n",
    "            drug = \"Levomefolic acid\"\n",
    "        elif (drug == \"6-thioguanine\"):\n",
    "            drug = \"Tioguanine\"\n",
    "        elif (drug == \"68Ga\"):\n",
    "            drug = \"Ga 68 PSMA-11\"\n",
    "        elif (drug == \"8-MOP\"):\n",
    "            drug = \"Methoxsalen\"\n",
    "        elif (drug == \"A77\"):\n",
    "            drug = \"Teriflunomide\"\n",
    "        elif (drug == \"ACTH(1-39\"):\n",
    "            drug = \"Corticotropin\"\n",
    "        elif (drug == \"AMD3100\"):\n",
    "            drug = \"Plerixafor\"\n",
    "        elif (drug == \"AN2690\"):\n",
    "            drug = \"Tavaborole\"\n",
    "        elif (drug == 'Almeta'):\n",
    "            drug = \"Alclometasone\"\n",
    "        elif (drug == \"Azarga\"):\n",
    "            drug = \"Brinzolamide\"\n",
    "        elif (drug == \"BCNU\"):\n",
    "            drug = \"Carmustine\"\n",
    "        elif (drug == \"BPTI\"):\n",
    "            drug = \"Aprotinin\"\n",
    "        elif (drug == \"Benicar-HCT\"):\n",
    "            drug = \"Olmesartan\"\n",
    "        elif (drug == \"Buscopan\"):\n",
    "            drug = \"Butylscopolamine\"\n",
    "        elif (drug == \"Cancidas\"):\n",
    "            drug = \"Caspofungin\"\n",
    "        elif (drug == \"Cantril\"):\n",
    "            drug = \"Mepenzolate\"\n",
    "        elif (drug == \"Colimycin\"):\n",
    "            drug = \"Colistin\"\n",
    "        elif (drug == \"DFMO\"):\n",
    "            drug = \"Eflornithine\"\n",
    "        elif (drug == \"DMSO\"):\n",
    "            drug = \"Dimethyl sulfoxide\"\n",
    "        elif (drug == \"Diane-35\"):\n",
    "            drug = \"Cyproterone acetate\"\n",
    "        elif (drug == \"EACA\"):\n",
    "            drug = \"Aminocaproic acid\"\n",
    "        elif (drug == \"EDTA\"):\n",
    "            drug = \"Edetic acid\"\n",
    "        elif (drug == \"Estrofem\"):\n",
    "            drug = \"Estradiol\"\n",
    "        elif (drug == \"FAMP\"):\n",
    "            drug = \"Fludarabine\"\n",
    "        elif (drug == \"FK463\"):\n",
    "            drug = \"Micafungin\"\n",
    "        elif (drug == \"FTY720\"):\n",
    "            drug = \"Fingolimod\"\n",
    "        elif (drug == \"FdUrd\"):\n",
    "            drug = \"Floxuridine\"\n",
    "        elif (drug == \"Fe(III\"):\n",
    "            drug = \"Ferric derisomaltose\"\n",
    "        elif (drug == \"Forteo\"):\n",
    "            drug = \"Teriparatide\"\n",
    "        elif (drug == \"Gd-DTPA\"):\n",
    "            drug = \"Gadopentetic acid\"\n",
    "        elif (drug == \"Gd-EOB-DTPA\"):\n",
    "            drug = \"Gadoxetic acid\"\n",
    "        elif (drug == 'Glat'):\n",
    "            drug = \"Glatiramer\"\n",
    "        elif (drug == \"Humalog\"):\n",
    "            drug = \"Insulin lispro\"\n",
    "        elif (drug == \"Implanon\"):\n",
    "            drug = \"Etonogestrel\"\n",
    "        elif (drug == \"Insulin\"):\n",
    "            drug = \"Insulin human\"\n",
    "        elif (drug == \"Kaluril\"):\n",
    "            drug = \"Amiloride\"\n",
    "        elif (drug == 'L-threo-DOPS'):\n",
    "            drug = \"Droxidopa\"\n",
    "        elif (drug == 'LMWH'):\n",
    "            drug = \"Heparin\"\n",
    "        elif (drug == 'LY146032'):\n",
    "            drug = \"Daptomycin\"\n",
    "        elif (drug == \"Lantus\"):\n",
    "            drug = \"Insulin glargine\"\n",
    "        elif (drug == \"Leuprorelin\"):\n",
    "            drug = \"Leuprolide\"\n",
    "        elif (drug == \"Locorten\"):\n",
    "            drug = \"Flumethasone\"\n",
    "        elif (drug == \"Lovaza\"):\n",
    "            drug = \"Omega-3-acid ethyl esters\"\n",
    "        elif (drug == \"Lyxumia\"):\n",
    "            drug = \"Lixisenatide\"\n",
    "        elif (drug == \"MDV3100\"):\n",
    "            drug = \"Enzalutamide\"\n",
    "        elif (drug == \"MK-462\"):\n",
    "            drug = \"Rizatriptan\"\n",
    "        elif (drug == \"Madopar\"):\n",
    "            drug = \"Carbidopa\"\n",
    "        elif (drug == \"Mersyndol\"):\n",
    "            drug = \"Codeine\"\n",
    "        elif (drug == \"Metrodin\"):\n",
    "            drug = \"Urofollitropin\"\n",
    "        elif (drug == \"MnDPDP\"):\n",
    "            drug = \"Mangafodipir\"\n",
    "        elif (drug == \"N-carbamylglutamate\"):\n",
    "            drug = \"Carglumic acid\"\n",
    "        elif (drug == \"N-methylscopolamine\"):\n",
    "            drug = \"Methscopolamine\"\n",
    "        elif (drug == 'NTBC'):\n",
    "            drug = \"Nitisinone\"\n",
    "        elif (drug == \"NovoLog\"):\n",
    "            drug = \"Insulin aspart\"\n",
    "        elif (drug == 'NuvaRing'):\n",
    "            drug = \"Etonogestrel\"\n",
    "        elif (drug == 'Nuvocid'):\n",
    "            drug = \"Oritavancin\"\n",
    "        elif (drug == \"Optison\"):\n",
    "            drug = \"Perflutren\"\n",
    "        elif (drug == \"PCI-32765\"):\n",
    "            drug = \"Ibrutinib\"\n",
    "        elif (drug == \"PEP005\"):\n",
    "            drug = \"Ingenol mebutate\"\n",
    "        elif (drug == \"PGE1\"):\n",
    "            drug = \"Alprostadil\"\n",
    "        elif (drug == \"PGE2\"):\n",
    "            drug = \"Dinoprostone\"\n",
    "        elif (drug == \"Paroven\"):\n",
    "            drug = \"Oxerutins\"\n",
    "        elif (drug == \"Photofrin\"):\n",
    "            drug = \"Porfimer sodium\"\n",
    "        elif (drug == 'Prednefrin'):\n",
    "            drug = \"Prednisolone acetate\"\n",
    "        elif (drug == \"Promacta\"):\n",
    "            drug = \"Eltrombopag\"\n",
    "        elif (drug == \"Prussian\"):\n",
    "            drug = \"Prussian blue\"\n",
    "        elif (drug == \"Refludan\"):\n",
    "            drug = \"Lepirudin\"\n",
    "        elif (drug == \"Revasc\"):\n",
    "            drug = \"Desirudin\"\n",
    "        elif (drug in [\"SOM230\", \"Signifor\"]):\n",
    "            drug = \"Pasireotide\"\n",
    "        elif (drug == \"SU5416\"):\n",
    "            drug = \"Semaxanib\"\n",
    "        elif (drug == 'Sativex'):\n",
    "            drug = \"Nabiximols\"\n",
    "        elif (drug == \"SonoVue\"):\n",
    "            drug = \"Sulfur hexafluoride\"\n",
    "        elif (drug == \"Stalevo\"):\n",
    "            drug = \"Carbidopa\"\n",
    "        elif (drug == 'Symlin'):\n",
    "            drug = \"Pramlintide\"\n",
    "        elif (drug == \"TMC207\"):\n",
    "            drug = \"Bedaquiline\"\n",
    "        elif (drug == \"TMC435\"):\n",
    "            drug = \"Simeprevir\"\n",
    "        elif (drug == \"TPGS\"):\n",
    "            drug = \"Tocofersolan\"\n",
    "        elif (drug == \"TR-700\"):\n",
    "            drug = \"Tedizolid\"\n",
    "        elif (drug == \"Tadenan\"):\n",
    "            drug = \"Pygeum africanum bark\"\n",
    "        elif (drug == \"Timentin\"):\n",
    "            drug = \"Ticarcillin\"\n",
    "        elif (drug == \"TlCl\"):\n",
    "            drug = \"Thallous Chloride\"\n",
    "        elif (drug == \"Triphasil\"):\n",
    "            drug = \"Levonorgestrel\"\n",
    "        elif (drug == \"Trisequens\"):\n",
    "            drug = \"Estradiol\"\n",
    "        elif (drug == \"UDCA\"):\n",
    "            drug = \"Ursodeoxycholic acid\"\n",
    "        elif (drug == 'VACV'):\n",
    "            drug = \"Human vaccinia virus immune globulin\"\n",
    "        elif (drug == \"Vallergan\"):\n",
    "            drug = \"Alimemazine\"\n",
    "        elif (drug == \"Vancocine\"):\n",
    "            drug = \"Vancomycin\"\n",
    "        elif (drug == \"Westcort\"):\n",
    "            drug = \"Hydrocortisone valerate\"\n",
    "        elif (drug == \"ZnCl2\"):\n",
    "            drug = \"Zinc chloride\"\n",
    "        elif (drug == \"Zyprexa\"):\n",
    "            drug = \"Olanzapine\"\n",
    "        elif (drug == \"acetohydroxamic\"):\n",
    "            drug = \"Acetohydroxamic acid\"\n",
    "        elif (drug == \"actinomycin\"):\n",
    "            drug = \"Dactinomycin\"\n",
    "        elif (drug == \"alendronate\"):\n",
    "            drug = \"Alendronic acid\"\n",
    "        elif (drug == \"alpha-methyl-p-tyrosine\"):\n",
    "            drug = \"Racemetyrosine\"\n",
    "        else:\n",
    "            drug = drug.split(\"-\")[-1]\n",
    "            drug = drug[0].upper()+drug[1:]\n",
    "        return drug\n",
    "\n",
    "    drugnames = list(se_similarity_df.index)\n",
    "    di_drugname2pubchemid = {di_pubchemid2drugname[k] : k for k in di_pubchemid2drugname}\n",
    "    pubchem_ids = [di_drugname2pubchemid.get(standard_name_from_SIDER(d), None) for d in drugnames]\n",
    "    notfound_ids = [standard_name_from_SIDER(drugnames[ip]) for ip, p in enumerate(pubchem_ids) if (not p)]\n",
    "    got_pubchem_ids = []\n",
    "    sz = 100\n",
    "    for i in range(0,len(notfound_ids),sz):\n",
    "        print(\"%d/%d\" % (i, len(notfound_ids)))\n",
    "        got_pubchem_ids += get_pubchem_id(notfound_ids[i:(i+sz)])\n",
    "        \n",
    "    pubchem_ids = [di_drugname2pubchemid.get(standard_name_from_SIDER(d), None) for d in drugnames]\n",
    "    idx_notfound=0\n",
    "    for i, p in enumerate(pubchem_ids):\n",
    "        if (not p):\n",
    "            pubchem_ids[i] = got_pubchem_ids[idx_notfound]\n",
    "            idx_notfound += 1\n",
    "    print((idx_notfound, len(got_pubchem_ids)))\n",
    "    \n",
    "    drug_ids = [ip for ip, p in enumerate(pubchem_ids) if (p)]\n",
    "    se_custom = se_similarity_df.iloc[drug_ids,drug_ids]\n",
    "    se_custom.index = [p for p in pubchem_ids if (p)]\n",
    "    se_custom.columns = se_custom.index\n",
    "\n",
    "    custom = pd.read_csv(paths_global.data_folder+\"FEATURELESS_v1.0.0/all_ratings.csv\", index_col=0, header=0)\n",
    "\n",
    "    ids = [str(x) for x in custom.index if (x not in se_custom.index.astype(str))]\n",
    "    se_custom = pd.concat((se_custom, pd.DataFrame([], index=ids, columns=ids)))\n",
    "    se_custom.index = se_custom.index.astype(str)\n",
    "    se_custom.columns = se_custom.columns.astype(str)\n",
    "    se_custom = se_custom.loc[~se_custom.index.duplicated()]\n",
    "    se_custom = se_custom.loc[chemical_custom.index][chemical_custom.index]\n",
    "    np.fill_diagonal(se_custom.values,1)\n",
    "\n",
    "    se_custom.index = se_custom.index.astype(str)\n",
    "    se_custom.columns = se_custom.columns.astype(str)\n",
    "    \n",
    "    se_custom.to_csv(se_similarity_file)\n",
    "se_custom = pd.read_csv(se_similarity_file, header=0, index_col=0)\n",
    "se_custom"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### II.3 Drug-drug similarity: (Target) sequence drug-drug similarities\n",
    "\n",
    "This yields a dataframe `sequence_similarity_df` of size #drugs x #drugs, where identifiers are PubChem or DrugBank identifiers, which contains, when computable, average drug-pairwise alignment scores between gene target sequences. We use the normalization suggested in [Bleakley and Yamanishi (2009)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2735674/)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Align\n",
    "from Bio import SeqIO\n",
    "from time import time\n",
    "\n",
    "seq_similarity_file = predict_folder+\"sequence_similarity_df.csv\"\n",
    "seq_cst_similarity_file = predict_folder+\"TARGET_SIMILARITY/sequence_cst_similarity_df.txt\"\n",
    "\n",
    "if (not os.path.exists(seq_cst_similarity_file)):\n",
    "    sb.Popen((\"mkdir -p \"+predict_folder+\"TARGET_SIMILARITY/\".split(\" \")))\n",
    "    sequence_aligner = Align.PairwiseAligner()\n",
    "    fname = drugbank_folder+\"TARGET SEQUENCES/Drug Target Sequences/gene.fasta\"\n",
    "    ## Target sequences\n",
    "    target_sequences = [record.seq for record in SeqIO.parse(fname, \"fasta\")]\n",
    "    \n",
    "    ## Following the normalization suggested in Bleakley and Yamanishi (2009)\n",
    "    from scipy.stats.mstats import gmean as geometric_mean\n",
    "    start=time()\n",
    "    sequence_CST_similarity = [sequence_aligner.score(seq, seq) for seq in target_sequences]\n",
    "    end=time()-start\n",
    "    print(\"Computation took %.2f sec.\" % (end))\n",
    "    norm_cst = geometric_mean(sequence_CST_similarity)\n",
    "    with open(seq_cst_similarity_file, \"w+\") as f:\n",
    "        f.write(str(norm_cst))\n",
    "else:\n",
    "    with open(seq_cst_similarity_file, \"r\") as f:\n",
    "        norm_cst = float(f.read())\n",
    "        \n",
    "if (not os.path.exists(predict_folder+\"TARGET_SIMILARITY/save_sequence_similarity_df.csv\")):\n",
    "    sequence_aligner = Align.PairwiseAligner()\n",
    "    fname = drugbank_folder+\"TARGET SEQUENCES/Drug Target Sequences/gene.fasta\"\n",
    "    ## Target sequences\n",
    "    target_sequences = [record.seq for record in SeqIO.parse(fname, \"fasta\")]\n",
    "    ## Associated DrugBank ids\n",
    "    sequence_drug_ids = sb.check_output(\"cat '\"+fname+\"' | grep '>drugbank_target|' | sed -e 's/^.*(//g' | sed -e 's/)$//g'\", shell=True).decode(\"utf-8\").split(\"\\n\")\n",
    "    drug_ids_per_seq = [idx.split(\"; \") for idx in sequence_drug_ids]\n",
    "    ## unique DrugBank ids\n",
    "    drug_ids = list(set(filter(lambda x : len(x) > 0, [y for x in drug_ids_per_seq for y in x])))\n",
    "    ## consider only drugbank ids in custom dataset for computational reasons\n",
    "    drugnames = [di_drugbankid2drugname.get(cid, None) for cid in drug_ids]\n",
    "    di_drugname2pubchemid = {di_pubchemid2drugname[k] : k for k in di_pubchemid2drugname}\n",
    "    pubchem_ids = [di_drugname2pubchemid.get(d, None) for d in drugnames]\n",
    "    drug_notnone_ids = [ip for ip, p in enumerate(pubchem_ids) if (p)]\n",
    "    drug_ids = [drug_ids[i] for i in drug_notnone_ids]\n",
    "    print(len(drug_ids))\n",
    "\n",
    "    ndrugs=len(drug_ids)\n",
    "    \n",
    "    def seq_function(i, j):\n",
    "        print((i,j,ndrugs))\n",
    "        if (i>j):\n",
    "            return 0.\n",
    "        else:\n",
    "            drug_i = drug_ids[i]\n",
    "            drug_j = drug_ids[j]\n",
    "            ## all sequences in which drug_i appears\n",
    "            target_i_ids = [idx for idx, drug_ls in enumerate(drug_ids_per_seq) if (drug_i in drug_ls)]\n",
    "            target_seq_i = [target_sequences[idx_i] for idx_i in target_i_ids]\n",
    "            if (i==j):\n",
    "                align_scores = [sequence_aligner.score(s1, s2) for s1 in target_seq_i for s2 in target_seq_i]\n",
    "            else:\n",
    "                ## all sequences in which drug_j appears\n",
    "                target_j_ids = [idx for idx, drug_ls in enumerate(drug_ids_per_seq) if (drug_j in drug_ls)]\n",
    "                target_seq_j = [target_sequences[idx_j] for idx_j in target_j_ids]\n",
    "                align_scores = [sequence_aligner.score(s1, s2) for s1 in target_seq_i for s2 in target_seq_j]\n",
    "            score = np.mean(align_scores) if (len(align_scores)>0) else np.nan\n",
    "            return score\n",
    "        \n",
    "    start = time()\n",
    "    if (os.path.exists(predict_folder+\"TARGET_SIMILARITY/sequence_similarity.out\")):\n",
    "        sequence_similarity = np.loadtxt(predict_folder+\"TARGET_SIMILARITY/sequence_similarity.out\")\n",
    "    else:\n",
    "        sequence_similarity = np.zeros((ndrugs, ndrugs))\n",
    "    start_i = 0\n",
    "    for i in range(start_i,ndrugs):\n",
    "        for j in range(i, ndrugs):\n",
    "            sequence_similarity[i,j] = seq_function(i, j)\n",
    "        np.savetxt(predict_folder+\"TARGET_SIMILARITY/sequence_similarity.out\", sequence_similarity)\n",
    "    end = time()-start\n",
    "    print(\"Computation took %.2f sec.\" % end)\n",
    "    \n",
    "    sequence_similarity = np.loadtxt(predict_folder+\"TARGET_SIMILARITY/sequence_similarity.out\")\n",
    "    sequence_similarity_df = pd.DataFrame(sequence_similarity, index=drug_ids, columns=drug_ids)\n",
    "    sequence_similarity_df.to_csv(predict_folder+\"TARGET_SIMILARITY/save_sequence_similarity_df.csv\")\n",
    "                                  \n",
    "sequence_similarity_df = pd.read_csv(predict_folder+\"TARGET_SIMILARITY/save_sequence_similarity_df.csv\", index_col=0)\n",
    "\n",
    "if (not os.path.exists(seq_similarity_file)):\n",
    "    diag = np.diag(sequence_similarity_df.values)\n",
    "    sequence_similarity_df += sequence_similarity_df.T\n",
    "    np.fill_diagonal(sequence_similarity_df.values, diag)\n",
    "    ## normalisation\n",
    "    sequence_similarity_df = sequence_similarity_df/norm_cst\n",
    "    \n",
    "    drugnames = [di_drugbankid2drugname.get(cid, None) for cid in sequence_similarity_df.index]\n",
    "    di_drugname2pubchemid = {di_pubchemid2drugname[k] : k for k in di_pubchemid2drugname}\n",
    "    pubchem_ids = [di_drugname2pubchemid.get(d, None) for d in drugnames]\n",
    "    drug_ids = [ip for ip, p in enumerate(pubchem_ids) if (p)]\n",
    "    sequence_custom = sequence_similarity_df.iloc[drug_ids,drug_ids]\n",
    "    sequence_custom.index = [p for p in pubchem_ids if (p)]\n",
    "    sequence_custom.columns = sequence_custom.index\n",
    "\n",
    "    sequence_custom.index = sequence_custom.index.astype(str)\n",
    "    sequence_custom.columns = sequence_custom.columns.astype(str)\n",
    "\n",
    "    sequence_custom.to_csv(seq_similarity_file)\n",
    "\n",
    "sequence_custom = pd.read_csv(seq_similarity_file, index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### II.4 Drug-drug similarity: Closeness in the human PPI network drug target to drug target similarities\n",
    "\n",
    "This yields a dataframe `network_similarity_df` of size #drugs x #drugs, where identifiers are PubChem or DrugBank identifiers, which contains, when computable, the transformed distance between drug targets in the PPI using the Floyd-Warshall algorithm. We apply to the Floyd-Warshall cost matrix the transformation suggested by the [PREDICT paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3159979/)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "network_folder = predict_folder+\"NETWORK_SIMILARITY/\"\n",
    "network_similarity_file = predict_folder+\"network_similarity_df.csv\"\n",
    "\n",
    "## 1. Mapping drugs and genes\n",
    "if (not os.path.exists(network_folder+\"SAVE_network_similarity_df.csv\")):\n",
    "    sb.Popen((\"mkdir -p \"+network_folder).split(\" \"))\n",
    "    fname=paths_global.drugbank_folder+\"PROTEIN IDENTIFIERS/Drug Target Identifiers/all.csv\"\n",
    "\n",
    "    protein_db = pd.read_csv(fname, index_col=5).query(\"Species=='Humans'\")[[\"Gene Name\", \"Drug IDs\"]]\n",
    "    drug_protein_db = protein_db[[\"Drug IDs\"]].groupby(level=0).apply(lambda x : \"; \".join(list(sorted(set(list(x.values.flatten()))))))\n",
    "    drug_gene_db = protein_db[[\"Gene Name\"]].groupby(level=0).apply(lambda x : \"; \".join(list(map(str,set(x.values.flatten())))))\n",
    "    protein_db = pd.DataFrame([], index=drug_protein_db.index)\n",
    "    protein_db[\"Drug IDs\"] = drug_protein_db\n",
    "    protein_db[\"Gene Name\"] = drug_gene_db\n",
    "\n",
    "    drug_ids = list(set([y for x in list(protein_db[\"Drug IDs\"]) for y in x.split(\"; \")]))\n",
    "    ## consider only drugbank ids in custom dataset for computational reasons\n",
    "    drugnames = [di_drugbankid2drugname.get(cid, None) for cid in drug_ids]\n",
    "    di_drugname2pubchemid = {di_pubchemid2drugname[k] : k for k in di_pubchemid2drugname}\n",
    "    pubchem_ids = [di_drugname2pubchemid.get(d, None) for d in drugnames]\n",
    "    drug_notnone_ids = [ip for ip, p in enumerate(pubchem_ids) if (p)]\n",
    "    drug_ids = [drug_ids[i] for i in drug_notnone_ids]\n",
    "\n",
    "    keep_ids = [ix for ix, x in enumerate(list(protein_db[\"Drug IDs\"])) if (any([y in drug_ids for y in x.split(\"; \")]))]\n",
    "    protein_db = protein_db.iloc[keep_ids]\n",
    "\n",
    "    from NORDic.UTILS.STRING_utils import get_app_name_STRING, get_network_from_STRING\n",
    "    network_fname = network_folder+\"human_ppi_network.csv\"\n",
    "    ## 9606=human, retrieve PPI from STRING\n",
    "    if (not os.path.exists(network_fname)):\n",
    "        gene_list = list(protein_db[\"Gene Name\"].unique())\n",
    "        taxon_id=9606 #human\n",
    "        app_name = get_app_name_STRING(paths_global.string_file)\n",
    "        ppi = get_network_from_STRING(gene_list, taxon_id, min_score=0, network_type=\"functional\", add_nodes=0, \n",
    "                                      app_name=app_name, version=\"11.5\", quiet=False)\n",
    "        ppi.to_csv(network_fname)\n",
    "    ppi = pd.read_csv(network_fname, index_col=0)\n",
    "    \n",
    "    ## selection of interactions\n",
    "    score_thres=0\n",
    "\n",
    "    ## Get drug-gene associations\n",
    "    gene_targets = list(protein_db[\"Gene Name\"])\n",
    "\n",
    "    ppi = pd.read_csv(network_fname, sep=\"\\t\", header=0, index_col=0)\n",
    "    ppi_subset = ppi.loc[ppi[\"score\"]>score_thres][[\"preferredName_A\",\"preferredName_B\",\"score\"]]\n",
    "    ppi_subset = ppi_subset.loc[[(a in gene_targets) and (b in gene_targets) for a,b in zip(list(ppi_subset[\"preferredName_A\"]),list(ppi_subset[\"preferredName_B\"]))]]\n",
    "    ppi_subset[\"cost\"] = 1-ppi_subset[[\"score\"]].values.flatten() #cost\n",
    "    ppi_subset = ppi_subset[[\"preferredName_A\",\"preferredName_B\",\"cost\"]]\n",
    "    ppi_subset = ppi_subset.drop_duplicates(keep=\"first\")\n",
    "\n",
    "    ## 1 is the maximum cost of edge\n",
    "    max_cost=1.\n",
    "    wam = ppi_subset.pivot(index='preferredName_A', columns='preferredName_B', values='cost').fillna(max_cost)\n",
    "    for g in wam.index:\n",
    "        if (g not in wam.columns):\n",
    "            wam[g] = [max_cost]*wam.shape[0]\n",
    "    for g in wam.columns:\n",
    "        if (g not in wam.index):\n",
    "            wam.loc[g] = [max_cost]*wam.shape[1]\n",
    "    wam = wam.loc[wam.index][wam.index]\n",
    "\n",
    "    ## Compute distance between target genes\n",
    "    from scipy.sparse.csgraph import floyd_warshall\n",
    "\n",
    "    dists = floyd_warshall(wam.values.copy(order='C'), directed=False)\n",
    "    dists_df = pd.DataFrame(dists, index=wam.index, columns=wam.columns)\n",
    "\n",
    "    A_mat, b_mat = 0.9*np.exp(1), 1.\n",
    "    network_similarity = A_mat*np.exp(-b_mat*dists_df)\n",
    "    network_similarity /= np.max(network_similarity)\n",
    "\n",
    "    network_similarity_1 = pd.DataFrame(0., index=drug_ids, columns=network_similarity.columns)\n",
    "    for drug in drug_ids:\n",
    "        keep_drug_ids = [idx for idx in protein_db.index if (drug in protein_db.loc[idx][\"Drug IDs\"])]\n",
    "        targeted_genes = [x for x in list(protein_db.loc[keep_drug_ids][\"Gene Name\"]) if (x in network_similarity.index)]\n",
    "        network_similarity_1.loc[drug] = network_similarity.loc[targeted_genes].mean(axis=0)\n",
    "        \n",
    "    network_similarity_df = pd.DataFrame([], index=drug_ids)\n",
    "    for drug in drug_ids:\n",
    "        keep_drug_ids = [idx for idx in protein_db.index if (drug in protein_db.loc[idx][\"Drug IDs\"])]\n",
    "        targeted_genes = [x for x in list(protein_db.loc[keep_drug_ids][\"Gene Name\"]) if (x in network_similarity.index)]\n",
    "        network_similarity_df[drug] = network_similarity_1[targeted_genes].mean(axis=1)\n",
    "        \n",
    "    network_similarity_df.to_csv(network_folder+\"SAVE_network_similarity_df.csv\")\n",
    "    \n",
    "network_similarity_df = pd.read_csv(network_folder+\"SAVE_network_similarity_df.csv\", index_col=0)\n",
    "\n",
    "if (not os.path.exists(network_similarity_file)):\n",
    "    drugnames = [di_drugbankid2drugname.get(cid, None) for cid in network_similarity_df.index]\n",
    "    di_drugname2pubchemid = {di_pubchemid2drugname[k] : k for k in di_pubchemid2drugname}\n",
    "    pubchem_ids = [di_drugname2pubchemid.get(d, None) for d in drugnames]\n",
    "    drug_ids = [ip for ip, p in enumerate(pubchem_ids) if (p)]\n",
    "    network_custom = network_similarity_df.iloc[drug_ids,drug_ids]\n",
    "    network_custom.index = [p for p in pubchem_ids if (p)]\n",
    "    network_custom.columns = network_custom.index\n",
    "\n",
    "    network_custom.index = network_custom.index.astype(str)\n",
    "    network_custom.columns = network_custom.columns.astype(str)\n",
    "\n",
    "    sb.call(\"mkdir -p \"+data_folder+\"custom_PREDICT/\", shell=True)\n",
    "    network_similarity_file = data_folder+\"custom_PREDICT/network_similarity_df.csv\"\n",
    "    network_custom.to_csv(network_similarity_file)\n",
    "    \n",
    "network_custom = pd.read_csv(network_similarity_file, index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### II.5 Drug-drug similarity: Gene Onthology drug-drug similarities\n",
    "\n",
    "This yields a dataframe `go_similarity_df` of size #drugs x #drugs, where identifiers are PubChem or DrugBank identifiers, which contains the similarity between Gene Onthology annotations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "go_similarity_file = predict_folder+\"go_similarity_df.csv\"\n",
    "    \n",
    "if (not os.path.exists(go_similarity_file)):\n",
    "    ## Get drug-protein-gene associations\n",
    "    fname=paths_global.drugbank_folder+\"PROTEIN IDENTIFIERS/Drug Target Identifiers/all.csv\"\n",
    "\n",
    "    protein_db = pd.read_csv(fname, index_col=5).query(\"Species=='Humans'\")[[\"Gene Name\", \"Drug IDs\"]]\n",
    "    drug_protein_db = protein_db[[\"Drug IDs\"]].groupby(level=0).apply(lambda x : \"; \".join(list(sorted(set(list(x.values.flatten()))))))\n",
    "    drug_gene_db = protein_db[[\"Gene Name\"]].groupby(level=0).apply(lambda x : \"; \".join(list(map(str,set(x.values.flatten())))))\n",
    "    protein_db = pd.DataFrame([], index=drug_protein_db.index)\n",
    "    protein_db[\"Drug IDs\"] = drug_protein_db\n",
    "    protein_db[\"Gene Name\"] = drug_gene_db\n",
    "\n",
    "    drug_ids = list(set([y for x in list(protein_db[\"Drug IDs\"]) for y in x.split(\"; \")]))\n",
    "    ## consider only drugbank ids in custom dataset for computational reasons\n",
    "    drugnames = [di_drugbankid2drugname.get(cid, None) for cid in drug_ids]\n",
    "    di_drugname2pubchemid = {di_pubchemid2drugname[k] : k for k in di_pubchemid2drugname}\n",
    "    pubchem_ids = [di_drugname2pubchemid.get(d, None) for d in drugnames]\n",
    "    drug_notnone_ids = [ip for ip, p in enumerate(pubchem_ids) if (p)]\n",
    "    drug_ids = [drug_ids[i] for i in drug_notnone_ids]\n",
    "\n",
    "    keep_ids = [ix for ix, x in enumerate(list(protein_db[\"Drug IDs\"])) if (any([y in drug_ids for y in x.split(\"; \")]))]\n",
    "    protein_db = protein_db.iloc[keep_ids]\n",
    "\n",
    "    go_sim_fname = \"go_similarity\"\n",
    "    if (not os.path.exists(predict_folder+\"GO_SIMILARITY/\"+go_sim_fname+\".csv\")):\n",
    "        sb.Popen((\"mkdir -p \"+predict_folder+\"GO_SIMILARITY/\").split(\" \"))\n",
    "        proteins = list(set(list(protein_db.index)))\n",
    "        with open(\"proteins.txt\", \"w+\") as f:\n",
    "            f.write(\",\".join(proteins))\n",
    "        process = sb.Popen([\"Rscript\", \"../src/Rscript_GOsim.R\"])\n",
    "        process.wait()\n",
    "        sb.Popen([\"mv\", \"go_similarity.csv\", \"go_similarity_MF.csv\", \n",
    "                  \"go_similarity_CC.csv\", \"go_similarity_BP.csv\", predict_folder+\"GO_SIMILARITY/\"])\n",
    "    ## annotated by protein        \n",
    "    go_similarity = pd.read_csv(predict_folder+\"GO_SIMILARITY/\"+go_sim_fname+\".csv\", index_col=0)\n",
    "\n",
    "    go_similarity_mat = go_similarity.fillna(0.)\n",
    "\n",
    "    go_similarity_1 = pd.DataFrame([], index=drug_ids, columns=go_similarity_mat.columns)\n",
    "    for drug in drug_ids:\n",
    "        keep_drug_ids = []\n",
    "        for idx in protein_db.index[3:]:\n",
    "            df = protein_db.loc[idx][\"Drug IDs\"]\n",
    "            if (\"str\" in str(type(df))):\n",
    "                drs = df.split(\"; \")\n",
    "            else:\n",
    "                drs = [x for ls in list(df) for x in ls.split(\"; \")]\n",
    "            if (drug in drs):\n",
    "                keep_drug_ids.append(idx)\n",
    "        keep_drug_ids = list(set(keep_drug_ids))\n",
    "        if (len(keep_drug_ids)==0):\n",
    "            continue\n",
    "        go_similarity_1.loc[drug] = list(go_similarity_mat.loc[keep_drug_ids].mean(axis=0))\n",
    "        \n",
    "    ## annotated by DrugBank ID \n",
    "    go_similarity_df = pd.DataFrame([], index=drug_ids)\n",
    "    for drug in drug_ids:\n",
    "        keep_drug_ids = []\n",
    "        for idx in protein_db.index[3:]:\n",
    "            df = protein_db.loc[idx][\"Drug IDs\"]\n",
    "            if (\"str\" in str(type(df))):\n",
    "                drs = df.split(\"; \")\n",
    "            else:\n",
    "                drs = [x for ls in list(df) for x in ls.split(\"; \")]\n",
    "            if (drug in drs):\n",
    "                keep_drug_ids.append(idx)\n",
    "        keep_drug_ids = list(set(keep_drug_ids))\n",
    "        if (len(keep_drug_ids)==0):\n",
    "            continue\n",
    "        go_similarity_df[drug] = go_similarity_1[keep_drug_ids].mean(axis=1)\n",
    "\n",
    "    drugnames = [di_drugbankid2drugname.get(cid, None) for cid in go_similarity_df.index]\n",
    "    di_drugname2pubchemid = {di_pubchemid2drugname[k] : k for k in di_pubchemid2drugname}\n",
    "    pubchem_ids = [di_drugname2pubchemid.get(d, None) for d in drugnames]\n",
    "    drug_ids = [ip for ip, p in enumerate(pubchem_ids) if (p)]\n",
    "    go_custom = go_similarity_df.iloc[drug_ids,drug_ids]\n",
    "    go_custom.index = [p for p in pubchem_ids if (p)]\n",
    "    go_custom.columns = go_custom.index\n",
    "\n",
    "    ## annotated by PubChem CID mainly\n",
    "    go_custom.index = go_custom.index.astype(str)\n",
    "    go_custom.columns = go_custom.columns.astype(str)\n",
    "\n",
    "    go_custom.to_csv(go_similarity_file)\n",
    "    \n",
    "go_custom = pd.read_csv(go_similarity_file, index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### II.6 Drug-drug similarity 6: Transcriptomic-based\n",
    "\n",
    "\"Given genetic signatures of diseases obtained from gene expression experiments, we used a Jaccard score between every pair of signatures, taking into account the direction of the response of each gene. That is, the total number of mutual upregulated genes and mutual downregulated genes over the unified list of all genes. Signature genes with inconsistent regulation directionality for the same disease across various experiments (i.e., registered as both upregulated and downregulated across various experiments for the same disease) were filtered, allowing for up to 10% expression measurement errors.\" \n",
    "\n",
    "This yields a dataframe `go_similarity_df` of size #drugs x #drugs, where identifiers are PubChem or DrugBank identifiers, which contains the similarity between Gene Onthology annotations.\n",
    "\n",
    "Signatures were obtained in the notebook TRANSCRIPT_dataset-v1.0.0.ipynb."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>36314</th>\n",
       "      <th>11354606</th>\n",
       "      <th>679</th>\n",
       "      <th>2554</th>\n",
       "      <th>126941</th>\n",
       "      <th>3033</th>\n",
       "      <th>444795</th>\n",
       "      <th>5757</th>\n",
       "      <th>19649</th>\n",
       "      <th>6279</th>\n",
       "      <th>...</th>\n",
       "      <th>5281078</th>\n",
       "      <th>16129706</th>\n",
       "      <th>2310</th>\n",
       "      <th>4375</th>\n",
       "      <th>3352</th>\n",
       "      <th>3373</th>\n",
       "      <th>5917</th>\n",
       "      <th>442872</th>\n",
       "      <th>442021</th>\n",
       "      <th>4843</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>36314</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.922423</td>\n",
       "      <td>0.917092</td>\n",
       "      <td>0.917092</td>\n",
       "      <td>0.919255</td>\n",
       "      <td>0.927368</td>\n",
       "      <td>0.922037</td>\n",
       "      <td>0.926905</td>\n",
       "      <td>0.918251</td>\n",
       "      <td>0.918251</td>\n",
       "      <td>...</td>\n",
       "      <td>0.205069</td>\n",
       "      <td>0.214495</td>\n",
       "      <td>0.210478</td>\n",
       "      <td>0.216195</td>\n",
       "      <td>0.216350</td>\n",
       "      <td>0.207078</td>\n",
       "      <td>0.218359</td>\n",
       "      <td>0.205378</td>\n",
       "      <td>0.211096</td>\n",
       "      <td>0.213723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11354606</th>\n",
       "      <td>0.922423</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.917246</td>\n",
       "      <td>0.917246</td>\n",
       "      <td>0.917710</td>\n",
       "      <td>0.918869</td>\n",
       "      <td>0.919873</td>\n",
       "      <td>0.926905</td>\n",
       "      <td>0.913769</td>\n",
       "      <td>0.913769</td>\n",
       "      <td>...</td>\n",
       "      <td>0.207696</td>\n",
       "      <td>0.208623</td>\n",
       "      <td>0.212023</td>\n",
       "      <td>0.215423</td>\n",
       "      <td>0.210168</td>\n",
       "      <td>0.213877</td>\n",
       "      <td>0.213414</td>\n",
       "      <td>0.211250</td>\n",
       "      <td>0.213723</td>\n",
       "      <td>0.209241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>679</th>\n",
       "      <td>0.917092</td>\n",
       "      <td>0.917246</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.924587</td>\n",
       "      <td>0.920028</td>\n",
       "      <td>0.919487</td>\n",
       "      <td>0.927755</td>\n",
       "      <td>0.915237</td>\n",
       "      <td>0.915237</td>\n",
       "      <td>...</td>\n",
       "      <td>0.222145</td>\n",
       "      <td>0.213182</td>\n",
       "      <td>0.209628</td>\n",
       "      <td>0.214418</td>\n",
       "      <td>0.219363</td>\n",
       "      <td>0.214264</td>\n",
       "      <td>0.213027</td>\n",
       "      <td>0.212255</td>\n",
       "      <td>0.213027</td>\n",
       "      <td>0.212718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2554</th>\n",
       "      <td>0.917092</td>\n",
       "      <td>0.917246</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.924587</td>\n",
       "      <td>0.920028</td>\n",
       "      <td>0.919487</td>\n",
       "      <td>0.927755</td>\n",
       "      <td>0.915237</td>\n",
       "      <td>0.915237</td>\n",
       "      <td>...</td>\n",
       "      <td>0.222145</td>\n",
       "      <td>0.213182</td>\n",
       "      <td>0.209628</td>\n",
       "      <td>0.214418</td>\n",
       "      <td>0.219363</td>\n",
       "      <td>0.214264</td>\n",
       "      <td>0.213027</td>\n",
       "      <td>0.212255</td>\n",
       "      <td>0.213027</td>\n",
       "      <td>0.212718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126941</th>\n",
       "      <td>0.919255</td>\n",
       "      <td>0.917710</td>\n",
       "      <td>0.924587</td>\n",
       "      <td>0.924587</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.923582</td>\n",
       "      <td>0.919332</td>\n",
       "      <td>0.928991</td>\n",
       "      <td>0.916783</td>\n",
       "      <td>0.916783</td>\n",
       "      <td>...</td>\n",
       "      <td>0.224308</td>\n",
       "      <td>0.215500</td>\n",
       "      <td>0.214573</td>\n",
       "      <td>0.216118</td>\n",
       "      <td>0.209473</td>\n",
       "      <td>0.214109</td>\n",
       "      <td>0.214418</td>\n",
       "      <td>0.213027</td>\n",
       "      <td>0.217663</td>\n",
       "      <td>0.210709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3373</th>\n",
       "      <td>0.207078</td>\n",
       "      <td>0.213877</td>\n",
       "      <td>0.214264</td>\n",
       "      <td>0.214264</td>\n",
       "      <td>0.214109</td>\n",
       "      <td>0.209396</td>\n",
       "      <td>0.218127</td>\n",
       "      <td>0.215268</td>\n",
       "      <td>0.205223</td>\n",
       "      <td>0.205223</td>\n",
       "      <td>...</td>\n",
       "      <td>0.186215</td>\n",
       "      <td>0.305362</td>\n",
       "      <td>0.331479</td>\n",
       "      <td>0.281409</td>\n",
       "      <td>0.159017</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.329161</td>\n",
       "      <td>0.214650</td>\n",
       "      <td>0.259002</td>\n",
       "      <td>0.189306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5917</th>\n",
       "      <td>0.218359</td>\n",
       "      <td>0.213414</td>\n",
       "      <td>0.213027</td>\n",
       "      <td>0.213027</td>\n",
       "      <td>0.214418</td>\n",
       "      <td>0.219286</td>\n",
       "      <td>0.219518</td>\n",
       "      <td>0.221140</td>\n",
       "      <td>0.211405</td>\n",
       "      <td>0.211405</td>\n",
       "      <td>...</td>\n",
       "      <td>0.205532</td>\n",
       "      <td>0.241539</td>\n",
       "      <td>0.248802</td>\n",
       "      <td>0.298563</td>\n",
       "      <td>0.160408</td>\n",
       "      <td>0.329161</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.231494</td>\n",
       "      <td>0.188688</td>\n",
       "      <td>0.155772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>442872</th>\n",
       "      <td>0.205378</td>\n",
       "      <td>0.211250</td>\n",
       "      <td>0.212255</td>\n",
       "      <td>0.212255</td>\n",
       "      <td>0.213027</td>\n",
       "      <td>0.212950</td>\n",
       "      <td>0.215809</td>\n",
       "      <td>0.220522</td>\n",
       "      <td>0.207850</td>\n",
       "      <td>0.207850</td>\n",
       "      <td>...</td>\n",
       "      <td>0.252357</td>\n",
       "      <td>0.260547</td>\n",
       "      <td>0.208159</td>\n",
       "      <td>0.257302</td>\n",
       "      <td>0.156854</td>\n",
       "      <td>0.214650</td>\n",
       "      <td>0.231494</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.242621</td>\n",
       "      <td>0.171380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>442021</th>\n",
       "      <td>0.211096</td>\n",
       "      <td>0.213723</td>\n",
       "      <td>0.213027</td>\n",
       "      <td>0.213027</td>\n",
       "      <td>0.217663</td>\n",
       "      <td>0.219595</td>\n",
       "      <td>0.212409</td>\n",
       "      <td>0.221913</td>\n",
       "      <td>0.214341</td>\n",
       "      <td>0.214341</td>\n",
       "      <td>...</td>\n",
       "      <td>0.251584</td>\n",
       "      <td>0.219595</td>\n",
       "      <td>0.246948</td>\n",
       "      <td>0.237058</td>\n",
       "      <td>0.225931</td>\n",
       "      <td>0.259002</td>\n",
       "      <td>0.188688</td>\n",
       "      <td>0.242621</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.266110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4843</th>\n",
       "      <td>0.213723</td>\n",
       "      <td>0.209241</td>\n",
       "      <td>0.212718</td>\n",
       "      <td>0.212718</td>\n",
       "      <td>0.210709</td>\n",
       "      <td>0.217586</td>\n",
       "      <td>0.216736</td>\n",
       "      <td>0.217432</td>\n",
       "      <td>0.212950</td>\n",
       "      <td>0.212950</td>\n",
       "      <td>...</td>\n",
       "      <td>0.263483</td>\n",
       "      <td>0.225313</td>\n",
       "      <td>0.202287</td>\n",
       "      <td>0.194251</td>\n",
       "      <td>0.221759</td>\n",
       "      <td>0.189306</td>\n",
       "      <td>0.155772</td>\n",
       "      <td>0.171380</td>\n",
       "      <td>0.266110</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>871 rows × 871 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             36314  11354606       679      2554    126941      3033   \n",
       "36314     1.000000  0.922423  0.917092  0.917092  0.919255  0.927368  \\\n",
       "11354606  0.922423  1.000000  0.917246  0.917246  0.917710  0.918869   \n",
       "679       0.917092  0.917246  1.000000  1.000000  0.924587  0.920028   \n",
       "2554      0.917092  0.917246  1.000000  1.000000  0.924587  0.920028   \n",
       "126941    0.919255  0.917710  0.924587  0.924587  1.000000  0.923582   \n",
       "...            ...       ...       ...       ...       ...       ...   \n",
       "3373      0.207078  0.213877  0.214264  0.214264  0.214109  0.209396   \n",
       "5917      0.218359  0.213414  0.213027  0.213027  0.214418  0.219286   \n",
       "442872    0.205378  0.211250  0.212255  0.212255  0.213027  0.212950   \n",
       "442021    0.211096  0.213723  0.213027  0.213027  0.217663  0.219595   \n",
       "4843      0.213723  0.209241  0.212718  0.212718  0.210709  0.217586   \n",
       "\n",
       "            444795      5757     19649      6279  ...   5281078  16129706   \n",
       "36314     0.922037  0.926905  0.918251  0.918251  ...  0.205069  0.214495  \\\n",
       "11354606  0.919873  0.926905  0.913769  0.913769  ...  0.207696  0.208623   \n",
       "679       0.919487  0.927755  0.915237  0.915237  ...  0.222145  0.213182   \n",
       "2554      0.919487  0.927755  0.915237  0.915237  ...  0.222145  0.213182   \n",
       "126941    0.919332  0.928991  0.916783  0.916783  ...  0.224308  0.215500   \n",
       "...            ...       ...       ...       ...  ...       ...       ...   \n",
       "3373      0.218127  0.215268  0.205223  0.205223  ...  0.186215  0.305362   \n",
       "5917      0.219518  0.221140  0.211405  0.211405  ...  0.205532  0.241539   \n",
       "442872    0.215809  0.220522  0.207850  0.207850  ...  0.252357  0.260547   \n",
       "442021    0.212409  0.221913  0.214341  0.214341  ...  0.251584  0.219595   \n",
       "4843      0.216736  0.217432  0.212950  0.212950  ...  0.263483  0.225313   \n",
       "\n",
       "              2310      4375      3352      3373      5917    442872   \n",
       "36314     0.210478  0.216195  0.216350  0.207078  0.218359  0.205378  \\\n",
       "11354606  0.212023  0.215423  0.210168  0.213877  0.213414  0.211250   \n",
       "679       0.209628  0.214418  0.219363  0.214264  0.213027  0.212255   \n",
       "2554      0.209628  0.214418  0.219363  0.214264  0.213027  0.212255   \n",
       "126941    0.214573  0.216118  0.209473  0.214109  0.214418  0.213027   \n",
       "...            ...       ...       ...       ...       ...       ...   \n",
       "3373      0.331479  0.281409  0.159017  1.000000  0.329161  0.214650   \n",
       "5917      0.248802  0.298563  0.160408  0.329161  1.000000  0.231494   \n",
       "442872    0.208159  0.257302  0.156854  0.214650  0.231494  1.000000   \n",
       "442021    0.246948  0.237058  0.225931  0.259002  0.188688  0.242621   \n",
       "4843      0.202287  0.194251  0.221759  0.189306  0.155772  0.171380   \n",
       "\n",
       "            442021      4843  \n",
       "36314     0.211096  0.213723  \n",
       "11354606  0.213723  0.209241  \n",
       "679       0.213027  0.212718  \n",
       "2554      0.213027  0.212718  \n",
       "126941    0.217663  0.210709  \n",
       "...            ...       ...  \n",
       "3373      0.259002  0.189306  \n",
       "5917      0.188688  0.155772  \n",
       "442872    0.242621  0.171380  \n",
       "442021    1.000000  0.266110  \n",
       "4843      0.266110  1.000000  \n",
       "\n",
       "[871 rows x 871 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if (not os.path.exists(predict_folder+\"signature_similarity_df.csv\")):\n",
    "\n",
    "    from sklearn.metrics import pairwise_distances\n",
    "\n",
    "    S = pd.read_csv(paths_global.data_folder+\"TRANSCRIPT_v1.0.0/all_drugs_+LINCS.csv\", index_col=0, header=0)\n",
    "    ## Look only at the direction of differential expression\n",
    "    S[S<0] = -1\n",
    "    S[S>0] = 1\n",
    "    # sum of mutual up/down-regulated genes divided by total number of genes\n",
    "    Dists = (S.shape[0]-pairwise_distances(S.T, metric='l1'))/S.shape[0] \n",
    "    signature_custom = pd.DataFrame(Dists, index=S.columns, columns=S.columns)\n",
    "    \n",
    "    signature_custom.index = signature_custom.index.astype(str)\n",
    "    signature_custom.columns = signature_custom.columns.astype(str)\n",
    "    \n",
    "    signature_custom.to_csv(predict_folder+\"signature_similarity_df.csv\")\n",
    "    \n",
    "signature_custom = pd.read_csv(predict_folder+\"signature_similarity_df.csv\", index_col=0)\n",
    "signature_custom"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## III. Disease-disease similarity database\n",
    "\n",
    "### III.1 Disease-disease similarity: Phenotype disease-disease similarities\n",
    "\n",
    "Identifying similarity between MeSH terms appearing in the medical description of diseases from the OMIM database. \n",
    "\n",
    "\"We used the phenotypic similarity constructed by [van Driel et al (2006)](https://www.nature.com/articles/5201585.pdf). The phenotypic similarity was constructed by identifying similarity between MeSH terms (Lipscomb, 2000) appearing in the medical description of diseases from the OMIM database (Hamosh et al, 2002)\". We actually use a posterior package for computing similarities, described in [this paper](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-020-03583-6).\n",
    "\n",
    "In order to run this code, you need to sign up to [a UMLS Terminology Services Account](https://uts.nlm.nih.gov/uts/signup-login) (with your academic e-mail address for instance). The review of the registration application takes up to 3 business days. Then\n",
    "1. Download file [public_mm_linux_main_2020.tar.bz2](https://data.lhncbc.nlm.nih.gov/umls-restricted/ii/tools/MetaMap/download/public_mm_linux_main_2020.tar.bz2) once logged in\n",
    "2. Place that file in folder `~/`\n",
    "\n",
    "For more information about MetaMap, please check the [documentation](https://lhncbc.nlm.nih.gov/ii/tools/MetaMap/Docs/README.html#Getting%20the%20distribution).\n",
    "\n",
    "This yields a dataframe `disease_phenotype_similarity_df.csv` of size #diseases x #diseases, where identifiers are MedGen Concept ID identifiers, which contains the similarity between disease phenotype annotations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>C0042133</th>\n",
       "      <th>C1704272</th>\n",
       "      <th>C1858361</th>\n",
       "      <th>C4310232</th>\n",
       "      <th>C0029456</th>\n",
       "      <th>C1851377</th>\n",
       "      <th>C0270327</th>\n",
       "      <th>C1833268</th>\n",
       "      <th>C0024141</th>\n",
       "      <th>C1852092</th>\n",
       "      <th>...</th>\n",
       "      <th>C0012813</th>\n",
       "      <th>C0014474</th>\n",
       "      <th>C1845903</th>\n",
       "      <th>C1956346</th>\n",
       "      <th>C0242770</th>\n",
       "      <th>C0022661</th>\n",
       "      <th>C0236792</th>\n",
       "      <th>C1135191</th>\n",
       "      <th>C0016667</th>\n",
       "      <th>C0039445</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C0042133</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.316805</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1704272</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.307626</td>\n",
       "      <td>0.307626</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.388069</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1858361</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.422435</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.390624</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.398025</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C4310232</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.430398</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.432798</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C0029456</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.422435</td>\n",
       "      <td>0.430398</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.477832</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C0022661</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.388069</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.427991</td>\n",
       "      <td>0.427991</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C0236792</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1135191</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.391671</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.245870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C0016667</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.481842</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.839769</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.401891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C0039445</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.466000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.409505</td>\n",
       "      <td>0.359964</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.24587</td>\n",
       "      <td>0.401891</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>906 rows × 906 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          C0042133  C1704272  C1858361  C4310232  C0029456  C1851377   \n",
       "C0042133       0.0  0.000000  0.000000  0.000000  0.000000  0.000000  \\\n",
       "C1704272       0.0  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "C1858361       0.0  0.000000  0.000000  0.000000  0.422435  0.000000   \n",
       "C4310232       0.0  0.000000  0.000000  0.000000  0.430398  0.000000   \n",
       "C0029456       0.0  0.000000  0.422435  0.430398  0.000000  0.000000   \n",
       "...            ...       ...       ...       ...       ...       ...   \n",
       "C0022661       0.0  0.388069  0.000000  0.000000  0.000000  0.000000   \n",
       "C0236792       0.0  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "C1135191       0.0  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "C0016667       0.0  0.000000  0.000000  0.000000  0.000000  0.481842   \n",
       "C0039445       0.0  0.000000  0.000000  0.000000  0.000000  0.466000   \n",
       "\n",
       "          C0270327  C1833268  C0024141  C1852092  ...  C0012813  C0014474   \n",
       "C0042133  0.000000  0.000000  0.000000  0.000000  ...  0.000000  0.316805  \\\n",
       "C1704272  0.307626  0.307626  0.000000  0.000000  ...  0.000000  0.000000   \n",
       "C1858361  0.000000  0.000000  0.390624  0.000000  ...  0.398025  0.000000   \n",
       "C4310232  0.000000  0.000000  0.000000  0.432798  ...  0.000000  0.000000   \n",
       "C0029456  0.000000  0.000000  0.000000  0.477832  ...  0.000000  0.000000   \n",
       "...            ...       ...       ...       ...  ...       ...       ...   \n",
       "C0022661  0.427991  0.427991  0.000000  0.000000  ...  0.000000  0.000000   \n",
       "C0236792  0.000000  0.000000  0.000000  0.000000  ...  0.000000  0.000000   \n",
       "C1135191  0.000000  0.000000  0.000000  0.000000  ...  0.000000  0.000000   \n",
       "C0016667  0.000000  0.000000  0.000000  0.000000  ...  0.000000  0.000000   \n",
       "C0039445  0.000000  0.000000  0.000000  0.000000  ...  0.000000  0.000000   \n",
       "\n",
       "          C1845903  C1956346  C0242770  C0022661  C0236792  C1135191   \n",
       "C0042133  0.000000  0.000000       0.0  0.000000       0.0   0.00000  \\\n",
       "C1704272  0.000000  0.000000       0.0  0.388069       0.0   0.00000   \n",
       "C1858361  0.000000  0.000000       0.0  0.000000       0.0   0.00000   \n",
       "C4310232  0.000000  0.000000       0.0  0.000000       0.0   0.00000   \n",
       "C0029456  0.000000  0.000000       0.0  0.000000       0.0   0.00000   \n",
       "...            ...       ...       ...       ...       ...       ...   \n",
       "C0022661  0.000000  0.000000       0.0  0.000000       0.0   0.00000   \n",
       "C0236792  0.000000  0.000000       0.0  0.000000       0.0   0.00000   \n",
       "C1135191  0.000000  0.391671       0.0  0.000000       0.0   0.00000   \n",
       "C0016667  0.839769  0.000000       0.0  0.000000       0.0   0.00000   \n",
       "C0039445  0.409505  0.359964       0.0  0.000000       0.0   0.24587   \n",
       "\n",
       "          C0016667  C0039445  \n",
       "C0042133  0.000000  0.000000  \n",
       "C1704272  0.000000  0.000000  \n",
       "C1858361  0.000000  0.000000  \n",
       "C4310232  0.000000  0.000000  \n",
       "C0029456  0.000000  0.000000  \n",
       "...            ...       ...  \n",
       "C0022661  0.000000  0.000000  \n",
       "C0236792  0.000000  0.000000  \n",
       "C1135191  0.000000  0.245870  \n",
       "C0016667  0.000000  0.401891  \n",
       "C0039445  0.401891  0.000000  \n",
       "\n",
       "[906 rows x 906 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if (not os.path.exists(predict_folder+\"disease_phenotype_similarity_df.csv\")):\n",
    "\n",
    "    A = pd.read_csv(paths_global.data_folder+\"FEATURELESS_v1.0.0/all_ratings.csv\", index_col=0)\n",
    "    sb.Popen([\"mkdir\", \"-p\", predict_folder+\"DISEASE_PHENOTYPE_SIMILARITY/\"])\n",
    "\n",
    "    ## Retrieve MeSH terms from MedGen Concept IDs\n",
    "    if (not os.path.exists(predict_folder+\"DISEASE_PHENOTYPE_SIMILARITY/disease_phenotype_MeshTERMS.out\")):\n",
    "        with open(\"diseases.txt\", \"w+\") as f:\n",
    "            f.write(','.join(list(A.columns)))  \n",
    "        process = sb.Popen(\"bash ../src/Bashscript_PHENOsim.sh\")\n",
    "        process.wait()\n",
    "        sb.Popen([\"mv\", \"disease_phenotype_MeshTERMS.out\", predict_folder+\"DISEASE_PHENOTYPE_SIMILARITY/\"])\n",
    "\n",
    "    disease_MeshTERMS = pd.read_csv(predict_folder+\"DISEASE_PHENOTYPE_SIMILARITY/disease_phenotype_MeshTERMS.out\", index_col=None, header=None)\n",
    "    disease_MeshTERMS.columns = [\"Disease UMLS\",\"MeSH ID\"]\n",
    "    disease_MeshTERMS = disease_MeshTERMS.loc[disease_MeshTERMS[\"MeSH ID\"].astype(str)!=\" None\"]\n",
    "    disease_MeshTERMS[\"MeSH ID\"] = [x[2:-1] for x in list(disease_MeshTERMS[\"MeSH ID\"])]\n",
    "    disease_MeshTERMS.index = list(disease_MeshTERMS[\"MeSH ID\"])\n",
    "    disease_MeshTERMS[\"Disease UMLS\"] = [x[1:-1] for x in list(disease_MeshTERMS[\"Disease UMLS\"])]\n",
    "    mesh_terms = list(disease_MeshTERMS[\"MeSH ID\"])\n",
    "\n",
    "    ## Compute MeSH term similarity\n",
    "    if (not os.path.exists(predict_folder+\"DISEASE_PHENOTYPE_SIMILARITY/disease_phenotype_similarity.out\")):\n",
    "        method, category = \"lin\", \"C\"\n",
    "        cmd = [\"source ~/miniconda3/etc/profile.d/conda.sh\"]\n",
    "        cmd += [\"conda activate pysim\"]\n",
    "        cmd += [\"~/public_mm/bin/skrmedpostctl start\"]\n",
    "        cmd += [\"~/public_mm/bin/wsdserverctl start\"]\n",
    "        cmd_py = [\"diseases_=[\\\"\"+\"\\\",\\\"\".join(mesh_terms)+\"\\\"]\"]\n",
    "        cmd_py += [\"from pyMeSHSim.Sim.similarity import termComp\"]\n",
    "        cmd_py += [\"import pandas as pd\"]\n",
    "        cmd_py += [\"import numpy as np\"]\n",
    "        cmd_py += [\"simCom = termComp()\"]\n",
    "        cmd_py += ['calcSim = lambda xii,xjj: simCom.termSim(dui1=xii, dui2=xjj, method=\\\"'+method+'\\\", category=\\\"'+category+'\\\")']\n",
    "        cmd_py += ['exec(\"def calcSimWrapper(xii, xjj):\\\\n\\\\ttry:\\\\n\\\\t\\\\treturn calcSim(xii,xjj)\\\\n\\\\texcept:\\\\n\\\\t\\\\treturn 0\\\\n\")']\n",
    "        cmd_py += [\"simmat = pd.DataFrame([],index=diseases_)\"]\n",
    "        cmd_py += ['exec(\"for i, xi in enumerate(list(set(diseases_))):\\\\n\\\\tsimmat[xi] = [0]*(i+1)+[calcSimWrapper(xi,xj) for j, xj in enumerate(diseases_[(i+1):])]\\\\n\\\\n\")']\n",
    "        cmd_py += ['simmat.to_csv(\"disease_phenotype_similarity.out\")']\n",
    "        cmd += [\"python3 -c \\'\"+\";\".join(cmd_py)+\"\\'\"]\n",
    "        cmd += [\"~/public_mm/bin/wsdserverctl stop\"]\n",
    "        cmd += [\"~/public_mm/bin/skrmedpostctl stop\"]\n",
    "        cmd += [\"conda deactivate\"]\n",
    "        sb.Popen([y for x in cmd for y in x.split(\" \")+[\";\"]])\n",
    "        sb.Popen([\"mv\", \"disease_phenotype_similarity.out\", predict_folder+\"DISEASE_PHENOTYPE_SIMILARITY/\"])\n",
    "\n",
    "    simmat = pd.read_csv(predict_folder+\"DISEASE_PHENOTYPE_SIMILARITY/disease_phenotype_similarity.out\", index_col=0)\n",
    "    disease_phenotype_similarity_df = simmat[mesh_terms]\n",
    "    disease_phenotype_similarity_df += disease_phenotype_similarity_df.T\n",
    "    np.fill_diagonal(disease_phenotype_similarity_df.values,1)\n",
    "\n",
    "    disease_id = [str(disease_MeshTERMS.loc[x][\"Disease UMLS\"]) for x in mesh_terms]\n",
    "    disease_phenotype_similarity_df.index = disease_id\n",
    "    disease_phenotype_similarity_df.columns = disease_id\n",
    "\n",
    "    disease_phenotype_similarity_df.index = disease_phenotype_similarity_df.index.astype(str)\n",
    "    disease_phenotype_similarity_df.columns = disease_phenotype_similarity_df.columns.astype(str)\n",
    "    \n",
    "    disease_phenotype_similarity_df.to_csv(predict_folder+\"disease_phenotype_similarity_df.csv\")\n",
    "\n",
    "disease_phenotype_custom = pd.read_csv(predict_folder+\"disease_phenotype_similarity_df.csv\", index_col=0)\n",
    "disease_phenotype_custom"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### III.2 Disease-disease similarity: Semantic phenotypic disease-disease similarities \n",
    "\n",
    "\"Hierarchical structure of the HPO together with the mapping provided by HPO between ontology nodes and OMIM diseases to construct a semantic similarity score based on Resnik.\"\n",
    "\n",
    "This yields a dataframe `disease_semantic_similarity_df.csv` of size #diseases x #diseases, where identifiers are MedGen Concept ID identifiers, which contains the semantic similarity score between disease ontologies. We use scripts from the following [repository](https://github.com/DGRC-PT/HPOSim_Helper_Scripts)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>C0008732</th>\n",
       "      <th>C1832855</th>\n",
       "      <th>C1868001</th>\n",
       "      <th>C1852223</th>\n",
       "      <th>C1843771</th>\n",
       "      <th>C1852092</th>\n",
       "      <th>C0270853</th>\n",
       "      <th>CN263178</th>\n",
       "      <th>C2750850</th>\n",
       "      <th>C1840264</th>\n",
       "      <th>...</th>\n",
       "      <th>C1840333</th>\n",
       "      <th>C5399971</th>\n",
       "      <th>C1861864</th>\n",
       "      <th>C1840172</th>\n",
       "      <th>C1850141</th>\n",
       "      <th>C1835047</th>\n",
       "      <th>C1843765</th>\n",
       "      <th>C1851055</th>\n",
       "      <th>C1840373</th>\n",
       "      <th>C3714927</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C0008732</th>\n",
       "      <td>6.385549</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.405463</td>\n",
       "      <td>3.836702</td>\n",
       "      <td>4.405463</td>\n",
       "      <td>4.405463</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.320906</td>\n",
       "      <td>2.202732</td>\n",
       "      <td>...</td>\n",
       "      <td>5.589068</td>\n",
       "      <td>3.031711</td>\n",
       "      <td>4.426020</td>\n",
       "      <td>3.070166</td>\n",
       "      <td>3.492540</td>\n",
       "      <td>0</td>\n",
       "      <td>4.405463</td>\n",
       "      <td>2.226376</td>\n",
       "      <td>3.086376</td>\n",
       "      <td>4.145003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1832855</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.439855</td>\n",
       "      <td>2.409154</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.606103</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.169734</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1868001</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.409154</td>\n",
       "      <td>5.359532</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.520726</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.409154</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.520726</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.619338</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.520726</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1852223</th>\n",
       "      <td>4.405463</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.421025</td>\n",
       "      <td>4.405463</td>\n",
       "      <td>4.579616</td>\n",
       "      <td>2.088279</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.088279</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.405463</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.405463</td>\n",
       "      <td>0</td>\n",
       "      <td>4.579616</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.405463</td>\n",
       "      <td>4.405463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1843771</th>\n",
       "      <td>3.836702</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.520726</td>\n",
       "      <td>4.405463</td>\n",
       "      <td>5.456336</td>\n",
       "      <td>4.405463</td>\n",
       "      <td>4.405463</td>\n",
       "      <td>2.077075</td>\n",
       "      <td>4.320906</td>\n",
       "      <td>4.282263</td>\n",
       "      <td>...</td>\n",
       "      <td>0.607193</td>\n",
       "      <td>4.614710</td>\n",
       "      <td>0.556594</td>\n",
       "      <td>3.020263</td>\n",
       "      <td>2.832083</td>\n",
       "      <td>0</td>\n",
       "      <td>4.884694</td>\n",
       "      <td>2.226376</td>\n",
       "      <td>4.085042</td>\n",
       "      <td>3.145428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1835047</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1843765</th>\n",
       "      <td>4.405463</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.579616</td>\n",
       "      <td>4.884694</td>\n",
       "      <td>4.798484</td>\n",
       "      <td>4.405463</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.667575</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.884694</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.405463</td>\n",
       "      <td>0</td>\n",
       "      <td>6.267000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.884694</td>\n",
       "      <td>4.405463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1851055</th>\n",
       "      <td>2.226376</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.226376</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.226376</td>\n",
       "      <td>2.226376</td>\n",
       "      <td>2.226376</td>\n",
       "      <td>4.401776</td>\n",
       "      <td>3.659651</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.672326</td>\n",
       "      <td>3.659651</td>\n",
       "      <td>2.226376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1840373</th>\n",
       "      <td>3.086376</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.520726</td>\n",
       "      <td>4.405463</td>\n",
       "      <td>4.085042</td>\n",
       "      <td>4.405463</td>\n",
       "      <td>4.405463</td>\n",
       "      <td>1.005913</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.282263</td>\n",
       "      <td>...</td>\n",
       "      <td>4.690214</td>\n",
       "      <td>3.350557</td>\n",
       "      <td>3.519859</td>\n",
       "      <td>1.994234</td>\n",
       "      <td>4.862964</td>\n",
       "      <td>0</td>\n",
       "      <td>4.884694</td>\n",
       "      <td>3.659651</td>\n",
       "      <td>4.559734</td>\n",
       "      <td>2.879483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3714927</th>\n",
       "      <td>4.145003</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.405463</td>\n",
       "      <td>3.145428</td>\n",
       "      <td>4.405463</td>\n",
       "      <td>4.405463</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.320906</td>\n",
       "      <td>2.202732</td>\n",
       "      <td>...</td>\n",
       "      <td>5.331829</td>\n",
       "      <td>3.031711</td>\n",
       "      <td>4.426020</td>\n",
       "      <td>3.010941</td>\n",
       "      <td>4.708483</td>\n",
       "      <td>0</td>\n",
       "      <td>4.405463</td>\n",
       "      <td>2.226376</td>\n",
       "      <td>2.879483</td>\n",
       "      <td>6.189577</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>264 rows × 264 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          C0008732  C1832855  C1868001  C1852223  C1843771  C1852092   \n",
       "C0008732  6.385549  0.000000  0.000000  4.405463  3.836702  4.405463  \\\n",
       "C1832855  0.000000  7.439855  2.409154  0.000000  0.000000  0.000000   \n",
       "C1868001  0.000000  2.409154  5.359532  0.000000  2.520726  0.000000   \n",
       "C1852223  4.405463  0.000000  0.000000  3.421025  4.405463  4.579616   \n",
       "C1843771  3.836702  0.000000  2.520726  4.405463  5.456336  4.405463   \n",
       "...            ...       ...       ...       ...       ...       ...   \n",
       "C1835047  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "C1843765  4.405463  0.000000  0.000000  4.579616  4.884694  4.798484   \n",
       "C1851055  2.226376  0.000000  0.000000  0.000000  2.226376  0.000000   \n",
       "C1840373  3.086376  0.000000  2.520726  4.405463  4.085042  4.405463   \n",
       "C3714927  4.145003  0.000000  0.000000  4.405463  3.145428  4.405463   \n",
       "\n",
       "          C0270853  CN263178  C2750850  C1840264  ...  C1840333  C5399971   \n",
       "C0008732  4.405463  0.000000  4.320906  2.202732  ...  5.589068  3.031711  \\\n",
       "C1832855  0.000000  1.606103  0.000000  0.000000  ...  0.000000  0.000000   \n",
       "C1868001  0.000000  2.409154  0.000000  0.000000  ...  0.000000  2.520726   \n",
       "C1852223  2.088279  0.000000  0.000000  2.088279  ...  0.000000  4.405463   \n",
       "C1843771  4.405463  2.077075  4.320906  4.282263  ...  0.607193  4.614710   \n",
       "...            ...       ...       ...       ...  ...       ...       ...   \n",
       "C1835047  0.000000  0.000000  0.000000  0.000000  ...  0.000000  0.000000   \n",
       "C1843765  4.405463  0.000000  0.000000  2.667575  ...  0.000000  4.884694   \n",
       "C1851055  0.000000  0.000000  0.000000  0.000000  ...  2.226376  2.226376   \n",
       "C1840373  4.405463  1.005913  0.000000  4.282263  ...  4.690214  3.350557   \n",
       "C3714927  4.405463  0.000000  4.320906  2.202732  ...  5.331829  3.031711   \n",
       "\n",
       "          C1861864  C1840172  C1850141  C1835047  C1843765  C1851055   \n",
       "C0008732  4.426020  3.070166  3.492540         0  4.405463  2.226376  \\\n",
       "C1832855  0.000000  5.169734  0.000000         0  0.000000  0.000000   \n",
       "C1868001  0.000000  3.619338  0.000000         0  0.000000  0.000000   \n",
       "C1852223  0.000000  0.000000  4.405463         0  4.579616  0.000000   \n",
       "C1843771  0.556594  3.020263  2.832083         0  4.884694  2.226376   \n",
       "...            ...       ...       ...       ...       ...       ...   \n",
       "C1835047  0.000000  0.000000  0.000000         0  0.000000  0.000000   \n",
       "C1843765  0.000000  0.000000  4.405463         0  6.267000  0.000000   \n",
       "C1851055  2.226376  4.401776  3.659651         0  0.000000  6.672326   \n",
       "C1840373  3.519859  1.994234  4.862964         0  4.884694  3.659651   \n",
       "C3714927  4.426020  3.010941  4.708483         0  4.405463  2.226376   \n",
       "\n",
       "          C1840373  C3714927  \n",
       "C0008732  3.086376  4.145003  \n",
       "C1832855  0.000000  0.000000  \n",
       "C1868001  2.520726  0.000000  \n",
       "C1852223  4.405463  4.405463  \n",
       "C1843771  4.085042  3.145428  \n",
       "...            ...       ...  \n",
       "C1835047  0.000000  0.000000  \n",
       "C1843765  4.884694  4.405463  \n",
       "C1851055  3.659651  2.226376  \n",
       "C1840373  4.559734  2.879483  \n",
       "C3714927  2.879483  6.189577  \n",
       "\n",
       "[264 rows x 264 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "semantic_folder = predict_folder+\"DISEASE_SEMANTIC_SIMILARITY/\"\n",
    "sb.Popen((\"mkdir -p \"+semantic_folder).split(\" \"))\n",
    "\n",
    "if (not os.path.exists(predict_folder+\"disease_semantic_similarity_df.csv\")):\n",
    "\n",
    "    A = pd.read_csv(paths_global.data_folder+\"FEATURELESS_v1.0.0/all_ratings.csv\", index_col=0)\n",
    "    diseases = list(A.columns) # MedGen Concept IDs\n",
    "    disease_list = list(set([di_medgenid2diseasename[disease.split(\".\")[0]] for disease in diseases]))\n",
    "\n",
    "    inv_map = {v:k for k, v in di_omimid2diseasename.items()}\n",
    "    disease_list_ = [\"OMIM:\"+(inv_map.get(d)[1:]) for d in disease_list if (d in inv_map)]\n",
    "\n",
    "    if (not os.path.exists(semantic_folder+\"disease_semantic_similarity.out\")):\n",
    "\n",
    "        process = sb.Popen(\"bash ../src/Bashscript_SEMANTICsim.sh\".split(\" \"))\n",
    "        process.wait()\n",
    "\n",
    "        with open(\"diseases.txt\", \"w\") as f:\n",
    "            f.write(\",\".join(disease_list_)+\"\\n\")\n",
    "\n",
    "        sb.Popen((\"mkdir -p \"+semantic_folder).split(\" \"))\n",
    "        cmd_R = [\"library(HPOSim);ls<-unlist(read.delim(\\\"diseases.txt\\\", sep=\\\",\\\", header=F))\"]\n",
    "        cmd_R += [\"mat<-getDiseaseListSim(ls,combinemethod=\\\"funSimMax\\\",method=\\\"Resnik\\\",verbose=TRUE)\"]\n",
    "        cmd_R += [\"write.csv(mat,\\\"disease_semantic_similarity.out\\\")\"]\n",
    "        sb.Popen(\"R -e '\"+(\";\".join(cmd_R)))\n",
    "\n",
    "        sb.Popen(\"rm diseases.txt\".split(\" \"))\n",
    "        sb.Popen((\"mv disease_semantic_similarity.out \"+semantic_folder).split(\" \"))\n",
    "\n",
    "    disease_semantic_similarity_df = pd.read_csv(semantic_folder+\"disease_semantic_similarity.out\", index_col=0)\n",
    "    disease_semantic_similarity_df.index = omim_list_\n",
    "    disease_semantic_similarity_df.columns = omim_list_\n",
    "    disease_list_ = [di_omimid2diseasename[\"D\"+x.split(\"OMIM:\")[-1]] for x in omim_list_]\n",
    "    invert_map = {v: k for k,v in di_medgenid2diseasename.items()}\n",
    "    disease_list_ = [invert_map.get(d,None) for d in disease_list_ if (invert_map.get(d,None) is not None)]\n",
    "    disease_semantic_similarity_df.index = disease_list_\n",
    "    disease_semantic_similarity_df.columns = disease_list_\n",
    "    disease_semantic_custom = disease_semantic_similarity_df\n",
    "    disease_semantic_custom.index = disease_semantic_custom.index.astype(str)\n",
    "    disease_semantic_custom.columns = disease_semantic_custom.columns.astype(str)\n",
    "    disease_semantic_custom.to_csv(predict_folder+\"disease_semantic_similarity_df.csv\")\n",
    "    \n",
    "disease_semantic_similarity_df = pd.read_csv(predict_folder+\"disease_semantic_similarity_df.csv\", index_col=0)\n",
    "disease_semantic_similarity_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### III.3 Disease-disease similarity: Genetic based disease-disease similarities\n",
    "\n",
    "\"Given genetic signatures of diseases obtained from gene expression experiments, we used a Jaccard score between every pair of signatures, taking into account the direction of the response of each gene. That is, the total number of mutual upregulated genes and mutual downregulated genes over the unified list of all genes. Signature genes with inconsistent regulation directionality for the same disease across various experiments (*i*.*e*., registered as both upregulated and downregulated across various experiments for the same disease) were filtered, allowing for up to 10% expression measurement errors.\" \n",
    "\n",
    "Transcriptomic phenotypes were obtained in the notebook *TRANSCRIPT_dataset.ipynb*.\n",
    "\n",
    "This yields a dataframe `disease_genetic_similarity_df.csv` of size #diseases x #diseases, where identifiers are MedGen Concept ID identifiers, which contains the Jaccard similarity score between disease transcriptomic phenotypes.\n",
    "\n",
    "Similar to Section [**II.6 Drug-drug similarity 6: Transcriptomic-based**](http://localhost:8888/notebooks/PREDICT_dataset.ipynb#II.6-Drug-drug-similarity-6:-Transcriptomic-based)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>C2936783</th>\n",
       "      <th>C2239176</th>\n",
       "      <th>C3553462</th>\n",
       "      <th>C0035235</th>\n",
       "      <th>C0032285</th>\n",
       "      <th>C0010346</th>\n",
       "      <th>C0009324</th>\n",
       "      <th>C0029408</th>\n",
       "      <th>C0001973</th>\n",
       "      <th>C3495559</th>\n",
       "      <th>...</th>\n",
       "      <th>C1260899</th>\n",
       "      <th>C1527336</th>\n",
       "      <th>C0275804</th>\n",
       "      <th>C0017168</th>\n",
       "      <th>C0003615</th>\n",
       "      <th>C0003872</th>\n",
       "      <th>C0014544</th>\n",
       "      <th>C0040028</th>\n",
       "      <th>C0040034</th>\n",
       "      <th>C0018802</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C2936783</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.913972</td>\n",
       "      <td>0.913972</td>\n",
       "      <td>0.920584</td>\n",
       "      <td>0.928940</td>\n",
       "      <td>0.928940</td>\n",
       "      <td>0.925816</td>\n",
       "      <td>0.927487</td>\n",
       "      <td>0.924435</td>\n",
       "      <td>...</td>\n",
       "      <td>0.924508</td>\n",
       "      <td>0.930102</td>\n",
       "      <td>0.933953</td>\n",
       "      <td>0.926760</td>\n",
       "      <td>0.931992</td>\n",
       "      <td>0.914554</td>\n",
       "      <td>0.926687</td>\n",
       "      <td>0.918768</td>\n",
       "      <td>0.930538</td>\n",
       "      <td>0.929449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C2239176</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.913972</td>\n",
       "      <td>0.913972</td>\n",
       "      <td>0.920584</td>\n",
       "      <td>0.928940</td>\n",
       "      <td>0.928940</td>\n",
       "      <td>0.925816</td>\n",
       "      <td>0.927487</td>\n",
       "      <td>0.924435</td>\n",
       "      <td>...</td>\n",
       "      <td>0.924508</td>\n",
       "      <td>0.930102</td>\n",
       "      <td>0.933953</td>\n",
       "      <td>0.926760</td>\n",
       "      <td>0.931992</td>\n",
       "      <td>0.914554</td>\n",
       "      <td>0.926687</td>\n",
       "      <td>0.918768</td>\n",
       "      <td>0.930538</td>\n",
       "      <td>0.929449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3553462</th>\n",
       "      <td>0.913972</td>\n",
       "      <td>0.913972</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.913900</td>\n",
       "      <td>0.914699</td>\n",
       "      <td>0.913246</td>\n",
       "      <td>0.913173</td>\n",
       "      <td>0.923418</td>\n",
       "      <td>0.913246</td>\n",
       "      <td>...</td>\n",
       "      <td>0.913754</td>\n",
       "      <td>0.914699</td>\n",
       "      <td>0.928141</td>\n",
       "      <td>0.913536</td>\n",
       "      <td>0.913100</td>\n",
       "      <td>0.915716</td>\n",
       "      <td>0.914045</td>\n",
       "      <td>0.913391</td>\n",
       "      <td>0.925452</td>\n",
       "      <td>0.923491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C0035235</th>\n",
       "      <td>0.913972</td>\n",
       "      <td>0.913972</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.913900</td>\n",
       "      <td>0.914699</td>\n",
       "      <td>0.913246</td>\n",
       "      <td>0.913173</td>\n",
       "      <td>0.923418</td>\n",
       "      <td>0.913246</td>\n",
       "      <td>...</td>\n",
       "      <td>0.913754</td>\n",
       "      <td>0.914699</td>\n",
       "      <td>0.928141</td>\n",
       "      <td>0.913536</td>\n",
       "      <td>0.913100</td>\n",
       "      <td>0.915716</td>\n",
       "      <td>0.914045</td>\n",
       "      <td>0.913391</td>\n",
       "      <td>0.925452</td>\n",
       "      <td>0.923491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C0032285</th>\n",
       "      <td>0.920584</td>\n",
       "      <td>0.920584</td>\n",
       "      <td>0.913900</td>\n",
       "      <td>0.913900</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.921456</td>\n",
       "      <td>0.921311</td>\n",
       "      <td>0.917024</td>\n",
       "      <td>0.927559</td>\n",
       "      <td>0.919712</td>\n",
       "      <td>...</td>\n",
       "      <td>0.922982</td>\n",
       "      <td>0.920294</td>\n",
       "      <td>0.931846</td>\n",
       "      <td>0.921601</td>\n",
       "      <td>0.922183</td>\n",
       "      <td>0.919858</td>\n",
       "      <td>0.921819</td>\n",
       "      <td>0.923491</td>\n",
       "      <td>0.927850</td>\n",
       "      <td>0.926179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C0003872</th>\n",
       "      <td>0.914554</td>\n",
       "      <td>0.914554</td>\n",
       "      <td>0.915716</td>\n",
       "      <td>0.915716</td>\n",
       "      <td>0.919858</td>\n",
       "      <td>0.915280</td>\n",
       "      <td>0.914554</td>\n",
       "      <td>0.913609</td>\n",
       "      <td>0.923563</td>\n",
       "      <td>0.921529</td>\n",
       "      <td>...</td>\n",
       "      <td>0.914336</td>\n",
       "      <td>0.915280</td>\n",
       "      <td>0.928577</td>\n",
       "      <td>0.913100</td>\n",
       "      <td>0.915861</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.918114</td>\n",
       "      <td>0.915571</td>\n",
       "      <td>0.925888</td>\n",
       "      <td>0.924217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C0014544</th>\n",
       "      <td>0.926687</td>\n",
       "      <td>0.926687</td>\n",
       "      <td>0.914045</td>\n",
       "      <td>0.914045</td>\n",
       "      <td>0.921819</td>\n",
       "      <td>0.930902</td>\n",
       "      <td>0.921601</td>\n",
       "      <td>0.929085</td>\n",
       "      <td>0.931047</td>\n",
       "      <td>0.932936</td>\n",
       "      <td>...</td>\n",
       "      <td>0.925307</td>\n",
       "      <td>0.926833</td>\n",
       "      <td>0.934026</td>\n",
       "      <td>0.923926</td>\n",
       "      <td>0.930902</td>\n",
       "      <td>0.918114</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.921601</td>\n",
       "      <td>0.931192</td>\n",
       "      <td>0.929521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C0040028</th>\n",
       "      <td>0.918768</td>\n",
       "      <td>0.918768</td>\n",
       "      <td>0.913391</td>\n",
       "      <td>0.913391</td>\n",
       "      <td>0.923491</td>\n",
       "      <td>0.924580</td>\n",
       "      <td>0.921093</td>\n",
       "      <td>0.922328</td>\n",
       "      <td>0.928795</td>\n",
       "      <td>0.922401</td>\n",
       "      <td>...</td>\n",
       "      <td>0.924798</td>\n",
       "      <td>0.923127</td>\n",
       "      <td>0.931338</td>\n",
       "      <td>0.924726</td>\n",
       "      <td>0.922110</td>\n",
       "      <td>0.915571</td>\n",
       "      <td>0.921601</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.927777</td>\n",
       "      <td>0.926978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C0040034</th>\n",
       "      <td>0.930538</td>\n",
       "      <td>0.930538</td>\n",
       "      <td>0.925452</td>\n",
       "      <td>0.925452</td>\n",
       "      <td>0.927850</td>\n",
       "      <td>0.928213</td>\n",
       "      <td>0.930538</td>\n",
       "      <td>0.929884</td>\n",
       "      <td>0.942454</td>\n",
       "      <td>0.934607</td>\n",
       "      <td>...</td>\n",
       "      <td>0.928722</td>\n",
       "      <td>0.929085</td>\n",
       "      <td>0.948194</td>\n",
       "      <td>0.929521</td>\n",
       "      <td>0.933735</td>\n",
       "      <td>0.925888</td>\n",
       "      <td>0.931192</td>\n",
       "      <td>0.927777</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.941510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C0018802</th>\n",
       "      <td>0.929449</td>\n",
       "      <td>0.929449</td>\n",
       "      <td>0.923491</td>\n",
       "      <td>0.923491</td>\n",
       "      <td>0.926179</td>\n",
       "      <td>0.926397</td>\n",
       "      <td>0.926978</td>\n",
       "      <td>0.927051</td>\n",
       "      <td>0.938022</td>\n",
       "      <td>0.927123</td>\n",
       "      <td>...</td>\n",
       "      <td>0.928359</td>\n",
       "      <td>0.927995</td>\n",
       "      <td>0.946233</td>\n",
       "      <td>0.927414</td>\n",
       "      <td>0.930466</td>\n",
       "      <td>0.924217</td>\n",
       "      <td>0.929521</td>\n",
       "      <td>0.926978</td>\n",
       "      <td>0.941510</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>144 rows × 144 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          C2936783  C2239176  C3553462  C0035235  C0032285  C0010346   \n",
       "C2936783  1.000000  1.000000  0.913972  0.913972  0.920584  0.928940  \\\n",
       "C2239176  1.000000  1.000000  0.913972  0.913972  0.920584  0.928940   \n",
       "C3553462  0.913972  0.913972  1.000000  1.000000  0.913900  0.914699   \n",
       "C0035235  0.913972  0.913972  1.000000  1.000000  0.913900  0.914699   \n",
       "C0032285  0.920584  0.920584  0.913900  0.913900  1.000000  0.921456   \n",
       "...            ...       ...       ...       ...       ...       ...   \n",
       "C0003872  0.914554  0.914554  0.915716  0.915716  0.919858  0.915280   \n",
       "C0014544  0.926687  0.926687  0.914045  0.914045  0.921819  0.930902   \n",
       "C0040028  0.918768  0.918768  0.913391  0.913391  0.923491  0.924580   \n",
       "C0040034  0.930538  0.930538  0.925452  0.925452  0.927850  0.928213   \n",
       "C0018802  0.929449  0.929449  0.923491  0.923491  0.926179  0.926397   \n",
       "\n",
       "          C0009324  C0029408  C0001973  C3495559  ...  C1260899  C1527336   \n",
       "C2936783  0.928940  0.925816  0.927487  0.924435  ...  0.924508  0.930102  \\\n",
       "C2239176  0.928940  0.925816  0.927487  0.924435  ...  0.924508  0.930102   \n",
       "C3553462  0.913246  0.913173  0.923418  0.913246  ...  0.913754  0.914699   \n",
       "C0035235  0.913246  0.913173  0.923418  0.913246  ...  0.913754  0.914699   \n",
       "C0032285  0.921311  0.917024  0.927559  0.919712  ...  0.922982  0.920294   \n",
       "...            ...       ...       ...       ...  ...       ...       ...   \n",
       "C0003872  0.914554  0.913609  0.923563  0.921529  ...  0.914336  0.915280   \n",
       "C0014544  0.921601  0.929085  0.931047  0.932936  ...  0.925307  0.926833   \n",
       "C0040028  0.921093  0.922328  0.928795  0.922401  ...  0.924798  0.923127   \n",
       "C0040034  0.930538  0.929884  0.942454  0.934607  ...  0.928722  0.929085   \n",
       "C0018802  0.926978  0.927051  0.938022  0.927123  ...  0.928359  0.927995   \n",
       "\n",
       "          C0275804  C0017168  C0003615  C0003872  C0014544  C0040028   \n",
       "C2936783  0.933953  0.926760  0.931992  0.914554  0.926687  0.918768  \\\n",
       "C2239176  0.933953  0.926760  0.931992  0.914554  0.926687  0.918768   \n",
       "C3553462  0.928141  0.913536  0.913100  0.915716  0.914045  0.913391   \n",
       "C0035235  0.928141  0.913536  0.913100  0.915716  0.914045  0.913391   \n",
       "C0032285  0.931846  0.921601  0.922183  0.919858  0.921819  0.923491   \n",
       "...            ...       ...       ...       ...       ...       ...   \n",
       "C0003872  0.928577  0.913100  0.915861  1.000000  0.918114  0.915571   \n",
       "C0014544  0.934026  0.923926  0.930902  0.918114  1.000000  0.921601   \n",
       "C0040028  0.931338  0.924726  0.922110  0.915571  0.921601  1.000000   \n",
       "C0040034  0.948194  0.929521  0.933735  0.925888  0.931192  0.927777   \n",
       "C0018802  0.946233  0.927414  0.930466  0.924217  0.929521  0.926978   \n",
       "\n",
       "          C0040034  C0018802  \n",
       "C2936783  0.930538  0.929449  \n",
       "C2239176  0.930538  0.929449  \n",
       "C3553462  0.925452  0.923491  \n",
       "C0035235  0.925452  0.923491  \n",
       "C0032285  0.927850  0.926179  \n",
       "...            ...       ...  \n",
       "C0003872  0.925888  0.924217  \n",
       "C0014544  0.931192  0.929521  \n",
       "C0040028  0.927777  0.926978  \n",
       "C0040034  1.000000  0.941510  \n",
       "C0018802  0.941510  1.000000  \n",
       "\n",
       "[144 rows x 144 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if (not os.path.exists(predict_folder+\"disease_signature_similarity_df.csv\")):\n",
    "\n",
    "    from sklearn.metrics import pairwise_distances\n",
    "\n",
    "    P = pd.read_csv(paths_global.data_folder+\"TRANSCRIPT_v1.0.0/all_diseases.csv\", index_col=0, header=0)\n",
    "    ## Look only at the direction of differential expression\n",
    "    P[P<0] = -1\n",
    "    P[P>0] = 1\n",
    "    # sum of mutual up/down-regulated genes divided by total number of genes\n",
    "    Dists = (P.shape[0]-pairwise_distances(P.T, metric='l1'))/P.shape[0] \n",
    "    disease_signature_custom = pd.DataFrame(Dists, index=P.columns, columns=P.columns)\n",
    "    disease_signature_custom.index = disease_signature_custom.index.astype(str)\n",
    "    disease_signature_custom.columns = disease_signature_custom.columns.astype(str)\n",
    "    disease_signature_custom.to_csv(predict_folder+\"disease_signature_similarity_df.csv\")\n",
    "\n",
    "disease_signature_custom = pd.read_csv(predict_folder+\"disease_signature_similarity_df.csv\", index_col=0)\n",
    "disease_signature_custom"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### III.4 Disease-disease similarity: Gene sequence based disease-disease similarities\n",
    "\n",
    "This yields a dataframe `disease_sequence_similarity_df.csv` of size #diseases x #diseases, where identifiers are MedGen Concept ID identifiers, which contains, when computable, average drug-pairwise alignment scores between gene target sequences. We use the normalization suggested in [Bleakley and Yamanishi (2009)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2735674/).\n",
    "\n",
    "Similar to Section [**II.3 Drug-drug similarity 3: (Target) sequence drug-drug similarities**](http://localhost:8888/notebooks/PREDICT_dataset.ipynb#II.3-Drug-drug-similarity-3:-(Target)-sequence-drug-drug-similarities). If one truly wants an equivalent to the computations made for drugs, they should use gene targets retrieved from DisGeNet."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "sb.Popen([\"mkdir\", \"-p\", predict_folder+\"DISEASE_TARGET_SIMILARITY/\"])\n",
    "\n",
    "from Bio import Align\n",
    "from Bio import SeqIO\n",
    "from time import time\n",
    "\n",
    "seq_cst_similarity_file = predict_folder+\"DISEASE_TARGET_SIMILARITY/disease_sequence_cst_similarity_df.txt\"\n",
    "if (not os.path.exists(seq_cst_similarity_file)):\n",
    "    sequence_aligner = Align.PairwiseAligner()\n",
    "    fname = drugbank_folder+\"TARGET SEQUENCES/Drug Target Sequences/gene.fasta\"\n",
    "    ## Target sequences\n",
    "    target_sequences = [record.seq for record in SeqIO.parse(fname, \"fasta\")]\n",
    "    \n",
    "    ## Following the normalization suggested in Bleakley and Yamanishi (2009)\n",
    "    from scipy.stats.mstats import gmean as geometric_mean\n",
    "    start=time()\n",
    "    sequence_CST_similarity = [sequence_aligner.score(seq, seq) for seq in target_sequences]\n",
    "    end=time()-end\n",
    "    print(\"Computation took %.2f sec.\" % (end))\n",
    "    norm_cst = geometric_mean(sequence_CST_similarity)\n",
    "    with open(seq_cst_similarity_file, \"w+\") as f:\n",
    "        f.write(str(norm_cst))\n",
    "else:\n",
    "    with open(seq_cst_similarity_file, \"r\") as f:\n",
    "        norm_cst = float(f.read())\n",
    "        \n",
    "seq_fname = paths_global.drugbank_folder+\"TARGET SEQUENCES/Drug Target Sequences/gene.fasta\"\n",
    "proteins_list = sb.check_output(\"cat '\"+seq_fname+\"' | grep '>drugbank_target|' | cut -d' ' -f1 | cut -d'|' -f2\", shell=True).decode(\"utf-8\").split(\"\\n\")[:-1]\n",
    "\n",
    "sb.Popen([\"mkdir\", \"-p\", predict_folder+\"DisGeNet/\"])\n",
    "\n",
    "if (not os.path.exists(predict_folder+\"DisGeNet/disease_db_DisGeNet.csv\")):\n",
    "    \n",
    "    from NORDic.UTILS.DISGENET_utils import get_user_key_DISGENET, get_genes_proteins_from_DISGENET\n",
    "    A = pd.read_csv(paths_global.data_folder+\"FEATURELESS_v1.0.0/all_ratings.csv\", index_col=0)\n",
    "    disease_list = list(A.columns)\n",
    "    \n",
    "    user_key = get_user_key_DISGENET(paths_global.disgenet_file)\n",
    "    disease_db = get_genes_proteins_from_DISGENET(disease_list, limit=3000, source=\"CURATED\", min_score=0.35, \n",
    "                                min_ei=1., min_dsi=0.25, min_dpi=0, chunksize=100, user_key=user_key, quiet=False)\n",
    "    \n",
    "    disease_db.to_csv(predict_folder+\"DisGeNet/disease_db_DisGeNet.csv\")\n",
    "    \n",
    "disease_db = pd.read_csv(predict_folder+\"DisGeNet/disease_db_DisGeNet.csv\", index_col=0)\n",
    "\n",
    "if (not os.path.exists(predict_folder+\"disease_sequence_similarity_df.csv\")):\n",
    "    sequence_aligner = Align.PairwiseAligner()\n",
    "    fname = paths_global.drugbank_folder+\"TARGET SEQUENCES/Drug Target Sequences/gene.fasta\"\n",
    "    ## Target sequences\n",
    "    target_sequences = [record.seq for record in SeqIO.parse(fname, \"fasta\")]\n",
    "    ndiseases=disease_db.shape[0]\n",
    "    \n",
    "    precompute_scores = {}\n",
    "    def compute_score(si,sj):\n",
    "        return precompute_scores.get((si,sj), sequence_aligner.score(si, sj))\n",
    "    \n",
    "    def disease_seq_function(i, j):\n",
    "        print((i,j,ndiseases))\n",
    "        if (i>j):\n",
    "            return 0.\n",
    "        else:\n",
    "            disease_i = disease_db.index[i]\n",
    "            disease_j = disease_db.index[j]\n",
    "            ## all sequences in which disease_i appears\n",
    "            target_i_ids = [ix for ix, protein in enumerate(proteins_list) if (protein in disease_db.loc[disease_i][\"Protein\"])]\n",
    "            target_seq_i = [target_sequences[idx_i] for idx_i in target_i_ids]\n",
    "            if (i==j):\n",
    "                align_scores = [compute_score(s1,s2) for s1 in target_seq_i for s2 in target_seq_i]\n",
    "            else:\n",
    "                ## all sequences in which disease_j appears\n",
    "                target_j_ids = [ix for ix, protein in enumerate(proteins_list) if (protein in disease_db.loc[disease_j][\"Protein\"])]\n",
    "                target_seq_j = [target_sequences[idx_j] for idx_j in target_j_ids]\n",
    "                align_scores = [compute_score(s1,s2) for s1 in target_seq_i for s2 in target_seq_j]\n",
    "            score = np.mean(align_scores) if (len(align_scores)>0) else np.nan\n",
    "            return score\n",
    "\n",
    "    start = time()\n",
    "    if (os.path.exists(predict_folder+\"DISEASE_TARGET_SIMILARITY/disease_sequence_similarity.out\")):\n",
    "        disease_sequence_similarity = np.loadtxt(predict_folder+\"DISEASE_TARGET_SIMILARITY/disease_sequence_similarity.out\")\n",
    "    else:\n",
    "        disease_sequence_similarity = np.zeros((ndiseases, ndiseases))\n",
    "    start_i = 0\n",
    "    for i in range(start_i, ndiseases):\n",
    "        for j in range(i, ndiseases):\n",
    "            disease_sequence_similarity[i,j] = disease_seq_function(i, j)\n",
    "        np.savetxt(predict_folder+\"DISEASE_TARGET_SIMILARITY/disease_sequence_similarity.out\", disease_sequence_similarity)\n",
    "    end = time()-start\n",
    "    print(\"Computation took %.2f sec.\" % end)\n",
    "\n",
    "    disease_sequence_similarity = np.loadtxt(predict_folder+\"DISEASE_TARGET_SIMILARITY/disease_sequence_similarity.out\")\n",
    "    disease_sequence_similarity_df = pd.DataFrame(disease_sequence_similarity, index=disease_db.index, columns=disease_db.index)\n",
    "    disease_sequence_similarity_df\n",
    "\n",
    "    diag = np.diag(disease_sequence_similarity_df.values)\n",
    "    disease_sequence_similarity_df += disease_sequence_similarity_df.T\n",
    "    np.fill_diagonal(disease_sequence_similarity_df.values, diag)\n",
    "    ## normalisation\n",
    "    disease_sequence_custom = disease_sequence_similarity_df/norm_cst\n",
    "    disease_sequence_custom.index = disease_sequence_custom.index.astype(str)\n",
    "    disease_sequence_custom.columns = disease_sequence_custom.columns.astype(str)\n",
    "    \n",
    "    disease_sequence_custom.to_csv(predict_folder+\"disease_sequence_similarity_df.csv\")\n",
    "\n",
    "disease_sequence_custom = pd.read_csv(predict_folder+\"disease_sequence_similarity_df.csv\", index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### III.5 Disease-disease similarity: Closeness in the human PPI network disease-disease similarities\n",
    "\n",
    "This yields a dataframe `disease_network_similarity_df.csv` of size #diseases x #diseases, where identifiers are MedGen Concept ID identifiers, which contains, when computable, the transformed distance between drug targets in the PPI using the Floyd-Warshall algorithm. We apply to the Floyd-Warshall cost matrix the transformation suggested by the [PREDICT paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3159979/).\n",
    "\n",
    "Similar to Section [**II.4 Drug-drug similarity 4: Closeness in the human PPI network drug-drug similarities**](http://localhost:8888/notebooks/PREDICT_dataset.ipynb#II.4-Drug-drug-similarity:-Closeness-in-the-human-PPI-network-drug-target-to-drug-target-similarities)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "network_folder = predict_folder+\"DISEASE_NETWORK_SIMILARITY/\"\n",
    "sb.Popen([\"mkdir\", \"-p\", network_folder])\n",
    "\n",
    "sb.Popen([\"mkdir\", \"-p\", predict_folder+\"DisGeNet/\"])\n",
    "if (not os.path.exists(predict_folder+\"DisGeNet/disease_db_DisGeNet.csv\")):\n",
    "    \n",
    "    from NORDic.UTILS.DISGENET_utils import get_user_key_DISGENET, get_genes_proteins_from_DISGENET\n",
    "    A = pd.read_csv(paths_global.data_folder+\"FEATURELESS_v1.0.0/all_ratings.csv\", index_col=0)\n",
    "    disease_list = list(A.columns)\n",
    "    \n",
    "    user_key = get_user_key_DISGENET(paths_global.disgenet_file)\n",
    "    disease_db = get_genes_proteins_from_DISGENET(disease_list, limit=3000, source=\"CURATED\", min_score=0.35, \n",
    "                                min_ei=1., min_dsi=0.25, min_dpi=0, chunksize=100, user_key=user_key, quiet=False)\n",
    "    \n",
    "    disease_db.to_csv(predict_folder+\"DisGeNet/disease_db_DisGeNet.csv\")\n",
    "    \n",
    "disease_db = pd.read_csv(predict_folder+\"DisGeNet/disease_db_DisGeNet.csv\", index_col=0)\n",
    "protein_targets = list(set([y for x in disease_db[\"Protein\"] for y in x.split(\"; \")]))\n",
    "\n",
    "if (not os.path.exists(predict_folder+\"disease_network_similarity_df.csv\")):\n",
    "    from NORDic.UTILS.STRING_utils import get_app_name_STRING, get_network_from_STRING\n",
    "    network_fname = network_folder+\"disease_human_ppi_network.csv\"\n",
    "    ## 9606=human, retrieve PPI from STRING\n",
    "    if (not os.path.exists(network_fname)):\n",
    "        gene_list = list(protein_db[\"Gene Name\"].unique())\n",
    "        taxon_id=9606 #human\n",
    "        app_name = get_app_name_STRING(paths_global.string_file)\n",
    "        ppi = get_network_from_STRING(gene_list, taxon_id, min_score=0, network_type=\"functional\", add_nodes=0, \n",
    "                                      app_name=app_name, version=\"11.5\", quiet=False)\n",
    "        ppi.to_csv(network_fname)\n",
    "    ppi = pd.read_csv(network_fname, index_col=0)\n",
    "\n",
    "    ## selection of interactions\n",
    "    score_thres=0\n",
    "\n",
    "    ## Get disease-gene associations\n",
    "    gene_targets = list(set([y for x in disease_db[\"Gene Name\"] for y in x.split(\"; \")]))\n",
    "\n",
    "    ppi = pd.read_csv(network_fname, sep=\"\\t\", header=0, index_col=0)\n",
    "    ppi_subset = ppi.loc[ppi[\"score\"]>score_thres][[\"preferredName_A\",\"preferredName_B\",\"score\"]]\n",
    "    ppi_subset = ppi_subset.loc[[(a in gene_targets) and (b in gene_targets) for a,b in zip(list(ppi_subset[\"preferredName_A\"]),list(ppi_subset[\"preferredName_B\"]))]]\n",
    "    ppi_subset[\"cost\"] = 1-ppi_subset[[\"score\"]].values.flatten() #cost\n",
    "    ppi_subset = ppi_subset[[\"preferredName_A\",\"preferredName_B\",\"cost\"]]\n",
    "    ppi_subset = ppi_subset.drop_duplicates(keep=\"first\")\n",
    "\n",
    "    ## 1 is the maximum cost of edge\n",
    "    max_cost=1.\n",
    "    wam = ppi_subset.pivot(index='preferredName_A', columns='preferredName_B', values='cost').fillna(max_cost)\n",
    "    for g in wam.index:\n",
    "        if (g not in wam.columns):\n",
    "            wam[g] = [max_cost]*wam.shape[0]\n",
    "    for g in wam.columns:\n",
    "        if (g not in wam.index):\n",
    "            wam.loc[g] = [max_cost]*wam.shape[1]\n",
    "    wam = wam.loc[wam.index][wam.index]\n",
    "\n",
    "    ## Compute distance between target genes\n",
    "    from scipy.sparse.csgraph import floyd_warshall\n",
    "\n",
    "    dists = floyd_warshall(wam.values.copy(order='C'), directed=False)\n",
    "    dists_df = pd.DataFrame(dists, index=wam.index, columns=wam.columns)\n",
    "\n",
    "    A_mat, b_mat = 0.9*np.exp(1), 1.\n",
    "    disease_network_similarity = A_mat*np.exp(-b_mat*dists_df)\n",
    "    disease_network_similarity /= np.max(disease_network_similarity)\n",
    "\n",
    "    diseases = pd.read_csv(ratings_file, index_col=0).columns\n",
    "    disease_network_similarity_1 = pd.DataFrame(0., index=diseases, columns=disease_network_similarity.columns)\n",
    "    for disease in diseases:\n",
    "        if (disease not in disease_db.index):\n",
    "            continue\n",
    "        genes_disease = disease_db.loc[disease][\"Gene Name\"].split(\"; \")\n",
    "        targeted_genes = [x for x in genes_disease if (x in disease_network_similarity.index)]\n",
    "        disease_network_similarity_1.loc[disease] = list(disease_network_similarity.loc[targeted_genes].mean(axis=0))\n",
    "\n",
    "    disease_network_similarity_df = pd.DataFrame([], index=diseases)\n",
    "    for disease in diseases:\n",
    "        if (disease not in disease_db.index):\n",
    "            disease_network_similarity_df[disease] = np.nan\n",
    "            continue\n",
    "        genes_disease = disease_db.loc[disease][\"Gene Name\"].split(\"; \")\n",
    "        targeted_genes = [x for x in genes_disease if (x in disease_network_similarity.index)]\n",
    "        disease_network_similarity_df[disease] = disease_network_similarity_1[targeted_genes].mean(axis=1)\n",
    "    disease_network_similarity_df.index = disease_network_similarity_df.index.astype(str)\n",
    "    disease_network_similarity_df.columns = disease_network_similarity_df.columns.astype(str)\n",
    "    \n",
    "    disease_network_similarity_df.to_csv(predict_folder+\"disease_network_similarity_df.csv\")\n",
    "    \n",
    "disease_network_custom = pd.read_csv(predict_folder+\"disease_network_similarity_df.csv\", index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### III.6 Disease-disease similarity: Gene Onthology disease-disease similarities\n",
    "\n",
    "This yields a dataframe `disease_go_similarity_df` of size #diseases x #diseases, where identifiers are MedGen Concept ID identifiers, which contains the similarity between Gene Onthology annotations.\n",
    "\n",
    "Similar to Section [**II.5 Drug-drug similarity: Gene Onthology drug-drug similarities**](http://localhost:8888/notebooks/PREDICT_dataset.ipynb#II.5-Drug-drug-similarity:-Gene-Onthology-drug-drug-similarities)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>C1851649</th>\n",
       "      <th>C0042133</th>\n",
       "      <th>C5193005</th>\n",
       "      <th>C2676676</th>\n",
       "      <th>C1704272</th>\n",
       "      <th>C4722327</th>\n",
       "      <th>C1858361</th>\n",
       "      <th>C2676676.1</th>\n",
       "      <th>C4310232</th>\n",
       "      <th>C0029456</th>\n",
       "      <th>...</th>\n",
       "      <th>C0242770</th>\n",
       "      <th>C1880129</th>\n",
       "      <th>C0022661</th>\n",
       "      <th>C0236792</th>\n",
       "      <th>C1135191</th>\n",
       "      <th>C0149516</th>\n",
       "      <th>C1835407</th>\n",
       "      <th>C0016667</th>\n",
       "      <th>C0039445</th>\n",
       "      <th>C5203670</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C1851649</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C0042133</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.417667</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.188667</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.142500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.099500</td>\n",
       "      <td>0.139000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.141179</td>\n",
       "      <td>0.136333</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.157333</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C5193005</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C2676676</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.188667</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.437333</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.142500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.124167</td>\n",
       "      <td>0.130567</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.148026</td>\n",
       "      <td>0.133333</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.191889</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1704272</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C0149516</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1835407</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C0016667</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C0039445</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.157333</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.191889</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.133722</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.097056</td>\n",
       "      <td>0.146489</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.142205</td>\n",
       "      <td>0.124000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.264185</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C5203670</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1599 rows × 1599 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          C1851649  C0042133  C5193005  C2676676  C1704272  C4722327   \n",
       "C1851649       NaN  0.000000       NaN  0.000000       NaN       NaN  \\\n",
       "C0042133       NaN  0.417667       NaN  0.188667       NaN       NaN   \n",
       "C5193005       NaN  0.000000       NaN  0.000000       NaN       NaN   \n",
       "C2676676       NaN  0.188667       NaN  0.437333       NaN       NaN   \n",
       "C1704272       NaN  0.000000       NaN  0.000000       NaN       NaN   \n",
       "...            ...       ...       ...       ...       ...       ...   \n",
       "C0149516       NaN  0.000000       NaN  0.000000       NaN       NaN   \n",
       "C1835407       NaN  0.000000       NaN  0.000000       NaN       NaN   \n",
       "C0016667       NaN  0.000000       NaN  0.000000       NaN       NaN   \n",
       "C0039445       NaN  0.157333       NaN  0.191889       NaN       NaN   \n",
       "C5203670       NaN  0.000000       NaN  0.000000       NaN       NaN   \n",
       "\n",
       "          C1858361  C2676676.1  C4310232  C0029456  ...  C0242770  C1880129   \n",
       "C1851649  0.000000         NaN  0.000000  0.000000  ...       NaN       NaN  \\\n",
       "C0042133  0.142500         NaN  0.099500  0.139000  ...       NaN       NaN   \n",
       "C5193005  0.000000         NaN  0.000000  0.000000  ...       NaN       NaN   \n",
       "C2676676  0.142500         NaN  0.124167  0.130567  ...       NaN       NaN   \n",
       "C1704272  0.000000         NaN  0.000000  0.000000  ...       NaN       NaN   \n",
       "...            ...         ...       ...       ...  ...       ...       ...   \n",
       "C0149516  0.000000         NaN  0.000000  0.000000  ...       NaN       NaN   \n",
       "C1835407  0.000000         NaN  0.000000  0.000000  ...       NaN       NaN   \n",
       "C0016667  0.000000         NaN  0.000000  0.000000  ...       NaN       NaN   \n",
       "C0039445  0.133722         NaN  0.097056  0.146489  ...       NaN       NaN   \n",
       "C5203670  0.000000         NaN  0.000000  0.000000  ...       NaN       NaN   \n",
       "\n",
       "          C0022661  C0236792  C1135191  C0149516  C1835407  C0016667   \n",
       "C1851649  0.000000  0.000000       NaN       NaN       NaN       NaN  \\\n",
       "C0042133  0.141179  0.136333       NaN       NaN       NaN       NaN   \n",
       "C5193005  0.000000  0.000000       NaN       NaN       NaN       NaN   \n",
       "C2676676  0.148026  0.133333       NaN       NaN       NaN       NaN   \n",
       "C1704272  0.000000  0.000000       NaN       NaN       NaN       NaN   \n",
       "...            ...       ...       ...       ...       ...       ...   \n",
       "C0149516  0.000000  0.000000       NaN       NaN       NaN       NaN   \n",
       "C1835407  0.000000  0.000000       NaN       NaN       NaN       NaN   \n",
       "C0016667  0.000000  0.000000       NaN       NaN       NaN       NaN   \n",
       "C0039445  0.142205  0.124000       NaN       NaN       NaN       NaN   \n",
       "C5203670  0.000000  0.000000       NaN       NaN       NaN       NaN   \n",
       "\n",
       "          C0039445  C5203670  \n",
       "C1851649  0.000000       NaN  \n",
       "C0042133  0.157333       NaN  \n",
       "C5193005  0.000000       NaN  \n",
       "C2676676  0.191889       NaN  \n",
       "C1704272  0.000000       NaN  \n",
       "...            ...       ...  \n",
       "C0149516  0.000000       NaN  \n",
       "C1835407  0.000000       NaN  \n",
       "C0016667  0.000000       NaN  \n",
       "C0039445  0.264185       NaN  \n",
       "C5203670  0.000000       NaN  \n",
       "\n",
       "[1599 rows x 1599 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sb.Popen([\"mkdir\", \"-p\", predict_folder+\"DisGeNet/\"])\n",
    "if (not os.path.exists(predict_folder+\"DisGeNet/disease_db_DisGeNet.csv\")):\n",
    "    \n",
    "    from NORDic.UTILS.DISGENET_utils import get_user_key_DISGENET, get_genes_proteins_from_DISGENET\n",
    "    A = pd.read_csv(paths_global.data_folder+\"FEATURELESS_v1.0.0/all_ratings.csv\", index_col=0)\n",
    "    disease_list = list(A.columns)\n",
    "    \n",
    "    user_key = get_user_key_DISGENET(paths_global.disgenet_file)\n",
    "    disease_db = get_genes_proteins_from_DISGENET(disease_list, limit=3000, source=\"CURATED\", min_score=0.35, \n",
    "                                min_ei=1., min_dsi=0.25, min_dpi=0, chunksize=100, user_key=user_key, quiet=False)\n",
    "    \n",
    "    disease_db.to_csv(predict_folder+\"DisGeNet/disease_db_DisGeNet.csv\")\n",
    "    \n",
    "disease_db = pd.read_csv(predict_folder+\"DisGeNet/disease_db_DisGeNet.csv\", index_col=0)\n",
    "\n",
    "if (not os.path.exists(predict_folder+\"disease_go_similarity_df.csv\")):\n",
    "    disease_go_sim_fname = \"disease_go_similarity\"\n",
    "    if (not os.path.exists(predict_folder+\"DISEASE_GO_SIMILARITY/\"+disease_go_sim_fname+\".csv\")):\n",
    "        sb.Popen((\"mkdir -p \"+predict_folder+\"DISEASE_GO_SIMILARITY/\").split(\" \"))\n",
    "        proteins = list(set([y for x in list(disease_db[\"Protein\"]) for y in x.split(\"; \")]))\n",
    "        with open(\"proteins.txt\", \"w+\") as f:\n",
    "            f.write(\",\".join(proteins))\n",
    "            process = sb.Popen([\"Rscript\", \"../src/Rscript_GOsim.R\"])\n",
    "            process.wait()\n",
    "            sb.Popen([\"mv\", \"disease_go_similarity.csv\", \"disease_go_similarity_MF.csv\", \n",
    "                      \"disease_go_similarity_CC.csv\", \"disease_go_similarity_BP.csv\", \n",
    "                      predict_folder+\"DISEASE_GO_SIMILARITY/\"])\n",
    "    ## annotated by protein\n",
    "    disease_go_similarity = pd.read_csv(predict_folder+\"DISEASE_GO_SIMILARITY/\"+disease_go_sim_fname+\".csv\", index_col=0)\n",
    "\n",
    "    disease_go_similarity_mat = disease_go_similarity.fillna(0.)\n",
    "\n",
    "    diseases = pd.read_csv(paths_global.data_folder+\"FEATURELESS_v1.0.0/all_ratings.csv\", index_col=0).columns\n",
    "    disease_go_similarity_1 = pd.DataFrame(0., index=diseases, columns=disease_go_similarity_mat.columns)\n",
    "    for disease in diseases:\n",
    "        if (disease not in disease_db.index):\n",
    "            continue\n",
    "        protein_disease = disease_db.loc[disease][\"Protein\"].split(\"; \")\n",
    "        targeted_proteins = [x for x in protein_disease if (x in disease_go_similarity_mat.index)]\n",
    "        disease_go_similarity_1.loc[disease] = list(disease_go_similarity_mat.loc[targeted_proteins].mean(axis=0))\n",
    "    disease_go_similarity_1\n",
    "\n",
    "    disease_go_similarity_df = []\n",
    "    for disease in diseases:\n",
    "        if (disease not in disease_db.index):\n",
    "            disease_go_similarity_df.append(pd.DataFrame([], index=diseases, columns=[disease]))\n",
    "            continue\n",
    "        proteins_disease = disease_db.loc[disease][\"Protein\"].split(\"; \")\n",
    "        targeted_proteins = [x for x in proteins_disease if (x in disease_go_similarity_1.columns)]\n",
    "        disease_go_similarity_df.append(pd.DataFrame(disease_go_similarity_1[targeted_proteins].mean(axis=1), columns=[disease]))\n",
    "    disease_go_similarity_df = disease_go_similarity_df[0].join(disease_go_similarity_df[1:], how=\"outer\")\n",
    "    disease_go_similarity_df.index = disease_go_similarity_df.index.astype(str)\n",
    "    disease_go_similarity_df.columns = disease_go_similarity_df.columns.astype(str)\n",
    "    \n",
    "    disease_go_similarity_df.to_csv(predict_folder+\"disease_go_similarity_df.csv\")\n",
    "    \n",
    "disease_go_similarity_df = pd.read_csv(predict_folder+\"disease_go_similarity_df.csv\", index_col=0)\n",
    "disease_go_similarity_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## IV. Build final  drug-drug, disease-disease and drug-disease matrices\n",
    "\n",
    "Disease ids: Concept IDs, drug ids: DrugBank or PubChem CIDs. Build matrices A (ratings) drugs $\\times$ diseases, P (disease features) disease features $\\times$ diseases, S (drug features) drug features $\\times$ drugs, with NaN for missing values.\n",
    "\n",
    "### IV.1. Merge matrices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sparsity = 0.3823575264626351%\n",
      "6030 drug features 2361 disease features\n",
      "Ndrugs=1395\tNdiseases=1501\n",
      "8240 positive\t295 negative\t2085360 unknown matchings\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>8525</th>\n",
       "      <th>8526</th>\n",
       "      <th>8527</th>\n",
       "      <th>8528</th>\n",
       "      <th>8529</th>\n",
       "      <th>8530</th>\n",
       "      <th>8531</th>\n",
       "      <th>8532</th>\n",
       "      <th>8533</th>\n",
       "      <th>8534</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ind_id</th>\n",
       "      <td>C0003467</td>\n",
       "      <td>C1269683.1</td>\n",
       "      <td>C0040558</td>\n",
       "      <td>C1332061</td>\n",
       "      <td>C0008513</td>\n",
       "      <td>C0040560</td>\n",
       "      <td>C0009324</td>\n",
       "      <td>C0037299</td>\n",
       "      <td>C0032302</td>\n",
       "      <td>C0151175</td>\n",
       "      <td>...</td>\n",
       "      <td>C1853965</td>\n",
       "      <td>C0009443</td>\n",
       "      <td>C0027424</td>\n",
       "      <td>C0006266</td>\n",
       "      <td>C1260880</td>\n",
       "      <td>C1859648</td>\n",
       "      <td>C0035455</td>\n",
       "      <td>C0037195</td>\n",
       "      <td>C1869116</td>\n",
       "      <td>C0265962</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>drug_id</th>\n",
       "      <td>3299</td>\n",
       "      <td>3299</td>\n",
       "      <td>5215</td>\n",
       "      <td>5215</td>\n",
       "      <td>5215</td>\n",
       "      <td>5215</td>\n",
       "      <td>5215</td>\n",
       "      <td>5215</td>\n",
       "      <td>54671203</td>\n",
       "      <td>54671203</td>\n",
       "      <td>...</td>\n",
       "      <td>9294</td>\n",
       "      <td>9294</td>\n",
       "      <td>9294</td>\n",
       "      <td>9294</td>\n",
       "      <td>9294</td>\n",
       "      <td>9294</td>\n",
       "      <td>9294</td>\n",
       "      <td>9294</td>\n",
       "      <td>9294</td>\n",
       "      <td>9294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rating</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 8535 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             0           1         2         3         4         5      \n",
       "ind_id   C0003467  C1269683.1  C0040558  C1332061  C0008513  C0040560  \\\n",
       "drug_id      3299        3299      5215      5215      5215      5215   \n",
       "rating        1.0         1.0       1.0       1.0       1.0       1.0   \n",
       "\n",
       "             6         7         8         9     ...      8525      8526   \n",
       "ind_id   C0009324  C0037299  C0032302  C0151175  ...  C1853965  C0009443  \\\n",
       "drug_id      5215      5215  54671203  54671203  ...      9294      9294   \n",
       "rating        1.0       1.0       1.0       1.0  ...       1.0       1.0   \n",
       "\n",
       "             8527      8528      8529      8530      8531      8532      8533   \n",
       "ind_id   C0027424  C0006266  C1260880  C1859648  C0035455  C0037195  C1869116  \\\n",
       "drug_id      9294      9294      9294      9294      9294      9294      9294   \n",
       "rating        1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "\n",
       "             8534  \n",
       "ind_id   C0265962  \n",
       "drug_id      9294  \n",
       "rating        1.0  \n",
       "\n",
       "[3 rows x 8535 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path_dataset = predict_folder\n",
    "\n",
    "## Drug-disease associations\n",
    "custom = pd.read_csv(paths_global.data_folder+\"FEATURELESS_v1.0.0/all_ratings.csv\", index_col=0)\n",
    "\n",
    "## Drug-drug similarities\n",
    "ftypes = [\"chemical\", \"se\", \"sequence\", \"network\", \"go\", \"signature\"]\n",
    "## Disease-disease similaries\n",
    "ftypes += [\"disease_\"+x for x in [\"phenotype\", \"semantic\", \"sequence\", \"network\", \"go\", \"signature\"]]\n",
    "\n",
    "fnames = [path_dataset+x+\"_similarity_df.csv\" for x in ftypes]\n",
    "fdict = {ftype: fnames[ift] for ift, ftype in enumerate(ftypes)}\n",
    "\n",
    "for ftype in fdict:\n",
    "    df = pd.read_csv(fdict[ftype], index_col=0, header=0)\n",
    "    df.index = [ftype+\"-\"+str(x) for x in df.index]\n",
    "    globals()[ftype+\"_custom\"] = df\n",
    "    \n",
    "## Concatenate all features\n",
    "S = pd.concat([eval(ftype+\"_custom\") for ftype in ftypes if (\"disease\" not in ftype)], axis=0, join=\"outer\")\n",
    "P = pd.concat([eval(ftype+\"_custom\") for ftype in ftypes if (\"disease\" in ftype)], axis=0, join=\"outer\")\n",
    "\n",
    "## Restrict to diseases in custom (association matrix)\n",
    "P = P.drop_duplicates(keep=\"first\")[list(set([p for p in custom.columns if (p in P.columns)]))]\n",
    "P_PREDICT = P.loc[~P.index.duplicated()]\n",
    "\n",
    "## Restrict to drugs in custom (association matrix)\n",
    "S = S.drop_duplicates(keep=\"first\")[list(set([x for x in list(custom.index) if (x in S.columns)]))]\n",
    "S_PREDICT = S.loc[~S.index.duplicated()]\n",
    "\n",
    "## Restrict association matrix\n",
    "A = custom[P.columns].loc[S.columns]\n",
    "A_PREDICT = A.loc[~A.index.duplicated()]\n",
    "A_PREDICT.index = A_PREDICT.index.astype(str)\n",
    "\n",
    "S_PREDICT.to_csv(predict_folder+\"items.csv\")\n",
    "P_PREDICT.to_csv(predict_folder+\"users.csv\")\n",
    "A_PREDICT.to_csv(predict_folder+\"ratings_mat.csv\")\n",
    "\n",
    "ratings_A = utils.matrix2ratings(A_PREDICT, \"ind_id\", \"drug_id\", \"rating\")\n",
    "print(\"Sparsity = \"+str(utils.compute_sparsity(A_PREDICT))+\"%\")\n",
    "print(\"%d drug features %d disease features\" % (S_PREDICT.shape[0], P_PREDICT.shape[0]))\n",
    "utils.print_dataset(ratings_A, \"ind_id\", \"drug_id\", \"rating\")\n",
    "ratings_A.T"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### IV.2. Generating matrices which can be shared "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "A_PREDICT = pd.read_csv(predict_folder+\"ratings_mat.csv\")\n",
    "drugs, diseases = list(A_PREDICT.index), list(A_PREDICT.columns)\n",
    "\n",
    "restrict = lambda df, ls : df[[s for s in ls if (s in df.columns)]]\n",
    "\n",
    "for matrix_name in [\"se\",\"signature\",\"disease_phenotype\",\"disease_semantic\"]:\n",
    "    df = restrict(eval(matrix_name+\"_custom\"), drugs if (\"disease\" not in matrix_name) else diseases)\n",
    "    df.to_csv(matrix_name+\"_PREDICT_matrix.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
