{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Creating the \"TRANSCRIPT\" dataset based on transcriptomics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Version 1.0.0 (December 28th 2022). Please run the notebook \"FEATURELESS-v1.0.0_dataset.ipynb\" beforehand."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import subprocess as sb\n",
    "import os\n",
    "\n",
    "from time import sleep\n",
    "from tqdm import tqdm\n",
    "import requests\n",
    "import pickle\n",
    "import json\n",
    "\n",
    "import sys\n",
    "sys.path.insert(0, \"../../src/\")\n",
    "\n",
    "import utils\n",
    "import paths_global\n",
    "import data_processing\n",
    "\n",
    "from joblib import Parallel, delayed\n",
    "from multiprocessing import cpu_count\n",
    "njobs=min(max(1,cpu_count()-2),3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Local paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "root_folder=\"/media/kali/1b80f30d-2803-4260-a792-9ae206084252/Code/M30/data/\"\n",
      "data_folder=\"/media/kali/1b80f30d-2803-4260-a792-9ae206084252/RECeSS/cfdr/data/\"\n"
     ]
    }
   ],
   "source": [
    "## Where database files are stored\n",
    "print('root_folder=\"%s\"' % paths_global.root_folder)\n",
    "## Where intermediary files are stored\n",
    "print('data_folder=\"%s\"' % paths_global.data_folder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "transcript_folder=\"/media/kali/1b80f30d-2803-4260-a792-9ae206084252/RECeSS/cfdr/data/TRANSCRIPT_v1.0.0/\"\n"
     ]
    }
   ],
   "source": [
    "transcript_folder = paths_global.data_folder+\"TRANSCRIPT_v1.0.0/\"\n",
    "sb.Popen([\"mkdir\", \"-p\", transcript_folder])\n",
    "## Where TRANSCRIPT dataset files are stored\n",
    "print('transcript_folder=\"%s\"' % transcript_folder)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Drug and disease identifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert os.path.exists(paths_global.data_folder+\"drugbankid2drugname.pck\")\n",
    "with open(paths_global.data_folder+\"drugbankid2drugname.pck\", \"rb\") as f:\n",
    "    di_drugbankid2drugname = pickle.load(f)\n",
    "    \n",
    "assert os.path.exists(paths_global.data_folder+\"omimid2diseasename.pck\")\n",
    "with open(paths_global.data_folder+\"omimid2diseasename.pck\", \"rb\") as f:\n",
    "    di_omimid2diseasename = pickle.load(f)\n",
    "    \n",
    "cids_file = paths_global.data_folder+\"medgenid2diseasename.pck\"\n",
    "if (not os.path.exists(cids_file)):\n",
    "    di_medgenid2diseasename = {}\n",
    "else:\n",
    "    with open(cids_file, \"rb\") as f:\n",
    "        di_medgenid2diseasename = pickle.load(f)\n",
    "        \n",
    "pubchem_file = paths_global.data_folder+\"pubchemid2drugname.pck\"\n",
    "if (not os.path.exists(pubchem_file)):\n",
    "    di_pubchemid2drugname = {}\n",
    "else:\n",
    "    with open(pubchem_file, \"rb\") as f:\n",
    "        di_pubchemid2drugname = pickle.load(f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## I. Matrix A : $N_S \\times N_D$ of drug-disease associations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>C1851649</th>\n",
       "      <th>C0042133</th>\n",
       "      <th>C5193005</th>\n",
       "      <th>C2676676</th>\n",
       "      <th>C1704272</th>\n",
       "      <th>C4722327</th>\n",
       "      <th>C1858361</th>\n",
       "      <th>C2676676.1</th>\n",
       "      <th>C4310232</th>\n",
       "      <th>C0029456</th>\n",
       "      <th>...</th>\n",
       "      <th>C0242770</th>\n",
       "      <th>C1880129</th>\n",
       "      <th>C0022661</th>\n",
       "      <th>C0236792</th>\n",
       "      <th>C1135191</th>\n",
       "      <th>C0149516</th>\n",
       "      <th>C1835407</th>\n",
       "      <th>C0016667</th>\n",
       "      <th>C0039445</th>\n",
       "      <th>C5203670</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>657181</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DB00010</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5311128</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DB00017</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5311065</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5917</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>442872</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>442021</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4843</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DB12466</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1601 rows × 1599 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         C1851649  C0042133  C5193005  C2676676  C1704272  C4722327  C1858361   \n",
       "657181        1.0       1.0       1.0       1.0       1.0       1.0       1.0  \\\n",
       "DB00010       0.0       0.0       0.0       0.0       0.0       0.0       1.0   \n",
       "5311128       1.0       0.0       0.0       1.0       0.0       1.0       0.0   \n",
       "DB00017       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "5311065       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "...           ...       ...       ...       ...       ...       ...       ...   \n",
       "5917          0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "442872        0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "442021        0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "4843          0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "DB12466       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "\n",
       "         C2676676.1  C4310232  C0029456  ...  C0242770  C1880129  C0022661   \n",
       "657181          0.0       0.0       0.0  ...       0.0       0.0       0.0  \\\n",
       "DB00010         0.0       0.0       0.0  ...       0.0       0.0       0.0   \n",
       "5311128         1.0       0.0       0.0  ...       0.0       0.0       0.0   \n",
       "DB00017         0.0       1.0       1.0  ...       0.0       0.0       0.0   \n",
       "5311065         0.0       0.0       0.0  ...       0.0       0.0       0.0   \n",
       "...             ...       ...       ...  ...       ...       ...       ...   \n",
       "5917            0.0       0.0       0.0  ...       0.0       0.0       0.0   \n",
       "442872          0.0       0.0       0.0  ...       0.0       0.0       0.0   \n",
       "442021          0.0       0.0       0.0  ...       0.0       0.0       0.0   \n",
       "4843            0.0       0.0       0.0  ...       0.0       0.0       0.0   \n",
       "DB12466         0.0       0.0       0.0  ...       0.0       0.0       0.0   \n",
       "\n",
       "         C0236792  C1135191  C0149516  C1835407  C0016667  C0039445  C5203670  \n",
       "657181        0.0       0.0       0.0       0.0       0.0       0.0       0.0  \n",
       "DB00010       0.0       0.0       0.0       0.0       0.0       0.0       0.0  \n",
       "5311128       0.0       0.0       0.0       0.0       0.0       0.0       0.0  \n",
       "DB00017       0.0       0.0       0.0       0.0       0.0       0.0       0.0  \n",
       "5311065       0.0       0.0       0.0       0.0       0.0       0.0       0.0  \n",
       "...           ...       ...       ...       ...       ...       ...       ...  \n",
       "5917          0.0       0.0       0.0       0.0       0.0       0.0       0.0  \n",
       "442872        0.0       0.0       0.0       0.0       0.0       0.0       0.0  \n",
       "442021        0.0       0.0       0.0       0.0       0.0       0.0       0.0  \n",
       "4843          0.0       0.0       0.0       0.0       0.0       0.0       0.0  \n",
       "DB12466       0.0       0.0       0.0       0.0       0.0       0.0      -1.0  \n",
       "\n",
       "[1601 rows x 1599 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "A = pd.read_csv(paths_global.data_folder+\"FEATURELESS_v1.0.0/all_ratings.csv\", index_col=0)\n",
    "A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sparsity = 0.35070326199346175%\n",
      "Ndrugs=1599\tNdiseases=1599\n",
      "8658 positive\t320 negative\t2547823 unknown matchings\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>8968</th>\n",
       "      <th>8969</th>\n",
       "      <th>8970</th>\n",
       "      <th>8971</th>\n",
       "      <th>8972</th>\n",
       "      <th>8973</th>\n",
       "      <th>8974</th>\n",
       "      <th>8975</th>\n",
       "      <th>8976</th>\n",
       "      <th>8977</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ind_id</th>\n",
       "      <td>C1851649</td>\n",
       "      <td>C0042133</td>\n",
       "      <td>C5193005</td>\n",
       "      <td>C2676676</td>\n",
       "      <td>C1704272</td>\n",
       "      <td>C4722327</td>\n",
       "      <td>C1858361</td>\n",
       "      <td>C0034013</td>\n",
       "      <td>C0014175</td>\n",
       "      <td>C1858361</td>\n",
       "      <td>...</td>\n",
       "      <td>C0014544</td>\n",
       "      <td>C0014544</td>\n",
       "      <td>C0014544</td>\n",
       "      <td>C0014544</td>\n",
       "      <td>C0014544</td>\n",
       "      <td>C0014544</td>\n",
       "      <td>C0014544</td>\n",
       "      <td>C0014544</td>\n",
       "      <td>C0014544</td>\n",
       "      <td>C5203670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>drug_id</th>\n",
       "      <td>657181</td>\n",
       "      <td>657181</td>\n",
       "      <td>657181</td>\n",
       "      <td>657181</td>\n",
       "      <td>657181</td>\n",
       "      <td>657181</td>\n",
       "      <td>657181</td>\n",
       "      <td>657181</td>\n",
       "      <td>657181</td>\n",
       "      <td>DB00010</td>\n",
       "      <td>...</td>\n",
       "      <td>2310</td>\n",
       "      <td>492405</td>\n",
       "      <td>4375</td>\n",
       "      <td>3352</td>\n",
       "      <td>3373</td>\n",
       "      <td>5917</td>\n",
       "      <td>442872</td>\n",
       "      <td>442021</td>\n",
       "      <td>4843</td>\n",
       "      <td>DB12466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rating</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 8978 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             0         1         2         3         4         5         6      \n",
       "ind_id   C1851649  C0042133  C5193005  C2676676  C1704272  C4722327  C1858361  \\\n",
       "drug_id    657181    657181    657181    657181    657181    657181    657181   \n",
       "rating        1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "\n",
       "             7         8         9     ...      8968      8969      8970   \n",
       "ind_id   C0034013  C0014175  C1858361  ...  C0014544  C0014544  C0014544  \\\n",
       "drug_id    657181    657181   DB00010  ...      2310    492405      4375   \n",
       "rating        1.0       1.0       1.0  ...      -1.0      -1.0      -1.0   \n",
       "\n",
       "             8971      8972      8973      8974      8975      8976      8977  \n",
       "ind_id   C0014544  C0014544  C0014544  C0014544  C0014544  C0014544  C5203670  \n",
       "drug_id      3352      3373      5917    442872    442021      4843   DB12466  \n",
       "rating       -1.0      -1.0      -1.0      -1.0      -1.0       1.0      -1.0  \n",
       "\n",
       "[3 rows x 8978 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ratings_A = utils.matrix2ratings(A, \"ind_id\", \"drug_id\", \"rating\")\n",
    "\n",
    "print(\"Sparsity = \"+str(utils.compute_sparsity(A))+\"%\")\n",
    "utils.print_dataset(ratings_A, \"ind_id\", \"drug_id\", \"rating\")\n",
    "ratings_A.T"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## II. Build matrix P : $N_F \\times N_D$ of disease features\n",
    "\n",
    "Here, features are disease \"differential\" phenotypes, that is vectors of genewise expression changes due to disease (by comparing differential expression between patient and healthy groups). We use [CREEDS](https://maayanlab.cloud/CREEDS/) (free access, without registration), which reports manually curated disease phenotypes and crowd-sourced phenotypes (see [paper](https://www.nature.com/articles/ncomms12846)).\n",
    "\n",
    "### II.a. Conversion of genes into their corresponding human orthologs\n",
    "\n",
    "Sometimes the phenotypes are computed on animal models, hence the need to convert (one-to-one) orthologs from that species to humans. Here, we only consider mice models outside of humans.\n",
    "- go to https://www.ensembl.org/biomart/martview/\n",
    "- select in \"Dataset\": \"Mouse genes (GRCm39)\"\n",
    "- \"Filters\": None\n",
    "- \"Attributes\": \n",
    "\tGene > \"Gene Stable ID\", \"Gene name\"\n",
    "\tHomologues > Human > \"Human gene stable ID\", \"Human homology type\", \"Human gene name\"\n",
    "- select \"Results\" and \"Go\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "orthologs_folder=\"/media/kali/1b80f30d-2803-4260-a792-9ae206084252/Code/M30/data/orthologs/\"\n"
     ]
    }
   ],
   "source": [
    "## Where the orthologs files are stored\n",
    "print('orthologs_folder=\"%s\"' % paths_global.orthologs_folder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "mouse_human = pd.read_csv(paths_global.orthologs_folder+\"mouse_human_matchings.tsv\", sep='\\t', header=0, index_col=None)\n",
    "mouse_human = mouse_human.loc[mouse_human[\"Human homology type\"]==\"ortholog_one2one\"][[\"Gene name\",\"Human gene name\"]]\n",
    "mouse_human.index = mouse_human[\"Gene name\"]\n",
    "mouse_human = mouse_human.to_dict()[\"Human gene name\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### II.b. Retrieval of differential phenotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "creeds_folder=\"/media/kali/1b80f30d-2803-4260-a792-9ae206084252/Code/M30/data/CREEDS/\"\n"
     ]
    }
   ],
   "source": [
    "## Where the CREEDS files are stored\n",
    "print('creeds_folder=\"%s\"' % paths_global.creeds_folder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>C2936783</th>\n",
       "      <th>C2239176</th>\n",
       "      <th>C3553462</th>\n",
       "      <th>C0035235</th>\n",
       "      <th>C0032285</th>\n",
       "      <th>C0010346</th>\n",
       "      <th>C0009324</th>\n",
       "      <th>C0029408</th>\n",
       "      <th>C0001973</th>\n",
       "      <th>C3495559</th>\n",
       "      <th>...</th>\n",
       "      <th>C1260899</th>\n",
       "      <th>C1527336</th>\n",
       "      <th>C0275804</th>\n",
       "      <th>C0017168</th>\n",
       "      <th>C0003615</th>\n",
       "      <th>C0003872</th>\n",
       "      <th>C0014544</th>\n",
       "      <th>C0040028</th>\n",
       "      <th>C0040034</th>\n",
       "      <th>C0018802</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>USMG5</th>\n",
       "      <td>-0.158146</td>\n",
       "      <td>-0.158146</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.038415</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.022347</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SLC26A2</th>\n",
       "      <td>-0.131811</td>\n",
       "      <td>-0.131811</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.075141</td>\n",
       "      <td>-0.112273</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.014750</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.036328</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GUCA2A</th>\n",
       "      <td>-0.130771</td>\n",
       "      <td>-0.130771</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.101133</td>\n",
       "      <td>-0.167974</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.026553</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AQP8</th>\n",
       "      <td>-0.126526</td>\n",
       "      <td>-0.126526</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.035313</td>\n",
       "      <td>-0.085438</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.033508</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CEACAM6</th>\n",
       "      <td>-0.125712</td>\n",
       "      <td>-0.125712</td>\n",
       "      <td>0.050276</td>\n",
       "      <td>0.050276</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.118054</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.018521</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.017911</td>\n",
       "      <td>-0.014329</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPRYD7</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.010207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ALKBH7</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.010192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SLC41A3</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.009935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TRIM54</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.017070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRKAB2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.010438</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13763 rows × 144 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         C2936783  C2239176  C3553462  C0035235  C0032285  C0010346  C0009324   \n",
       "USMG5   -0.158146 -0.158146  0.000000  0.000000       0.0  0.038415  0.000000  \\\n",
       "SLC26A2 -0.131811 -0.131811  0.000000  0.000000       0.0 -0.075141 -0.112273   \n",
       "GUCA2A  -0.130771 -0.130771  0.000000  0.000000       0.0 -0.101133 -0.167974   \n",
       "AQP8    -0.126526 -0.126526  0.000000  0.000000       0.0 -0.035313 -0.085438   \n",
       "CEACAM6 -0.125712 -0.125712  0.050276  0.050276       0.0  0.118054  0.000000   \n",
       "...           ...       ...       ...       ...       ...       ...       ...   \n",
       "SPRYD7   0.000000  0.000000  0.000000  0.000000       0.0  0.000000  0.000000   \n",
       "ALKBH7   0.000000  0.000000  0.000000  0.000000       0.0  0.000000  0.000000   \n",
       "SLC41A3  0.000000  0.000000  0.000000  0.000000       0.0  0.000000  0.000000   \n",
       "TRIM54   0.000000  0.000000  0.000000  0.000000       0.0  0.000000  0.000000   \n",
       "PRKAB2   0.000000  0.000000  0.000000  0.000000       0.0  0.000000  0.000000   \n",
       "\n",
       "         C0029408  C0001973  C3495559  ...  C1260899  C1527336  C0275804   \n",
       "USMG5         0.0       0.0       0.0  ...       0.0  0.000000       0.0  \\\n",
       "SLC26A2       0.0       0.0       0.0  ...       0.0 -0.014750       0.0   \n",
       "GUCA2A        0.0       0.0       0.0  ...       0.0  0.000000       0.0   \n",
       "AQP8          0.0       0.0       0.0  ...       0.0  0.000000       0.0   \n",
       "CEACAM6       0.0       0.0       0.0  ...       0.0  0.018521       0.0   \n",
       "...           ...       ...       ...  ...       ...       ...       ...   \n",
       "SPRYD7        0.0       0.0       0.0  ...       0.0  0.000000       0.0   \n",
       "ALKBH7        0.0       0.0       0.0  ...       0.0  0.000000       0.0   \n",
       "SLC41A3       0.0       0.0       0.0  ...       0.0  0.000000       0.0   \n",
       "TRIM54        0.0       0.0       0.0  ...       0.0  0.000000       0.0   \n",
       "PRKAB2        0.0       0.0       0.0  ...       0.0  0.000000       0.0   \n",
       "\n",
       "         C0017168  C0003615  C0003872  C0014544  C0040028  C0040034  C0018802  \n",
       "USMG5    0.000000  0.000000       0.0  0.022347       0.0       0.0  0.000000  \n",
       "SLC26A2  0.000000 -0.036328       0.0  0.000000       0.0       0.0  0.000000  \n",
       "GUCA2A   0.000000 -0.026553       0.0  0.000000       0.0       0.0  0.000000  \n",
       "AQP8     0.000000 -0.033508       0.0  0.000000       0.0       0.0  0.000000  \n",
       "CEACAM6 -0.017911 -0.014329       0.0  0.000000       0.0       0.0  0.000000  \n",
       "...           ...       ...       ...       ...       ...       ...       ...  \n",
       "SPRYD7   0.000000  0.000000       0.0  0.000000       0.0       0.0 -0.010207  \n",
       "ALKBH7   0.000000  0.000000       0.0  0.000000       0.0       0.0 -0.010192  \n",
       "SLC41A3  0.000000  0.000000       0.0  0.000000       0.0       0.0 -0.009935  \n",
       "TRIM54   0.000000  0.000000       0.0  0.000000       0.0       0.0  0.017070  \n",
       "PRKAB2   0.000000  0.000000       0.0  0.000000       0.0       0.0  0.010438  \n",
       "\n",
       "[13763 rows x 144 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if (not os.path.exists(transcript_folder+\"all_diseases.csv\")):\n",
    "    P_manual = {}\n",
    "    for k in [\"manual\"]:\n",
    "        fname=creeds_folder+\"disease_signatures-\"+type_di[k]+\"1.0.json\"\n",
    "        ii = 1\n",
    "        while (True):\n",
    "            content = iterator_json(fname, ii)\n",
    "            if (str(content)==\"None\"):\n",
    "                break\n",
    "            disease_cid = content[\"umls_cui\"]\n",
    "            organism = content[\"organism\"]\n",
    "            print((disease_cid, organism))\n",
    "            ii += 1\n",
    "            if (organism not in [\"human\", \"mouse\"]):\n",
    "                continue\n",
    "            if (disease_cid in A.columns and disease_cid not in P_manual):\n",
    "                if (organism == \"human\"):\n",
    "                    sig = dict(content['down_genes']+content['up_genes'])\n",
    "                elif (organism == \"mouse\"):\n",
    "                    sig = dict([[mouse_human[g],gi] for g,gi in content['down_genes']+content['up_genes'] if (g in mouse_human)])\n",
    "                P_manual.setdefault(disease_cid, sig)\n",
    "                \n",
    "    P_auto = {}\n",
    "    seen_diseases = []\n",
    "    seen_diseases = seen_diseases[:-1]\n",
    "    for k in [\"automatic\"]:\n",
    "        fname=creeds_folder+\"disease_signatures-\"+type_di[k]+\"1.0.json\"\n",
    "        ii = 1\n",
    "        while (True):\n",
    "            content = iterator_json(fname, ii)\n",
    "            if (str(content)==\"None\"):\n",
    "                print(\"End of file\")\n",
    "                break\n",
    "            disease_cid = []   \n",
    "            ii += 1\n",
    "            organism = content[\"organism\"]\n",
    "            if (not \"disease_name\" in content):\n",
    "                continue\n",
    "            if (not content[\"disease_name\"] in seen_diseases):\n",
    "                print((ii-1,organism, content[\"disease_name\"]))\n",
    "            if (organism not in [\"mouse\", \"human\"] and str(organism)!=str(list(sorted(['human', 'mouse'])))):\n",
    "                continue\n",
    "            if (not content[\"disease_name\"] in seen_diseases):\n",
    "                seen_diseases.append(content[\"disease_name\"])\n",
    "                for disease in content[\"disease_name\"].split(\"|\"):\n",
    "                    unspecified_diseases = [\"Syndrome\", \"incurable diseases\",\"Virus Diseases\"]\n",
    "                    unspecified_diseases += [\"Possessed\",\"Disease\",\"Disease Progression\",\"disease transmission\"]\n",
    "                    if (disease in unspecified_diseases):\n",
    "                        ## unspecified diseases\n",
    "                        continue\n",
    "                    try:\n",
    "                        cid = get_concept_id(disease)\n",
    "                        if (cid in A.columns and cid not in P_auto):\n",
    "                            disease_cid.append(cid)\n",
    "                    except:\n",
    "                        if (disease == \"Seryl-tRNA synthetase, mitochondrial\"):\n",
    "                            disease = \"Seryl-tRNA synthetase\"\n",
    "                        elif (disease == \"Alzheimer's Disease Pathway KEGG\"):\n",
    "                            disease = \"Alzheimer Disease\"\n",
    "                        elif (disease == \"Metastatic Renal Cell Cancer\"):\n",
    "                            disease = \"Renal Cell Cancer\"\n",
    "                        elif (disease == \"1-acylglycerol-3-phosphate O-acyltransferase ABHD5\"):\n",
    "                            disease = \"ABHD5\"\n",
    "                        elif (disease == \"S-adenosyl-L-methionine\"):\n",
    "                            ## unspecified\n",
    "                            continue\n",
    "                        elif (disease == \"metaplastic cell transformation\"):\n",
    "                            ## unspecified\n",
    "                            continue\n",
    "                        elif (disease == \"Lipoma-preferred partner\"):\n",
    "                            disease = \"Lipoma\"\n",
    "                        elif (disease == \"lipophosphoglycan\"):\n",
    "                            continue\n",
    "                        elif (disease == \"Refractory anaemia with excess blasts\"):\n",
    "                            disease = \"Refractory cytopenias\"\n",
    "                        else:\n",
    "                            print(disease)\n",
    "                            raise ValueError\n",
    "                            continue\n",
    "                        cid = get_concept_id(disease)\n",
    "                        if (cid in A.columns and cid not in P_auto):\n",
    "                            disease_cid.append(cid)\n",
    "            if (len(disease_cid) == 0):\n",
    "                continue\n",
    "            for dcid in disease_cid:\n",
    "                if ((organism == \"human\") or (\"human\" in organism)):\n",
    "                    sig = dict(content['down_genes']+content['up_genes'])\n",
    "                elif (organism == \"mouse\"):\n",
    "                    sig = dict([[mouse_human[g],gi] for g,gi in content['down_genes']+content['up_genes'] if (g in mouse_human)])\n",
    "                P_auto.setdefault(dcid, sig)\n",
    "    \n",
    "    ## Merge the two datasets (favouring manually curated signatures over the other ones)\n",
    "    from copy import deepcopy\n",
    "    P = deepcopy(P_auto)\n",
    "    P.update(P_manual)\n",
    "    \n",
    "    pd.DataFrame(P).fillna(0.).to_csv(transcript_folder+\"all_diseases.csv\")\n",
    "    \n",
    "P = pd.read_csv(transcript_folder+\"all_diseases.csv\", index_col=0)\n",
    "P"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'144/1599 (9.01 perc.)'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"%d/%d (%.2f perc.)\" % (len([x for x in A.columns if (x in P.columns)]), A.shape[1], len([x for x in A.columns if (x in P.columns)])*100/A.shape[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'144/144 (100.00 perc.)'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"%d/%d (%.2f perc.)\" % (len([x for x in P.columns if (x in A.columns)]), P.shape[1], len([x for x in P.columns if (x in A.columns)])*100/P.shape[1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## III. Build matrix S : $N_F \\times N_S$ of drug features\n",
    "\n",
    "### III.a. Using [CREEDS](https://maayanlab.cloud/CREEDS/) dataset\n",
    "\n",
    "Here, features are drug signatures, that is vectors of genewise expression changes due to treatment (by comparing differential expression between treated and untreated groups). We use [CREEDS](https://maayanlab.cloud/CREEDS/) (free access, without registration), which reports manually curated and crowd-sourced drug signatures (see [paper](https://www.nature.com/articles/ncomms12846)). Contrary to [LINCS](https://lincsproject.org/LINCS/tools/workflows/find-the-best-place-to-obtain-the-lincs-l1000-data) L1000 data (see [paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5990023/)), those drug signatures are computed on GEO datasets, which might avoid the caveat of using measures from immortalized cells, and to rely on the computational inference for most genes. Similarly, whenever necessary, we convert gene names to their one-to-one orthologs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "orthologs_folder=\"/media/kali/1b80f30d-2803-4260-a792-9ae206084252/Code/M30/data/orthologs/\"\n",
      "creeds_folder=\"/media/kali/1b80f30d-2803-4260-a792-9ae206084252/Code/M30/data/CREEDS/\"\n"
     ]
    }
   ],
   "source": [
    "## Where the orthologs files are stored\n",
    "print('orthologs_folder=\"%s\"' % paths_global.orthologs_folder)\n",
    "## Where the CREEDS files are stored\n",
    "print('creeds_folder=\"%s\"' % paths_global.creeds_folder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>36314</th>\n",
       "      <th>11354606</th>\n",
       "      <th>679</th>\n",
       "      <th>2554</th>\n",
       "      <th>126941</th>\n",
       "      <th>3033</th>\n",
       "      <th>444795</th>\n",
       "      <th>5757</th>\n",
       "      <th>19649</th>\n",
       "      <th>6279</th>\n",
       "      <th>...</th>\n",
       "      <th>149096</th>\n",
       "      <th>10836</th>\n",
       "      <th>5379</th>\n",
       "      <th>446220</th>\n",
       "      <th>57469</th>\n",
       "      <th>3410</th>\n",
       "      <th>5754</th>\n",
       "      <th>6322</th>\n",
       "      <th>31401</th>\n",
       "      <th>5394</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>IGFBP3</th>\n",
       "      <td>-0.305380</td>\n",
       "      <td>-0.099444</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.01587</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.012214</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.015497</td>\n",
       "      <td>-0.048162</td>\n",
       "      <td>-0.229562</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.150907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADIRF</th>\n",
       "      <td>-0.158368</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>THBS2</th>\n",
       "      <td>-0.134381</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.017583</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.054800</td>\n",
       "      <td>-0.018192</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C10orf10</th>\n",
       "      <td>-0.128742</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AQP1</th>\n",
       "      <td>-0.128104</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EHBP1L1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.016950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C19orf60</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.014467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TBX2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.013415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF395</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.027301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TMEM237</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.014068</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10702 rows × 146 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             36314  11354606  679  2554  126941  3033   444795  5757  19649   \n",
       "IGFBP3   -0.305380 -0.099444  0.0   0.0     0.0   0.0  0.01587   0.0    0.0  \\\n",
       "ADIRF    -0.158368  0.000000  0.0   0.0     0.0   0.0  0.00000   0.0    0.0   \n",
       "THBS2    -0.134381  0.000000  0.0   0.0     0.0   0.0  0.00000   0.0    0.0   \n",
       "C10orf10 -0.128742  0.000000  0.0   0.0     0.0   0.0  0.00000   0.0    0.0   \n",
       "AQP1     -0.128104  0.000000  0.0   0.0     0.0   0.0  0.00000   0.0    0.0   \n",
       "...            ...       ...  ...   ...     ...   ...      ...   ...    ...   \n",
       "EHBP1L1   0.000000  0.000000  0.0   0.0     0.0   0.0  0.00000   0.0    0.0   \n",
       "C19orf60  0.000000  0.000000  0.0   0.0     0.0   0.0  0.00000   0.0    0.0   \n",
       "TBX2      0.000000  0.000000  0.0   0.0     0.0   0.0  0.00000   0.0    0.0   \n",
       "ZNF395    0.000000  0.000000  0.0   0.0     0.0   0.0  0.00000   0.0    0.0   \n",
       "TMEM237   0.000000  0.000000  0.0   0.0     0.0   0.0  0.00000   0.0    0.0   \n",
       "\n",
       "          6279  ...  149096  10836      5379    446220     57469      3410   \n",
       "IGFBP3     0.0  ...     0.0    0.0 -0.012214  0.000000 -0.015497 -0.048162  \\\n",
       "ADIRF      0.0  ...     0.0    0.0  0.000000  0.000000  0.000000  0.000000   \n",
       "THBS2      0.0  ...     0.0    0.0  0.000000  0.017583  0.000000  0.054800   \n",
       "C10orf10   0.0  ...     0.0    0.0  0.000000  0.000000  0.000000  0.000000   \n",
       "AQP1       0.0  ...     0.0    0.0  0.000000  0.000000  0.000000  0.000000   \n",
       "...        ...  ...     ...    ...       ...       ...       ...       ...   \n",
       "EHBP1L1    0.0  ...     0.0    0.0  0.000000  0.000000  0.000000  0.000000   \n",
       "C19orf60   0.0  ...     0.0    0.0  0.000000  0.000000  0.000000  0.000000   \n",
       "TBX2       0.0  ...     0.0    0.0  0.000000  0.000000  0.000000  0.000000   \n",
       "ZNF395     0.0  ...     0.0    0.0  0.000000  0.000000  0.000000  0.000000   \n",
       "TMEM237    0.0  ...     0.0    0.0  0.000000  0.000000  0.000000  0.000000   \n",
       "\n",
       "              5754  6322  31401      5394  \n",
       "IGFBP3   -0.229562   0.0    0.0  0.150907  \n",
       "ADIRF     0.000000   0.0    0.0  0.000000  \n",
       "THBS2    -0.018192   0.0    0.0  0.000000  \n",
       "C10orf10  0.000000   0.0    0.0  0.000000  \n",
       "AQP1      0.000000   0.0    0.0  0.000000  \n",
       "...            ...   ...    ...       ...  \n",
       "EHBP1L1   0.000000   0.0    0.0 -0.016950  \n",
       "C19orf60  0.000000   0.0    0.0 -0.014467  \n",
       "TBX2      0.000000   0.0    0.0 -0.013415  \n",
       "ZNF395    0.000000   0.0    0.0  0.027301  \n",
       "TMEM237   0.000000   0.0    0.0  0.014068  \n",
       "\n",
       "[10702 rows x 146 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if (not os.path.exists(transcript_folder+\"all_drugs.csv\")):\n",
    "\n",
    "    A = pd.read_csv(paths_global.data_folder+\"FEATURELESS_v1.0.0/all_ratings.csv\", engine=\"python\", index_col=0)\n",
    "\n",
    "    type_di = {\"manual\": \"v\", \"automatic\": \"p\"}\n",
    "    fs = [\"disease_signatures\", \"single_drug_perturbations\"]\n",
    "    base_url = \"https://maayanlab.cloud/CREEDS/download/\"\n",
    "\n",
    "    sb.call(\"mkdir -p \"+paths_global.creeds_folder, shell=True)\n",
    "    for f in fs:\n",
    "        for kt in type_di:\n",
    "            fname = f+\"-\"+type_di[kt]+\"1.0.json\"\n",
    "            if (not os.path.exists(paths_global.creeds_folder+fname)):\n",
    "                sb.call(\"wget -O \"+paths_global.creeds_folder+fname+\" \"+base_url+fname, shell=True)\n",
    "\n",
    "    def iterator_json(fname, idi):\n",
    "        cmd = \"cat \"+fname+\" | sed 's/\\\\}/\\\\n/g' | head -n\"+str(idi+1)+\" | tail -n1\"\n",
    "        res = sb.check_output(cmd, shell=True).decode(\"utf-8\")[:-1]\n",
    "        if (len(res) == 0):\n",
    "            return None\n",
    "        res = res[1:-1]+\"\\\"}\"\n",
    "        content = json.loads(res)\n",
    "        return content\n",
    "\n",
    "    S_manual = {}\n",
    "    for k in [\"manual\"]:\n",
    "        fname=creeds_folder+\"single_drug_perturbations-\"+type_di[k]+\"1.0.json\"\n",
    "        ii = 1\n",
    "        while (True):\n",
    "            content = iterator_json(fname, ii)\n",
    "            if (str(content)==\"None\"):\n",
    "                print(\"End of file\")\n",
    "                break\n",
    "            drug_cid = content[\"pubchem_cid\"]\n",
    "            organism = content[\"organism\"]\n",
    "            print((ii,drug_cid, organism))\n",
    "            ii += 1\n",
    "            if (organism not in [\"human\", \"mouse\"]):\n",
    "                continue\n",
    "            if (str(drug_cid)==\"None\"):\n",
    "                continue\n",
    "            if (drug_cid in A.index and drug_cid not in S_manual):\n",
    "                if (organism == \"human\"):\n",
    "                    sig = dict(content['down_genes']+content['up_genes'])\n",
    "                elif (organism == \"mouse\"):\n",
    "                    sig = dict([[mouse_human[g],gi] for g,gi in content['down_genes']+content['up_genes'] if (g in mouse_human)])\n",
    "                S_manual.setdefault(drug_cid, sig)\n",
    "\n",
    "    def iterator_json(fname, idi):\n",
    "        cmd = \"cat \"+fname+\" | sed 's/\\\\}/\\\\n/g' | head -n\"+str(idi+1)+\" | tail -n1\"\n",
    "        res = sb.check_output(cmd, shell=True).decode(\"utf-8\")[:-1]\n",
    "        if (len(res) == 0):\n",
    "            return None\n",
    "        res = res[1:-1]+\"\\\"}\"\n",
    "        try:\n",
    "            content = json.loads(res)\n",
    "            return content\n",
    "        except:\n",
    "            ## sanitize...\n",
    "            res_ = res.split(\"drug_name\\\": \\\"\")[0]+\"drug_name\\\": \\\"\"\n",
    "            res_ += \"\".join(\"\".join(\"\".join(\"\".join(\"\".join(\"\".join(res.split(\"drug_name\\\": \\\"\")[-1].split(\"\\\"\")[0].split(\"{\")).split(\"}\")).split(\"[\")).split(\"]\")).split(\"\\'\")).split(\"\\\\\"))\n",
    "            res_ += \"\\\"\"+res.split(\"drug_name\\\": \\\"\")[-1].split(\"\\\"\")[-1]\n",
    "            res = res_\n",
    "            try:\n",
    "                content = json.loads(res)\n",
    "                return content\n",
    "            except:\n",
    "                return 0\n",
    "\n",
    "    S_auto = {}\n",
    "    seen_drugs = []\n",
    "    for k in [\"automatic\"]:\n",
    "        fname=creeds_folder+\"single_drug_perturbations-\"+type_di[k]+\"1.0.json\"\n",
    "        ii = 1\n",
    "        while (True):\n",
    "            content = iterator_json(fname, ii)\n",
    "            if (str(content)==\"None\"):\n",
    "                ## end of file\n",
    "                print(\"End of file\")\n",
    "                break\n",
    "            ii += 1\n",
    "            if (str(content)==\"0\"):\n",
    "                ## ill-formated \n",
    "                continue\n",
    "            if (not \"organism\" in content):\n",
    "                continue\n",
    "            organism = content[\"organism\"]\n",
    "            if (not \"drug_name\" in content):\n",
    "                continue\n",
    "            if (not content[\"drug_name\"] in seen_drugs):\n",
    "                print((ii-1,organism, content[\"drug_name\"]))\n",
    "            drug_cid = []\n",
    "            if (not content[\"drug_name\"] in seen_drugs):\n",
    "                seen_drugs.append(content[\"drug_name\"])\n",
    "                cids = utils.get_pubchem_id(content[\"drug_name\"].split(\"|\"))\n",
    "                drug_cid = [cid for cid in cids if (cid in A.index and cid not in S_auto)]\n",
    "            if (len(drug_cid) == 0):\n",
    "                continue\n",
    "            for dcid in drug_cid:\n",
    "                if ((organism == \"human\") or (\"human\" in organism)):\n",
    "                    sig = dict(content['down_genes']+content['up_genes'])\n",
    "                elif (organism == \"mouse\"):\n",
    "                    sig = dict([[mouse_human[g],gi] for g,gi in content['down_genes']+content['up_genes'] if (g in mouse_human)])\n",
    "                S_auto.setdefault(dcid, sig)\n",
    "\n",
    "    ## Merge the two datasets (favouring manually curated signatures over the other ones)\n",
    "    from copy import deepcopy\n",
    "    S = deepcopy(S_auto)\n",
    "    S.update(S_manual)\n",
    "    pd.DataFrame(S).fillna(0.)\n",
    "\n",
    "    pd.DataFrame(S).fillna(0.).to_csv(transcript_folder+\"all_drugs.csv\")\n",
    "\n",
    "S = pd.read_csv(transcript_folder+\"all_drugs.csv\", index_col=0)\n",
    "S"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### III.b. Using [LINCS L1000](https://maayanlab.cloud/Harmonizome/resource/LINCS+L1000+Connectivity+Map)\n",
    "\n",
    "In order to populate further matrix S, one can get profiles from LINCS L$1000$ Level $3$ data, and apply Characteristic Direction on them to get signatures similar to the ones in CREEDS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "lincs_folder=\"/media/kali/1b80f30d-2803-4260-a792-9ae206084252/Code/M30/NetworkOrientedRepurposingofDrugs/lincs/\"\n"
     ]
    }
   ],
   "source": [
    "## Where LINCS L1000 files are stored\n",
    "print('lincs_folder=\"%s\"' % paths_global.lincs_folder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1456\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "15744"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "A = pd.read_csv(paths_global.data_folder+\"FEATURELESS_v1.0.0/all_ratings.csv\", engine=\"python\", index_col=0)\n",
    "P = pd.read_csv(transcript_folder+\"all_diseases.csv\", engine=\"python\", index_col=0)\n",
    "S = pd.read_csv(transcript_folder+\"all_drugs.csv\", engine=\"python\", index_col=0)\n",
    "\n",
    "other_drugs = [x for x in A.index if (str(x) not in S.columns)]\n",
    "print(len(other_drugs))\n",
    "## reduce the number of genes to ease computations?\n",
    "var_thres=0\n",
    "genes = list(S.loc[S.var(axis=1) > var_thres].index)\n",
    "genes += list(P.loc[P.var(axis=1) > var_thres].index)\n",
    "genes = list(set(genes))\n",
    "len(genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Convert gene names into EntrezGene CIDs and EntrezGene CIDs into LINCS L1000 gene symbols using package NORDic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15744/15744\n",
      "9/902\n",
      "Not found genes:\n",
      "893/15744\n"
     ]
    }
   ],
   "source": [
    "from NORDic.UTILS.LINCS_utils import *\n",
    "from NORDic.UTILS.utils_data import *\n",
    "\n",
    "## Convert gene names into EntrezGene CID\n",
    "entrezgene_fname=transcript_folder+\"entrezgenes_ids.csv\"\n",
    "if (not os.path.exists(entrezgene_fname)):\n",
    "    probes = request_biodbnet(genes, \"Gene Symbol and Synonyms\", \"Gene ID\", 9606, chunksize=100)\n",
    "    probes.to_csv(entrezgene_fname)\n",
    "probes = pd.read_csv(entrezgene_fname,index_col=0)\n",
    "\n",
    "print(\"%d/%d\" % (probes.shape[0], len(genes)))\n",
    "\n",
    "other_genes = list(probes[probes[\"Gene ID\"]==\"-\"].index)\n",
    "entrezgene_fname=transcript_folder+\"entrezgenes_ids_all.csv\"\n",
    "if (not os.path.exists(entrezgene_fname)):\n",
    "    #list_genes = [data_processing.get_biomart(\"%d/%d\" % (x_id+1,len(other_genes)), x) for x_id, x in enumerate(other_genes)]\n",
    "    list_genes = Parallel(n_jobs=njobs, backend='loky')(delayed(data_processing.get_biomart)(\"%d/%d\" % (x_id+1,len(other_genes)), x) for x_id, x in enumerate(other_genes))\n",
    "    other_probes = pd.DataFrame(list_genes, index=other_genes)\n",
    "    other_probes.to_csv(entrezgene_fname)\n",
    "other_probes = pd.read_csv(entrezgene_fname,index_col=0)\n",
    "other_probes.columns = [\"Gene ID\"]\n",
    "print(\"%d/%d\" % (other_probes[other_probes[\"Gene ID\"]!=\"-\"].shape[0], len(other_genes)))\n",
    "\n",
    "probes = pd.concat((probes.loc[[g for g in genes if ((g.upper() not in other_genes) and (g in probes.index))]], other_probes), axis=0)\n",
    "\n",
    "if (len(list(probes[probes[\"Gene ID\"]==\"-\"].index))>0):\n",
    "    print(\"Not found genes:\")\n",
    "    print(\"%d/%d\" % (len(list(probes[probes[\"Gene ID\"]==\"-\"].index)), len(genes)))\n",
    "probes = probes[probes[\"Gene ID\"]!=\"-\"]\n",
    "\n",
    "model_genes = list(probes.index)\n",
    "\n",
    "## convert EntrezGene ids back to (LINCS L1000) gene symbols\n",
    "if (not os.path.exists(transcript_folder+\"entrezid2symbols.csv\")):\n",
    "    pert_inames = [None]*len(model_genes)\n",
    "    entrez_ids = [None]*len(model_genes)\n",
    "    for ig, g in enumerate(model_genes):\n",
    "        endpoint=\"genes\"\n",
    "        method=\"filter\"\n",
    "        all_entrezid = str(probes.loc[g][\"Gene ID\"]).split(\"; \")\n",
    "        for entrezid in all_entrezid:\n",
    "            params = {\"where\":{\"entrez_id\": str(entrezid)},\"fields\":[\"gene_symbol\"]}\n",
    "            request_url = build_url(endpoint, method, params=params, user_key=user_key)\n",
    "            try:\n",
    "                data = post_request(request_url, quiet=True, stime=0)\n",
    "            except:\n",
    "                continue\n",
    "            if (len(data)==0):\n",
    "                continue\n",
    "            else:\n",
    "                pert_inames[ig] = data[0][\"gene_symbol\"]\n",
    "                entrez_ids[ig] = entrezid\n",
    "                if (pert_inames[ig]==g):\n",
    "                    break\n",
    "        print(\"\\t\".join([g, str(pert_inames[ig]), str(ig+1), str(len(model_genes))]))\n",
    "    pert_iname_ids = [ip for ip,p in enumerate(pert_inames) if (p is not None)]\n",
    "    pd.DataFrame([[pert_inames[i] for i in pert_iname_ids], [entrez_ids[i] for i in pert_iname_ids]], columns=[model_genes[i] for i in pert_iname_ids], index=[\"Gene Symbol\", \"Entrez ID\"]).T.to_csv(sigs_data_folder+\"entrezid2symbols.csv\")\n",
    "pert_df = pd.read_csv(transcript_folder+\"entrezid2symbols.csv\", index_col=0)\n",
    "pert_inames = list(pert_df[\"Gene Symbol\"])\n",
    "entrez_ids = list(pert_df[\"Entrez ID\"])\n",
    "model_genes = list(pert_df.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "93.66 perc. (N=10024)\n",
      "87.47 perc. (N=12038)\n"
     ]
    }
   ],
   "source": [
    "print(\"%.2f perc. (N=%d)\" % (len([g for g in S.index if (g in pert_df.index)])*100/S.shape[0],len([g for g in S.index if (g in pert_df.index)])))\n",
    "print(\"%.2f perc. (N=%d)\" % (len([g for g in P.index if (g in pert_df.index)])*100/P.shape[0],len([g for g in P.index if (g in pert_df.index)])))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Convert missing DrugBank drug names into PubChem CIDs and get associated signatures from LINCS L1000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>5311128</th>\n",
       "      <th>5311498</th>\n",
       "      <th>493570</th>\n",
       "      <th>445354</th>\n",
       "      <th>1054</th>\n",
       "      <th>54687</th>\n",
       "      <th>5324346</th>\n",
       "      <th>5362129</th>\n",
       "      <th>82153</th>\n",
       "      <th>3958</th>\n",
       "      <th>...</th>\n",
       "      <th>5281078</th>\n",
       "      <th>16129706</th>\n",
       "      <th>2310</th>\n",
       "      <th>4375</th>\n",
       "      <th>3352</th>\n",
       "      <th>3373</th>\n",
       "      <th>5917</th>\n",
       "      <th>442872</th>\n",
       "      <th>442021</th>\n",
       "      <th>4843</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>0.004211</td>\n",
       "      <td>-0.007953</td>\n",
       "      <td>0.012509</td>\n",
       "      <td>-0.051398</td>\n",
       "      <td>-0.009290</td>\n",
       "      <td>0.004976</td>\n",
       "      <td>0.040918</td>\n",
       "      <td>0.002669</td>\n",
       "      <td>-0.003704</td>\n",
       "      <td>0.010427</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.011196</td>\n",
       "      <td>-0.004524</td>\n",
       "      <td>0.022839</td>\n",
       "      <td>0.034118</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.027960</td>\n",
       "      <td>-0.005206</td>\n",
       "      <td>0.020435</td>\n",
       "      <td>-0.003454</td>\n",
       "      <td>-0.019129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NAT1</th>\n",
       "      <td>-0.002460</td>\n",
       "      <td>0.020545</td>\n",
       "      <td>-0.021153</td>\n",
       "      <td>0.005485</td>\n",
       "      <td>0.002446</td>\n",
       "      <td>-0.031114</td>\n",
       "      <td>-0.021160</td>\n",
       "      <td>0.019177</td>\n",
       "      <td>0.029274</td>\n",
       "      <td>-0.002868</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.013530</td>\n",
       "      <td>0.007315</td>\n",
       "      <td>-0.006189</td>\n",
       "      <td>0.000281</td>\n",
       "      <td>-0.008344</td>\n",
       "      <td>-0.002534</td>\n",
       "      <td>0.030247</td>\n",
       "      <td>-0.005044</td>\n",
       "      <td>-0.019232</td>\n",
       "      <td>-0.003018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SERPINA3</th>\n",
       "      <td>0.001860</td>\n",
       "      <td>0.012869</td>\n",
       "      <td>-0.009555</td>\n",
       "      <td>-0.003553</td>\n",
       "      <td>0.000099</td>\n",
       "      <td>0.001667</td>\n",
       "      <td>-0.009011</td>\n",
       "      <td>0.023065</td>\n",
       "      <td>0.031211</td>\n",
       "      <td>-0.019435</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.030425</td>\n",
       "      <td>0.008625</td>\n",
       "      <td>-0.000381</td>\n",
       "      <td>0.019200</td>\n",
       "      <td>-0.055678</td>\n",
       "      <td>-0.023840</td>\n",
       "      <td>0.008077</td>\n",
       "      <td>0.000588</td>\n",
       "      <td>0.016774</td>\n",
       "      <td>0.003450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AADAC</th>\n",
       "      <td>0.011732</td>\n",
       "      <td>-0.000800</td>\n",
       "      <td>-0.013999</td>\n",
       "      <td>0.014293</td>\n",
       "      <td>-0.003196</td>\n",
       "      <td>-0.001094</td>\n",
       "      <td>-0.002626</td>\n",
       "      <td>-0.009416</td>\n",
       "      <td>-0.013727</td>\n",
       "      <td>0.002335</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.007428</td>\n",
       "      <td>0.015931</td>\n",
       "      <td>0.003481</td>\n",
       "      <td>-0.013923</td>\n",
       "      <td>0.003363</td>\n",
       "      <td>-0.003875</td>\n",
       "      <td>0.003461</td>\n",
       "      <td>-0.010424</td>\n",
       "      <td>-0.005539</td>\n",
       "      <td>0.009355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAMP</th>\n",
       "      <td>0.000900</td>\n",
       "      <td>0.007262</td>\n",
       "      <td>0.001338</td>\n",
       "      <td>0.009476</td>\n",
       "      <td>0.012429</td>\n",
       "      <td>0.000898</td>\n",
       "      <td>-0.009574</td>\n",
       "      <td>0.007500</td>\n",
       "      <td>-0.001397</td>\n",
       "      <td>0.011513</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.004937</td>\n",
       "      <td>0.017714</td>\n",
       "      <td>-0.007305</td>\n",
       "      <td>0.002826</td>\n",
       "      <td>-0.001552</td>\n",
       "      <td>-0.003753</td>\n",
       "      <td>-0.001625</td>\n",
       "      <td>-0.004519</td>\n",
       "      <td>0.007522</td>\n",
       "      <td>0.013908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JMJD7</th>\n",
       "      <td>-0.004519</td>\n",
       "      <td>-0.008332</td>\n",
       "      <td>-0.004328</td>\n",
       "      <td>0.008560</td>\n",
       "      <td>-0.003966</td>\n",
       "      <td>0.001151</td>\n",
       "      <td>-0.006685</td>\n",
       "      <td>0.002405</td>\n",
       "      <td>-0.001554</td>\n",
       "      <td>-0.001448</td>\n",
       "      <td>...</td>\n",
       "      <td>0.006941</td>\n",
       "      <td>0.002554</td>\n",
       "      <td>-0.010250</td>\n",
       "      <td>-0.002952</td>\n",
       "      <td>0.000698</td>\n",
       "      <td>0.002259</td>\n",
       "      <td>0.002246</td>\n",
       "      <td>0.007328</td>\n",
       "      <td>0.008599</td>\n",
       "      <td>-0.007155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TIMM23</th>\n",
       "      <td>-0.002968</td>\n",
       "      <td>-0.001522</td>\n",
       "      <td>0.003604</td>\n",
       "      <td>-0.003464</td>\n",
       "      <td>0.001232</td>\n",
       "      <td>0.011251</td>\n",
       "      <td>-0.003626</td>\n",
       "      <td>-0.011817</td>\n",
       "      <td>0.010872</td>\n",
       "      <td>-0.002438</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.007431</td>\n",
       "      <td>-0.000748</td>\n",
       "      <td>-0.002906</td>\n",
       "      <td>-0.002370</td>\n",
       "      <td>0.005608</td>\n",
       "      <td>-0.004449</td>\n",
       "      <td>-0.001422</td>\n",
       "      <td>-0.009624</td>\n",
       "      <td>-0.008588</td>\n",
       "      <td>0.002467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF783</th>\n",
       "      <td>-0.001060</td>\n",
       "      <td>-0.009225</td>\n",
       "      <td>-0.002874</td>\n",
       "      <td>-0.005501</td>\n",
       "      <td>-0.007082</td>\n",
       "      <td>-0.000083</td>\n",
       "      <td>0.000620</td>\n",
       "      <td>0.005173</td>\n",
       "      <td>0.000937</td>\n",
       "      <td>0.004329</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.002663</td>\n",
       "      <td>0.002190</td>\n",
       "      <td>-0.008245</td>\n",
       "      <td>0.003452</td>\n",
       "      <td>-0.003703</td>\n",
       "      <td>-0.007072</td>\n",
       "      <td>-0.002508</td>\n",
       "      <td>0.004237</td>\n",
       "      <td>0.002847</td>\n",
       "      <td>-0.007339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>THAP9-AS1</th>\n",
       "      <td>-0.011316</td>\n",
       "      <td>0.011730</td>\n",
       "      <td>0.008664</td>\n",
       "      <td>0.000761</td>\n",
       "      <td>0.012618</td>\n",
       "      <td>-0.004360</td>\n",
       "      <td>-0.005032</td>\n",
       "      <td>-0.011883</td>\n",
       "      <td>-0.006719</td>\n",
       "      <td>0.008573</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.016185</td>\n",
       "      <td>0.006754</td>\n",
       "      <td>-0.006451</td>\n",
       "      <td>-0.002437</td>\n",
       "      <td>-0.000758</td>\n",
       "      <td>-0.009862</td>\n",
       "      <td>0.017386</td>\n",
       "      <td>-0.005379</td>\n",
       "      <td>0.001778</td>\n",
       "      <td>0.017479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MICA</th>\n",
       "      <td>0.005473</td>\n",
       "      <td>0.001818</td>\n",
       "      <td>-0.011232</td>\n",
       "      <td>-0.002826</td>\n",
       "      <td>-0.002988</td>\n",
       "      <td>0.001228</td>\n",
       "      <td>0.000956</td>\n",
       "      <td>0.005389</td>\n",
       "      <td>0.002518</td>\n",
       "      <td>-0.009410</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.001467</td>\n",
       "      <td>0.020665</td>\n",
       "      <td>0.000645</td>\n",
       "      <td>-0.001116</td>\n",
       "      <td>0.003404</td>\n",
       "      <td>-0.008615</td>\n",
       "      <td>0.001577</td>\n",
       "      <td>-0.004617</td>\n",
       "      <td>-0.009393</td>\n",
       "      <td>-0.004475</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10071 rows × 725 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            5311128   5311498    493570    445354      1054     54687   \n",
       "A2M        0.004211 -0.007953  0.012509 -0.051398 -0.009290  0.004976  \\\n",
       "NAT1      -0.002460  0.020545 -0.021153  0.005485  0.002446 -0.031114   \n",
       "SERPINA3   0.001860  0.012869 -0.009555 -0.003553  0.000099  0.001667   \n",
       "AADAC      0.011732 -0.000800 -0.013999  0.014293 -0.003196 -0.001094   \n",
       "AAMP       0.000900  0.007262  0.001338  0.009476  0.012429  0.000898   \n",
       "...             ...       ...       ...       ...       ...       ...   \n",
       "JMJD7     -0.004519 -0.008332 -0.004328  0.008560 -0.003966  0.001151   \n",
       "TIMM23    -0.002968 -0.001522  0.003604 -0.003464  0.001232  0.011251   \n",
       "ZNF783    -0.001060 -0.009225 -0.002874 -0.005501 -0.007082 -0.000083   \n",
       "THAP9-AS1 -0.011316  0.011730  0.008664  0.000761  0.012618 -0.004360   \n",
       "MICA       0.005473  0.001818 -0.011232 -0.002826 -0.002988  0.001228   \n",
       "\n",
       "            5324346   5362129     82153      3958  ...   5281078  16129706   \n",
       "A2M        0.040918  0.002669 -0.003704  0.010427  ... -0.011196 -0.004524  \\\n",
       "NAT1      -0.021160  0.019177  0.029274 -0.002868  ... -0.013530  0.007315   \n",
       "SERPINA3  -0.009011  0.023065  0.031211 -0.019435  ... -0.030425  0.008625   \n",
       "AADAC     -0.002626 -0.009416 -0.013727  0.002335  ... -0.007428  0.015931   \n",
       "AAMP      -0.009574  0.007500 -0.001397  0.011513  ... -0.004937  0.017714   \n",
       "...             ...       ...       ...       ...  ...       ...       ...   \n",
       "JMJD7     -0.006685  0.002405 -0.001554 -0.001448  ...  0.006941  0.002554   \n",
       "TIMM23    -0.003626 -0.011817  0.010872 -0.002438  ... -0.007431 -0.000748   \n",
       "ZNF783     0.000620  0.005173  0.000937  0.004329  ... -0.002663  0.002190   \n",
       "THAP9-AS1 -0.005032 -0.011883 -0.006719  0.008573  ... -0.016185  0.006754   \n",
       "MICA       0.000956  0.005389  0.002518 -0.009410  ... -0.001467  0.020665   \n",
       "\n",
       "               2310      4375      3352      3373      5917    442872   \n",
       "A2M        0.022839  0.034118  0.000014  0.027960 -0.005206  0.020435  \\\n",
       "NAT1      -0.006189  0.000281 -0.008344 -0.002534  0.030247 -0.005044   \n",
       "SERPINA3  -0.000381  0.019200 -0.055678 -0.023840  0.008077  0.000588   \n",
       "AADAC      0.003481 -0.013923  0.003363 -0.003875  0.003461 -0.010424   \n",
       "AAMP      -0.007305  0.002826 -0.001552 -0.003753 -0.001625 -0.004519   \n",
       "...             ...       ...       ...       ...       ...       ...   \n",
       "JMJD7     -0.010250 -0.002952  0.000698  0.002259  0.002246  0.007328   \n",
       "TIMM23    -0.002906 -0.002370  0.005608 -0.004449 -0.001422 -0.009624   \n",
       "ZNF783    -0.008245  0.003452 -0.003703 -0.007072 -0.002508  0.004237   \n",
       "THAP9-AS1 -0.006451 -0.002437 -0.000758 -0.009862  0.017386 -0.005379   \n",
       "MICA       0.000645 -0.001116  0.003404 -0.008615  0.001577 -0.004617   \n",
       "\n",
       "             442021      4843  \n",
       "A2M       -0.003454 -0.019129  \n",
       "NAT1      -0.019232 -0.003018  \n",
       "SERPINA3   0.016774  0.003450  \n",
       "AADAC     -0.005539  0.009355  \n",
       "AAMP       0.007522  0.013908  \n",
       "...             ...       ...  \n",
       "JMJD7      0.008599 -0.007155  \n",
       "TIMM23    -0.008588  0.002467  \n",
       "ZNF783     0.002847 -0.007339  \n",
       "THAP9-AS1  0.001778  0.017479  \n",
       "MICA      -0.009393 -0.004475  \n",
       "\n",
       "[10071 rows x 725 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if (not os.path.exists(transcript_folder+\"save_signatures.csv\")):\n",
    "    other_drugs_converted = [di_drugbankid2drugname.get(d, d) for d in other_drugs]\n",
    "    inv_map = {v: k for k, v in di_pubchemid2drugname.items()}\n",
    "    for idx, d in enumerate(other_drugs):\n",
    "        dx = other_drugs_converted[idx]\n",
    "        try:\n",
    "            other_drugs_converted[idx] = int(dx) if (int(dx) in di_pubchemid2drugname) else inv_map.get(dx, None)\n",
    "        except:\n",
    "            other_drugs_converted[idx] = inv_map.get(dx, None)\n",
    "    other_drugs_converted = [dx for dx in other_drugs_converted if (dx is not None)]\n",
    "    \"%.2f perc. (N=%d)\" % (len(other_drugs_converted)*100/len(other_drugs),len(other_drugs_converted))\n",
    "\n",
    "    signatures = get_profiles(other_drugs_converted, entrez_ids, path_to_lincs=path_to_lincs, nsigs=2, verbose=0)\n",
    "    signatures.to_csv(transcript_folder+\"save_signatures.csv\")\n",
    "    \n",
    "signatures = pd.read_csv(transcript_folder+\"save_signatures.csv\", index_col=0)\n",
    "ids = [pert_df.iloc[list(pert_df[\"Entrez ID\"]).index(int(idx)),:][\"Gene Symbol\"] for idx in signatures.index]\n",
    "signatures.index = ids\n",
    "signatures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>36314</th>\n",
       "      <th>11354606</th>\n",
       "      <th>679</th>\n",
       "      <th>2554</th>\n",
       "      <th>126941</th>\n",
       "      <th>3033</th>\n",
       "      <th>444795</th>\n",
       "      <th>5757</th>\n",
       "      <th>19649</th>\n",
       "      <th>6279</th>\n",
       "      <th>...</th>\n",
       "      <th>5281078</th>\n",
       "      <th>16129706</th>\n",
       "      <th>2310</th>\n",
       "      <th>4375</th>\n",
       "      <th>3352</th>\n",
       "      <th>3373</th>\n",
       "      <th>5917</th>\n",
       "      <th>442872</th>\n",
       "      <th>442021</th>\n",
       "      <th>4843</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.001028</td>\n",
       "      <td>-0.000363</td>\n",
       "      <td>0.004807</td>\n",
       "      <td>-0.000522</td>\n",
       "      <td>-0.000997</td>\n",
       "      <td>0.008423</td>\n",
       "      <td>0.002083</td>\n",
       "      <td>0.005748</td>\n",
       "      <td>0.001114</td>\n",
       "      <td>0.014441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.011196</td>\n",
       "      <td>-0.004524</td>\n",
       "      <td>0.022839</td>\n",
       "      <td>0.034118</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.027960</td>\n",
       "      <td>-0.005206</td>\n",
       "      <td>0.020435</td>\n",
       "      <td>-0.003454</td>\n",
       "      <td>-0.019129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAS</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.001392</td>\n",
       "      <td>-0.001070</td>\n",
       "      <td>-0.009262</td>\n",
       "      <td>0.001435</td>\n",
       "      <td>-0.001956</td>\n",
       "      <td>0.003047</td>\n",
       "      <td>0.001742</td>\n",
       "      <td>0.004964</td>\n",
       "      <td>0.012373</td>\n",
       "      <td>0.006155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACS</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003157</td>\n",
       "      <td>-0.003778</td>\n",
       "      <td>-0.015237</td>\n",
       "      <td>0.007205</td>\n",
       "      <td>-0.012258</td>\n",
       "      <td>-0.006826</td>\n",
       "      <td>-0.003272</td>\n",
       "      <td>-0.001598</td>\n",
       "      <td>0.001768</td>\n",
       "      <td>0.007927</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.008635</td>\n",
       "      <td>0.012023</td>\n",
       "      <td>-0.010802</td>\n",
       "      <td>-0.011957</td>\n",
       "      <td>0.007483</td>\n",
       "      <td>-0.000572</td>\n",
       "      <td>-0.003034</td>\n",
       "      <td>-0.007464</td>\n",
       "      <td>0.006834</td>\n",
       "      <td>-0.009079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000688</td>\n",
       "      <td>-0.001199</td>\n",
       "      <td>0.000477</td>\n",
       "      <td>-0.005147</td>\n",
       "      <td>-0.016044</td>\n",
       "      <td>-0.003332</td>\n",
       "      <td>-0.000163</td>\n",
       "      <td>-0.001941</td>\n",
       "      <td>-0.005947</td>\n",
       "      <td>0.006733</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.010314</td>\n",
       "      <td>0.003458</td>\n",
       "      <td>-0.009071</td>\n",
       "      <td>0.011935</td>\n",
       "      <td>0.011555</td>\n",
       "      <td>-0.004139</td>\n",
       "      <td>0.010807</td>\n",
       "      <td>0.000995</td>\n",
       "      <td>0.007324</td>\n",
       "      <td>0.003029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cra</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12942 rows × 871 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        36314  11354606  679  2554  126941  3033  444795  5757  19649  6279   \n",
       "A1BG      0.0       0.0  0.0   0.0     0.0   0.0     0.0   0.0    0.0   0.0  \\\n",
       "A1CF      0.0       0.0  0.0   0.0     0.0   0.0     0.0   0.0    0.0   0.0   \n",
       "A2M       0.0       0.0  0.0   0.0     0.0   0.0     0.0   0.0    0.0   0.0   \n",
       "AAAS      0.0       0.0  0.0   0.0     0.0   0.0     0.0   0.0    0.0   0.0   \n",
       "AACS      0.0       0.0  0.0   0.0     0.0   0.0     0.0   0.0    0.0   0.0   \n",
       "...       ...       ...  ...   ...     ...   ...     ...   ...    ...   ...   \n",
       "ZYG11B    0.0       0.0  0.0   0.0     0.0   0.0     0.0   0.0    0.0   0.0   \n",
       "ZYX       0.0       0.0  0.0   0.0     0.0   0.0     0.0   0.0    0.0   0.0   \n",
       "ZZEF1     0.0       0.0  0.0   0.0     0.0   0.0     0.0   0.0    0.0   0.0   \n",
       "ZZZ3      0.0       0.0  0.0   0.0     0.0   0.0     0.0   0.0    0.0   0.0   \n",
       "cra       0.0       0.0  0.0   0.0     0.0   0.0     0.0   0.0    0.0   0.0   \n",
       "\n",
       "        ...   5281078  16129706      2310      4375      3352      3373   \n",
       "A1BG    ...  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \\\n",
       "A1CF    ... -0.001028 -0.000363  0.004807 -0.000522 -0.000997  0.008423   \n",
       "A2M     ... -0.011196 -0.004524  0.022839  0.034118  0.000014  0.027960   \n",
       "AAAS    ... -0.001392 -0.001070 -0.009262  0.001435 -0.001956  0.003047   \n",
       "AACS    ...  0.003157 -0.003778 -0.015237  0.007205 -0.012258 -0.006826   \n",
       "...     ...       ...       ...       ...       ...       ...       ...   \n",
       "ZYG11B  ...  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "ZYX     ...  0.008635  0.012023 -0.010802 -0.011957  0.007483 -0.000572   \n",
       "ZZEF1   ... -0.000688 -0.001199  0.000477 -0.005147 -0.016044 -0.003332   \n",
       "ZZZ3    ... -0.010314  0.003458 -0.009071  0.011935  0.011555 -0.004139   \n",
       "cra     ...  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "\n",
       "            5917    442872    442021      4843  \n",
       "A1BG    0.000000  0.000000  0.000000  0.000000  \n",
       "A1CF    0.002083  0.005748  0.001114  0.014441  \n",
       "A2M    -0.005206  0.020435 -0.003454 -0.019129  \n",
       "AAAS    0.001742  0.004964  0.012373  0.006155  \n",
       "AACS   -0.003272 -0.001598  0.001768  0.007927  \n",
       "...          ...       ...       ...       ...  \n",
       "ZYG11B  0.000000  0.000000  0.000000  0.000000  \n",
       "ZYX    -0.003034 -0.007464  0.006834 -0.009079  \n",
       "ZZEF1  -0.000163 -0.001941 -0.005947  0.006733  \n",
       "ZZZ3    0.010807  0.000995  0.007324  0.003029  \n",
       "cra     0.000000  0.000000  0.000000  0.000000  \n",
       "\n",
       "[12942 rows x 871 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if (not os.path.exists(transcript_folder+\"all_drugs_+LINCS.csv\")):\n",
    "    S_LINCS = S[[c for c in S.columns if (c not in signatures.columns)]].join(signatures, how=\"outer\")\n",
    "    S_LINCS.fillna(0).to_csv(transcript_folder+\"all_drugs_+LINCS.csv\")\n",
    "    \n",
    "S_LINCS = pd.read_csv(transcript_folder+\"all_drugs_+LINCS.csv\", index_col=0)\n",
    "S_LINCS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## IV. Merge A, P and S to get the final \"TRANSCRIPT\" dataset\n",
    "\n",
    "Disease ids: Concept IDs, drug ids: DrugBank or PubChem CIDs. Build matrices A (ratings) drugs $\\times$ diseases, P (disease features) disease features $\\times$ diseases, S (drug features) drug features $\\times$ drugs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sparsity = 0.7606199770378874%\n",
      "10811 drug features 10811 disease features\n",
      "Ndrugs=558\tNdiseases=118\n",
      "773 positive\t181 negative\t64890 unknown matchings\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>944</th>\n",
       "      <th>945</th>\n",
       "      <th>946</th>\n",
       "      <th>947</th>\n",
       "      <th>948</th>\n",
       "      <th>949</th>\n",
       "      <th>950</th>\n",
       "      <th>951</th>\n",
       "      <th>952</th>\n",
       "      <th>953</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ind_id</th>\n",
       "      <td>C0020538</td>\n",
       "      <td>C1140680</td>\n",
       "      <td>C0007131</td>\n",
       "      <td>C0036220</td>\n",
       "      <td>C0236780</td>\n",
       "      <td>C0036341</td>\n",
       "      <td>C0014544</td>\n",
       "      <td>C2936783</td>\n",
       "      <td>C0033860</td>\n",
       "      <td>C0149925</td>\n",
       "      <td>...</td>\n",
       "      <td>C0014544</td>\n",
       "      <td>C0023418</td>\n",
       "      <td>C0014544</td>\n",
       "      <td>C0014544</td>\n",
       "      <td>C0014544</td>\n",
       "      <td>C0014544</td>\n",
       "      <td>C0014544</td>\n",
       "      <td>C0014544</td>\n",
       "      <td>C0014544</td>\n",
       "      <td>C0014544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>drug_id</th>\n",
       "      <td>36314</td>\n",
       "      <td>36314</td>\n",
       "      <td>36314</td>\n",
       "      <td>36314</td>\n",
       "      <td>2554</td>\n",
       "      <td>2554</td>\n",
       "      <td>2554</td>\n",
       "      <td>126941</td>\n",
       "      <td>126941</td>\n",
       "      <td>126941</td>\n",
       "      <td>...</td>\n",
       "      <td>9915886</td>\n",
       "      <td>5281078</td>\n",
       "      <td>2310</td>\n",
       "      <td>4375</td>\n",
       "      <td>3352</td>\n",
       "      <td>3373</td>\n",
       "      <td>5917</td>\n",
       "      <td>442872</td>\n",
       "      <td>442021</td>\n",
       "      <td>4843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rating</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 954 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              0         1         2         3         4         5         6     \n",
       "ind_id   C0020538  C1140680  C0007131  C0036220  C0236780  C0036341  C0014544  \\\n",
       "drug_id     36314     36314     36314     36314      2554      2554      2554   \n",
       "rating        1.0       1.0       1.0       1.0       1.0       1.0       1.0   \n",
       "\n",
       "              7         8         9    ...       944       945       946   \n",
       "ind_id   C2936783  C0033860  C0149925  ...  C0014544  C0023418  C0014544  \\\n",
       "drug_id    126941    126941    126941  ...   9915886   5281078      2310   \n",
       "rating        1.0       1.0       1.0  ...      -1.0      -1.0      -1.0   \n",
       "\n",
       "              947       948       949       950       951       952       953  \n",
       "ind_id   C0014544  C0014544  C0014544  C0014544  C0014544  C0014544  C0014544  \n",
       "drug_id      4375      3352      3373      5917    442872    442021      4843  \n",
       "rating       -1.0      -1.0      -1.0      -1.0      -1.0      -1.0       1.0  \n",
       "\n",
       "[3 rows x 954 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "A = pd.read_csv(paths_global.data_folder+\"FEATURELESS_v1.0.0/all_ratings.csv\", engine=\"python\", index_col=0)\n",
    "P = pd.read_csv(transcript_folder+\"all_diseases.csv\", engine=\"python\", index_col=0)\n",
    "S = pd.read_csv(transcript_folder+\"all_drugs_+LINCS.csv\", engine=\"python\", index_col=0)\n",
    "\n",
    "A_TRANSCRIPT = A[P.columns].loc[[c for c in S.columns if (c in A.index)]]\n",
    "SP = S.join(P, how=\"inner\").fillna(0)\n",
    "S_TRANSCRIPT = SP[S.columns]\n",
    "P_TRANSCRIPT = SP[P.columns]\n",
    "\n",
    "S_TRANSCRIPT.to_csv(transcript_folder+\"items.csv\")\n",
    "P_TRANSCRIPT.to_csv(transcript_folder+\"users.csv\")\n",
    "A_TRANSCRIPT.to_csv(transcript_folder+\"ratings_mat.csv\")\n",
    "\n",
    "ratings_A = utils.matrix2ratings(A_TRANSCRIPT, \"ind_id\", \"drug_id\", \"rating\")\n",
    "print(\"Sparsity = \"+str(utils.compute_sparsity(A_TRANSCRIPT))+\"%\")\n",
    "print(\"%d drug features %d disease features\" % (S_TRANSCRIPT.shape[0], P_TRANSCRIPT.shape[0]))\n",
    "utils.print_dataset(ratings_A, \"ind_id\", \"drug_id\", \"rating\")\n",
    "ratings_A.T"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
